diff --git a/.devcontainer/Dockerfile b/.devcontainer/Dockerfile
index a1b5cfd6be..94497f52fb 100644
--- a/.devcontainer/Dockerfile
+++ b/.devcontainer/Dockerfile
@@ -1,5 +1,5 @@
 FROM ubuntu:24.04
-LABEL org.opencontainers.image.source=https://github.com/riscv-software-src/riscv-unified-db
+LABEL org.opencontainers.image.source=https://github.com/riscv/riscv-unified-db
 
 ENV DEBIAN_FRONTEND=noninteractive
 
diff --git a/bin/.container-tag b/bin/.container-tag
index a4d2aceeb8..caa4836d8e 100644
--- a/bin/.container-tag
+++ b/bin/.container-tag
@@ -1 +1 @@
-0.18
+0.19
