diff --git a/iopmp_ref_model/include/iopmp.h b/iopmp_ref_model/include/iopmp.h
index 505f88c..81ab8b0 100644
--- a/iopmp_ref_model/include/iopmp.h
+++ b/iopmp_ref_model/include/iopmp.h
@@ -107,13 +107,11 @@ typedef struct iopmp_cfg_t {
     bool imp_msi;                       // IOPMP implements message-signaled interrupts (MSI)
 } iopmp_cfg_t;
 
-// Enumerates specific match and error statuses for transactions
+// Enumerates specific match statuses for transactions
 typedef enum {
     ENTRY_NOTMATCH,                     // Entry doesn't match any byte of a transation
     ENTRY_PARTIAL_MATCH,                // Entry matches partial bytes of a transation
     ENTRY_MATCH,                        // Entry matches all bytes of a transation
-    ENTRY_MATCH_NOTGRANT,               // Entry matches all bytes of a transaction, but doesn't grant transaction permission
-    ENTRY_MATCH_GRANT,                  // Entry matches all bytes of a transaction and grants transaction permission
 } iopmpMatchStatus_t;
 
 // Enumerates the type of violation for transactions
@@ -128,10 +126,30 @@ typedef enum {
     STALLED_TRANSACTION     = 0x07,     // Error due to a stalled transaction
 } iopmpErrorType_t;
 
+// The information the rule analyzer needs
+typedef struct iopmp_rule_analyzer_input_t {
+    uint16_t rrid;
+    uint64_t prev_iopmpaddr;
+    uint64_t iopmpaddr;
+    entry_cfg_t iopmpcfg;
+    uint64_t trans_start;
+    uint64_t trans_end;
+    perm_type_e perm;
+    bool is_amo;
+    uint8_t md;
+} iopmp_rule_analyzer_input_t;
+
+// The output of the rule analyzer
+typedef struct iopmp_rule_analyzer_output_t {
+    iopmpMatchStatus_t match_status;
+    bool grant_perm;
+} iopmp_rule_analyzer_output_t;
+
 uint8_t write_memory(uint64_t *data, uint64_t addr, uint32_t size);
 
 // Function Declarations: Core IOPMP operations
-iopmpMatchStatus_t iopmpRuleAnalyzer(iopmp_dev_t *iopmp, iopmp_trans_req_t trans_req, uint64_t prev_iopmpaddr, uint64_t iopmpaddr, entry_cfg_t iopmpcfg, uint8_t md);
+void iopmpRuleAnalyzer(iopmp_dev_t *iopmp, iopmp_rule_analyzer_input_t *input,
+                       iopmp_rule_analyzer_output_t *output);
 void errorCapture(iopmp_dev_t *iopmp, perm_type_e trans_type, uint8_t error_type, uint16_t rrid, uint16_t entry_id, uint64_t err_addr, uint8_t *intrpt);
 void generate_interrupt(iopmp_dev_t *iopmp, uint8_t *intrpt);
 
diff --git a/iopmp_ref_model/src/iopmp_reg.c b/iopmp_ref_model/src/iopmp_reg.c
index b2837b1..99c82d3 100644
--- a/iopmp_ref_model/src/iopmp_reg.c
+++ b/iopmp_ref_model/src/iopmp_reg.c
@@ -98,6 +98,8 @@ int reset_iopmp(iopmp_dev_t *iopmp, iopmp_cfg_t *cfg)
 
     // Hardware Configuration
     iopmp->reg_file.hwcfg0.enable           = cfg->enable;
+    iopmp->reg_file.hwcfg0.HWCFG2_en        = 1;    // HWCFG2 is always implemented by reference model
+    iopmp->reg_file.hwcfg0.HWCFG3_en        = 1;    // HWCFG3 is always implemented by reference model
     iopmp->reg_file.hwcfg0.md_num           = cfg->md_num;
     iopmp->reg_file.hwcfg0.addrh_en         = cfg->addrh_en;
     iopmp->reg_file.hwcfg0.tor_en           = cfg->tor_en;
@@ -114,8 +116,6 @@ int reset_iopmp(iopmp_dev_t *iopmp, iopmp_cfg_t *cfg)
     iopmp->reg_file.hwcfg2.sps_en           = cfg->sps_en;
     iopmp->reg_file.hwcfg2.stall_en         = cfg->stall_en;
     iopmp->reg_file.hwcfg2.mfr_en           = cfg->mfr_en;
-    /* Set HWCFG2_en if HWCFG2 is not zero */
-    iopmp->reg_file.hwcfg0.HWCFG2_en        = (iopmp->reg_file.hwcfg2.raw) != 0 ? true : false;
 
     iopmp->reg_file.hwcfg3.mdcfg_fmt        = cfg->mdcfg_fmt;
     iopmp->reg_file.hwcfg3.srcmd_fmt        = cfg->srcmd_fmt;
@@ -129,8 +129,6 @@ int reset_iopmp(iopmp_dev_t *iopmp, iopmp_cfg_t *cfg)
         iopmp->reg_file.hwcfg3.rrid_transl_prog = cfg->rrid_transl_prog;
         iopmp->reg_file.hwcfg3.rrid_transl      = cfg->rrid_transl;
     }
-    /* Set HWCFG3_en if HWCFG3 is not zero */
-    iopmp->reg_file.hwcfg0.HWCFG3_en        = (iopmp->reg_file.hwcfg3.raw) != 0 ? true : false;
 
     iopmp->reg_file.entryoffset.offset      = cfg->entryoffset;
 
diff --git a/iopmp_ref_model/src/iopmp_rule_analyzer.c b/iopmp_ref_model/src/iopmp_rule_analyzer.c
index 3bfc4b6..51f0318 100644
--- a/iopmp_ref_model/src/iopmp_rule_analyzer.c
+++ b/iopmp_ref_model/src/iopmp_rule_analyzer.c
@@ -70,7 +70,8 @@ static int iopmpAddrRange(uint64_t *startAddr, uint64_t *endAddr, uint64_t prev_
 /**
   * @brief Determine the matching status of transaction request address against IOPMP entry range.
   *
-  * @param trans_req Transaction request containing address, size, and length
+  * @param trans_start Start address of the transaction
+  * @param trans_end End address of the transaction
   * @param lo Lower bound of the IOPMP range
   * @param hi Upper bound of the IOPMP range
   * @return iopmpMatchStatus_t Status of the match:
@@ -78,18 +79,17 @@ static int iopmpAddrRange(uint64_t *startAddr, uint64_t *endAddr, uint64_t prev_
   *         - ENTRY_PARTIAL_MATCH: Entry matches partial bytes of a transation
   *         - ENTRY_NOTMATCH: Entry doesn't cover any byte of the transaction
  **/
-static iopmpMatchStatus_t iopmpMatchAddr(iopmp_trans_req_t trans_req, uint64_t lo, uint64_t hi) {
+static iopmpMatchStatus_t iopmpMatchAddr(uint64_t trans_start,
+                                         uint64_t trans_end,
+                                         uint64_t lo, uint64_t hi) {
     // Validate range
     if (hi < lo) { return ENTRY_NOTMATCH; }  // Invalid range, no match
 
-    // Compute the end address of the transaction
-    uint64_t trans_end = trans_req.addr + ((int)pow(2, trans_req.size) * (trans_req.length + 1));
-
     // Check if transaction falls outside the IOPMP entry range
-    if (trans_end <= lo || trans_req.addr >= hi) { return ENTRY_NOTMATCH; } // No match, transaction outside range
+    if (trans_end <= lo || trans_start >= hi) { return ENTRY_NOTMATCH; } // No match, transaction outside range
 
     // Determine if there's a full match
-    if (trans_req.addr >= lo && trans_end <= hi) { return ENTRY_MATCH; }
+    if (trans_start >= lo && trans_end <= hi) { return ENTRY_MATCH; }
 
     return ENTRY_PARTIAL_MATCH;
 }
@@ -103,11 +103,10 @@ static iopmpMatchStatus_t iopmpMatchAddr(iopmp_trans_req_t trans_req, uint64_t l
   * @param iopmpcfg IOPMP configuration for the entry
   * @param md Respective Memory Domain
   * @param is_amo Indicates the AMO Access
-  * @return iopmpMatchStatus_t Status of the match and permission check:
-  *         - ENTRY_MATCH_GRANT: Entry matches all bytes of a transaction and grants transation permission
-  *         - ENTRY_MATCH_NOTGRANT: Entry matches all bytes of a transaction but doesn't grant transation permission
+  * @return - true if entry grants transation permission
+  *         - false if entry doesn't grant transation permission
  **/
-static iopmpMatchStatus_t iopmpCheckPerms(iopmp_dev_t *iopmp, uint16_t rrid, perm_type_e req_perm, entry_cfg_t iopmpcfg, uint8_t md, bool is_amo) {
+static bool iopmpCheckPerms(iopmp_dev_t *iopmp, uint16_t rrid, perm_type_e req_perm, entry_cfg_t iopmpcfg, uint8_t md, bool is_amo) {
     // Common permission checks
     bool read_allowed    = false;
     bool write_allowed   = false;
@@ -156,14 +155,14 @@ static iopmpMatchStatus_t iopmpCheckPerms(iopmp_dev_t *iopmp, uint16_t rrid, per
             iopmp->intrpt_suppress = iopmpcfg.sire;
             iopmp->error_suppress  = iopmpcfg.sere | iopmp->reg_file.err_cfg.rs;
         }
-        return read_allowed ? ENTRY_MATCH_GRANT : ENTRY_MATCH_NOTGRANT;
+        return read_allowed;
 
     case WRITE_ACCESS:
         if (!write_allowed) {
             iopmp->intrpt_suppress = iopmpcfg.siwe;
             iopmp->error_suppress  = iopmpcfg.sewe | iopmp->reg_file.err_cfg.rs;
         }
-        return write_allowed ? ENTRY_MATCH_GRANT : ENTRY_MATCH_NOTGRANT;
+        return write_allowed;
 
     case INSTR_FETCH:
         if (iopmp->reg_file.hwcfg2.chk_x) {
@@ -171,16 +170,12 @@ static iopmpMatchStatus_t iopmpCheckPerms(iopmp_dev_t *iopmp, uint16_t rrid, per
                 iopmp->intrpt_suppress = iopmpcfg.sixe;
                 iopmp->error_suppress  = iopmpcfg.sexe | iopmp->reg_file.err_cfg.rs;
             }
-            return execute_allowed ? ENTRY_MATCH_GRANT : ENTRY_MATCH_NOTGRANT;
-        } else if (read_allowed) {
-            return ENTRY_MATCH_GRANT;   // Grant Execute permission via Read fallback
+            return execute_allowed;
         }
-        iopmp->intrpt_suppress = iopmpcfg.sixe;
-        iopmp->error_suppress  = iopmpcfg.sexe | iopmp->reg_file.err_cfg.rs;
-        return ENTRY_MATCH_NOTGRANT;
+        return false;   // Instruction fetch check not implemented
 
     default:
-        return ENTRY_MATCH_NOTGRANT;    // Default case for invalid permission request
+        return false;   // Default case for invalid permission request
     }
 }
 
@@ -188,40 +183,34 @@ static iopmpMatchStatus_t iopmpCheckPerms(iopmp_dev_t *iopmp, uint16_t rrid, per
   * @brief Analyze the matching status and permissions of a transaction request
   *
   * @param iopmp The IOPMP instance.
-  * @param trans_req Transaction request containing address, permissions, etc.
-  * @param prev_iopmpaddr Previous IOPMP address (used in TOR mode)
-  * @param iopmpaddr Current IOPMP address
-  * @param iopmpcfg IOPMP entry configuration
-  * @param md Respective Memory Domain
-  * @return iopmpMatchStatus_t Status of the match and permission check:
-  *         - ENTRY_NOTMATCH: Entry doesn't cover any byte of the transaction
-  *         - ENTRY_PARTIAL_MATCH: Entry matches partial bytes of a transation
-  *         - ENTRY_MATCH_NOTGRANT: Entry matches all bytes of a transaction, but doesn't grant transaction permission
-  *         - ENTRY_MATCH_GRANT: Entry matches all bytes of a transaction and grants transaction permission
+  * @param input All the information the analyzer needs.
+  * @param output Address matching status and permission grant result.
  **/
-iopmpMatchStatus_t iopmpRuleAnalyzer(iopmp_dev_t *iopmp, iopmp_trans_req_t trans_req, uint64_t prev_iopmpaddr, uint64_t iopmpaddr, entry_cfg_t iopmpcfg, uint8_t md) {
-    iopmpMatchStatus_t match_status = ENTRY_MATCH;  // Default to full match
+void iopmpRuleAnalyzer(iopmp_dev_t *iopmp,
+                       iopmp_rule_analyzer_input_t *input,
+                       iopmp_rule_analyzer_output_t *output)
+{
     uint64_t start_addr, end_addr;
 
     // Set up the address range; if range setup fails, return not matched
-    if (iopmpAddrRange(&start_addr, &end_addr, prev_iopmpaddr, iopmpaddr, iopmpcfg)) { return ENTRY_NOTMATCH; }
+    if (iopmpAddrRange(&start_addr, &end_addr, input->prev_iopmpaddr,
+                       input->iopmpaddr, input->iopmpcfg)) {
+        output->match_status = ENTRY_NOTMATCH;
+        return;
+    }
 
     // Match transaction address to the IOPMP address range
     // Multiply the addresses with 4 is equivalent to (Address << 2)
-    iopmpMatchStatus_t addr_match_status = iopmpMatchAddr(trans_req, (start_addr * 4), (end_addr * 4));
-    if (addr_match_status == ENTRY_NOTMATCH || addr_match_status == ENTRY_PARTIAL_MATCH) {
+    output->match_status = iopmpMatchAddr(input->trans_start, input->trans_end,
+                                          (start_addr * 4), (end_addr * 4));
+    if (output->match_status == ENTRY_NOTMATCH ||
+        output->match_status == ENTRY_PARTIAL_MATCH) {
         // It's unnecessary to check entry permissions in these two cases
-        return addr_match_status;
+        return;
     }
 
     // iopmpMatchAddr() returns ENTRY_MATCH. Further checks entry permissions
-    #if (SRC_ENFORCEMENT_EN)
-        match_status = iopmpCheckPerms(iopmp, 0, trans_req.perm, iopmpcfg, md, trans_req.is_amo);
-    #else
-    // Check access permissions
-        match_status = iopmpCheckPerms(iopmp, trans_req.rrid, trans_req.perm, iopmpcfg, md, trans_req.is_amo);
-    #endif
-
-    // If all checks pass, return full match status
-    return match_status;
+    output->grant_perm = iopmpCheckPerms(iopmp, input->rrid, input->perm,
+                                         input->iopmpcfg, input->md,
+                                         input->is_amo);
 }
diff --git a/iopmp_ref_model/src/iopmp_validate.c b/iopmp_ref_model/src/iopmp_validate.c
index b6d3d52..8190e9e 100644
--- a/iopmp_ref_model/src/iopmp_validate.c
+++ b/iopmp_ref_model/src/iopmp_validate.c
@@ -61,6 +61,13 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
         assert(trans_req->is_amo == 0);
     }
 
+#if (SRC_ENFORCEMENT_EN == 1)
+    // Enforce RRID=0 for Source-Enforcement
+    const uint16_t rrid = 0;
+#else
+    const uint16_t rrid = trans_req->rrid;
+#endif
+    perm_type_e trans_perm = trans_req->perm;
     iopmpErrorType_t error_type = NO_ERROR;
     uint16_t error_eid = 0;
 
@@ -71,7 +78,6 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
     srcmd_en_t  srcmd_en;
     srcmd_enh_t srcmd_enh;
 
-    iopmpMatchStatus_t iopmpMatchStatus;
     int nonPrioErrorSup    = 0;
     int nonPrioIntrSup     = 0;
     int firstIllegalAccess = 1;
@@ -90,14 +96,14 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
     }
 
     // Check for valid RRID; if invalid, capture error and return
-    if (trans_req->rrid >= iopmp->reg_file.hwcfg1.rrid_num) {
+    if (rrid >= iopmp->reg_file.hwcfg1.rrid_num) {
         // Initially, check for global error suppression
         iopmp->error_suppress = iopmp->reg_file.err_cfg.rs;
         error_type = UNKNOWN_RRID;
         goto stop_and_report_fault;
     }
 
-    if (iopmp->rrid_stall[trans_req->rrid]) {
+    if (iopmp->rrid_stall[rrid]) {
         if (iopmp->stall_cntr != STALL_BUF_DEPTH){
             iopmp_trans_rsp->rrid_stalled = 1;
             iopmp->stall_cntr++;
@@ -113,13 +119,13 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
     // When no_w is set to 1, the IOPMP denies all write transactions regardless
     // of entry rule configurations, reporting them with error type
     // "not hit any rule" (0x05).
-    if (trans_req->perm == WRITE_ACCESS && iopmp->reg_file.hwcfg3.no_w) {
+    if (trans_perm == WRITE_ACCESS && iopmp->reg_file.hwcfg3.no_w) {
         iopmp->error_suppress = iopmp->reg_file.err_cfg.rs;
         error_type = NOT_HIT_ANY_RULE;
         goto stop_and_report_fault;
     }
 
-    if (trans_req->perm == INSTR_FETCH) {
+    if (trans_perm == INSTR_FETCH) {
         // When chk_x and no_x are set to 1, the IOPMP denies all instruction
         // fetch transactions regardless of entry rule configurations, reporting
         // them with error type "not hit any rule" (0x05).
@@ -128,17 +134,18 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
             error_type = NOT_HIT_ANY_RULE;
             goto stop_and_report_fault;
         }
+        // When chk_x = 0, The IOPMP doesn't perform instruction fetch
+        // permission checking. Instead, the IOPMP treats instruction fetch as
+        // read access.
+        if (!iopmp->reg_file.hwcfg2.chk_x) {
+            trans_perm = READ_ACCESS;
+        }
     }
 
     // Read SRCMD table based on `rrid`
     if (iopmp->reg_file.hwcfg3.srcmd_fmt == 0) {
-        #if (SRC_ENFORCEMENT_EN == 1)
-            srcmd_en  = iopmp->reg_file.srcmd_table[0].srcmd_en;
-            srcmd_enh = iopmp->reg_file.srcmd_table[0].srcmd_enh;
-        #else
-            srcmd_en  = iopmp->reg_file.srcmd_table[trans_req->rrid].srcmd_en;
-            srcmd_enh = iopmp->reg_file.srcmd_table[trans_req->rrid].srcmd_enh;
-        #endif
+        srcmd_en  = iopmp->reg_file.srcmd_table[rrid].srcmd_en;
+        srcmd_enh = iopmp->reg_file.srcmd_table[rrid].srcmd_enh;
     }
 
     int start_md_num;
@@ -149,10 +156,22 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
         start_md_num = 0;
         end_md_num   = iopmp->reg_file.hwcfg0.md_num;
     } else if (iopmp->reg_file.hwcfg3.srcmd_fmt == 1) {
-        start_md_num = SRC_ENFORCEMENT_EN ? 0 : trans_req->rrid;
-        end_md_num   = SRC_ENFORCEMENT_EN ? 1 : trans_req->rrid + 1;
+        start_md_num = rrid;
+        end_md_num   = rrid + 1;
     }
 
+    /* Prepare input of the rule analyzer which are fixed during entry checks */
+    iopmp_rule_analyzer_input_t rule_analyzer_i;
+    iopmp_rule_analyzer_output_t rule_analyzer_o;
+    rule_analyzer_i.rrid         = rrid;
+    rule_analyzer_i.trans_start  = trans_req->addr;
+    rule_analyzer_i.trans_end    = trans_req->addr +
+                          ((int)pow(2, trans_req->size) * (trans_req->length + 1));
+    rule_analyzer_i.perm         = trans_perm;
+    rule_analyzer_i.is_amo       = trans_req->is_amo;
+    rule_analyzer_o.match_status = ENTRY_NOTMATCH;
+    rule_analyzer_o.grant_perm   = false;
+
     // Traverse each MD entry and perform address/permission checks
     for (int cur_md = start_md_num; cur_md < end_md_num; ++cur_md) {
         if (iopmp->reg_file.hwcfg3.srcmd_fmt == 0) {
@@ -168,17 +187,25 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
         }
 
         for (int cur_entry = lwr_entry; cur_entry < upr_entry; cur_entry++) {
-            uint64_t prev_addr     = (cur_entry == 0) ? 0 : CONCAT32(iopmp->iopmp_entries.entry_table[cur_entry - 1].entry_addrh.addrh, iopmp->iopmp_entries.entry_table[cur_entry - 1].entry_addr.addr);
-            uint64_t curr_addr     = CONCAT32(iopmp->iopmp_entries.entry_table[cur_entry].entry_addrh.addrh, iopmp->iopmp_entries.entry_table[cur_entry].entry_addr.addr);
-            entry_cfg_t entry_cfg  = iopmp->iopmp_entries.entry_table[cur_entry].entry_cfg;
+            /* Assign necessary input information */
+            rule_analyzer_i.prev_iopmpaddr =
+                (cur_entry == 0) ? 0 : CONCAT32(iopmp->iopmp_entries.entry_table[cur_entry - 1].entry_addrh.addrh,
+                                            iopmp->iopmp_entries.entry_table[cur_entry - 1].entry_addr.addr);
+            rule_analyzer_i.iopmpaddr = CONCAT32(iopmp->iopmp_entries.entry_table[cur_entry].entry_addrh.addrh,
+                                            iopmp->iopmp_entries.entry_table[cur_entry].entry_addr.addr);
+            rule_analyzer_i.iopmpcfg = iopmp->iopmp_entries.entry_table[cur_entry].entry_cfg;
+            rule_analyzer_i.md       = cur_md;
+            /* Reset output information */
+            rule_analyzer_o.match_status = ENTRY_NOTMATCH;
+            rule_analyzer_o.grant_perm   = false;
 
-            // Analyze entry for match
-            iopmpMatchStatus = iopmpRuleAnalyzer(iopmp, *trans_req, prev_addr, curr_addr, entry_cfg, cur_md);
-            if (iopmpMatchStatus == ENTRY_MATCH_GRANT) {
+            // Analyze entry for matching and permission granting
+            iopmpRuleAnalyzer(iopmp, &rule_analyzer_i, &rule_analyzer_o);
+            if (rule_analyzer_o.match_status == ENTRY_MATCH && rule_analyzer_o.grant_perm) {
                 // If the entry matches all bytes of the transaction and grants
                 // transaction permission to operate, the transaction is legal.
                 goto pass_checks;
-            } else if (iopmpMatchStatus == ENTRY_PARTIAL_MATCH) {
+            } else if (rule_analyzer_o.match_status == ENTRY_PARTIAL_MATCH) {
                 // If the partial matching entry is non-priority entry, just
                 // keep checking next entry.
                 if (iopmp->reg_file.hwcfg2.non_prio_en && cur_entry >= iopmp->reg_file.hwcfg2.prio_entry)
@@ -192,7 +219,7 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
                 error_type = PARTIAL_HIT_ON_PRIORITY;
                 error_eid  = cur_entry;
                 goto stop_and_report_fault;
-            } else if (iopmpMatchStatus == ENTRY_MATCH_NOTGRANT) {
+            } else if (rule_analyzer_o.match_status == ENTRY_MATCH && !rule_analyzer_o.grant_perm) {
                 // If the matching entry is non-priority entry but doesn't grant
                 // transaction permission to operate, the model records this
                 // access as "first illegal access". This "first illegal access"
@@ -203,7 +230,7 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
                     nonPrioErrorSup |= iopmp->error_suppress;
                     nonPrioIntrSup  |= iopmp->intrpt_suppress;
                     if (firstIllegalAccess) {
-                        nonPrioRuleStatus = perm_to_etype(trans_req->perm);
+                        nonPrioRuleStatus = perm_to_etype(trans_perm);
                         nonPrioRuleNum    = cur_entry;
                         firstIllegalAccess = 0;
                     }
@@ -215,7 +242,7 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
                 // with error type = "illegal read access" (0x01) for read
                 // access transaction, "illegal write access/AMO" (0x02) for
                 // write access/atomic memory operation (AMO) transaction.
-                error_type = perm_to_etype(trans_req->perm);
+                error_type = perm_to_etype(trans_perm);
                 error_eid  = cur_entry;
                 goto stop_and_report_fault;
             }
@@ -258,7 +285,7 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
     // If IOPMP implements error capture feature, IOPMP triggers error capture
     // to log the error information into the registers.
     if (iopmp->imp_error_capture) {
-        errorCapture(iopmp, trans_req->perm, error_type, trans_req->rrid, error_eid, trans_req->addr, intrpt);
+        errorCapture(iopmp, trans_perm, error_type, rrid, error_eid, trans_req->addr, intrpt);
     }
     // Return response with default status if no match/error occurs
     // In case of error suppression, success response is returned, with user defined value on initiator port
diff --git a/iopmp_ref_model/verif/tests/fullmodel.c b/iopmp_ref_model/verif/tests/fullmodel.c
index b49588d..b9c3600 100644
--- a/iopmp_ref_model/verif/tests/fullmodel.c
+++ b/iopmp_ref_model/verif/tests/fullmodel.c
@@ -548,6 +548,24 @@ int main()
     write_register(&iopmp, ERR_INFO_OFFSET, 0, 4);
     END_TEST();)
 
+    START_TEST("Test NAPOT - 8 Byte Instruction access when chk_x=0");
+    cfg.chk_x = false;
+    reset_iopmp(&iopmp, &cfg);
+    configure_srcmd_n(&iopmp, SRCMD_EN, 32, 0x10, 4);
+    configure_srcmd_n(&iopmp, SRCMD_R, 32, 0x10, 4);
+    configure_mdcfg_n(&iopmp, 3, 2, 4);
+    configure_entry_n(&iopmp, ENTRY_ADDR, 1, 90, 4); // (364 >> 2) and keeping lsb 0
+    configure_entry_n(&iopmp, ENTRY_CFG, 1, (NAPOT), 4);
+    set_hwcfg0_enable(&iopmp);
+    receiver_port(32, 360, 0, 3, INSTR_FETCH, 0, &iopmp_trans_req);
+    // requestor Port Signals
+    iopmp_validate_access(&iopmp, &iopmp_trans_req, &iopmp_trans_rsp, &intrpt);
+    iopmp_trans_req.perm = READ_ACCESS; // Since chk_x=0, we expect ttype="Read access" in CHECK_IOPMP_TRANS()
+    CHECK_IOPMP_TRANS(&iopmp, IOPMP_ERROR, ILLEGAL_READ_ACCESS);
+    write_register(&iopmp, ERR_INFO_OFFSET, 0, 4);
+    cfg.chk_x = true;
+    END_TEST();    
+
     START_TEST_IF(iopmp.imp_mdlck, "Test MDLCK, updating locked srcmd_en field",
     reset_iopmp(&iopmp, &cfg);
     write_register(&iopmp, MDLCK_OFFSET, 0x10, 4);
@@ -1047,7 +1065,7 @@ int main()
     configure_srcmd_n(&iopmp, SRCMD_EN, 0, 0x02, 4);
     configure_srcmd_n(&iopmp, SRCMD_R, 0, 0x02, 4);
     configure_srcmd_n(&iopmp, SRCMD_W, 0, 0x02, 4);
-    configure_mdcfg_n(0, 2, 4);
+    configure_mdcfg_n(&iopmp, 0, 2, 4);
     configure_entry_n(&iopmp, ENTRY_ADDR, 0, 90, 4); // (364 >> 2) and keeping lsb 0
     configure_entry_n(&iopmp, ENTRY_CFG, 0, (NAPOT | W | R), 4);
     set_hwcfg0_enable(&iopmp);
