diff --git a/tools/ruby-gems/idlc/lib/idlc/ast.rb b/tools/ruby-gems/idlc/lib/idlc/ast.rb
index 76760111b5..d67cdd45e0 100644
--- a/tools/ruby-gems/idlc/lib/idlc/ast.rb
+++ b/tools/ruby-gems/idlc/lib/idlc/ast.rb
@@ -7149,12 +7149,6 @@ def to_s
     end
   end
 
-  # TODO: move this into a unit test
-  tmp = UnknownLiteral.new(5, 4)
-  raise tmp.to_s unless tmp.to_s == "3'bx01"
-  tmp = UnknownLiteral.new(0x7fff_ffff, 0b1000_0000_0000)
-  raise tmp.to_s unless tmp.to_s == "31'b1111111111111111111x11111111111"
-
   # represents an integer literal
   class IntLiteralAst < AstNode
     include Rvalue
diff --git a/tools/ruby-gems/idlc/test/run.rb b/tools/ruby-gems/idlc/test/run.rb
index 450e78f07a..ebec3eaca9 100644
--- a/tools/ruby-gems/idlc/test/run.rb
+++ b/tools/ruby-gems/idlc/test/run.rb
@@ -34,3 +34,4 @@
 require_relative "test_ast_type"
 require_relative "test_values"
 require_relative "test_std"
+require_relative "test_unknown_literal"
diff --git a/tools/ruby-gems/idlc/test/test_unknown_literal.rb b/tools/ruby-gems/idlc/test/test_unknown_literal.rb
new file mode 100644
index 0000000000..faffcd3594
--- /dev/null
+++ b/tools/ruby-gems/idlc/test/test_unknown_literal.rb
@@ -0,0 +1,21 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# typed: false
+# frozen_string_literal: true
+
+require "idlc"
+require "idlc/ast"
+require "minitest/autorun"
+
+module Idl
+  class TestUnknownLiteral < Minitest::Test
+    def test_to_s
+      tmp = UnknownLiteral.new(5, 4)
+      assert_equal "3'bx01", tmp.to_s
+
+      tmp = UnknownLiteral.new(0x7fff_ffff, 0b1000_0000_0000)
+      assert_equal "31'b1111111111111111111x11111111111", tmp.to_s
+    end
+  end
+end
