diff --git a/iommu_ref_model/libiommu/src/iommu_two_stage_trans.c b/iommu_ref_model/libiommu/src/iommu_two_stage_trans.c
index 60d2cbd0..5eabd003 100644
--- a/iommu_ref_model/libiommu/src/iommu_two_stage_trans.c
+++ b/iommu_ref_model/libiommu/src/iommu_two_stage_trans.c
@@ -21,7 +21,7 @@ two_stage_address_translation(
     pte_t amo_pte;
     gpte_t gpte;
     uint8_t NL_G = 0;
-    uint8_t is_implicit;
+    uint8_t is_implicit, is_implicit_write = 0;
     uint8_t PTESIZE, LEVELS, status, pte_changed, gst_fault;
     int8_t i;
     uint64_t a, a_gpa, masked_upper_bits, mask;
@@ -317,6 +317,7 @@ two_stage_address_translation(
     if ( SADE == 0 ) goto page_fault;
 
     // If G-stage does not provide write permission then cause guest page fault
+    is_implicit_write = 1;
     if ( gpte.W == 0 ) goto guest_page_fault;
 
     // Count S/VS stage page walks
@@ -401,7 +402,7 @@ two_stage_address_translation(
     // VS-stage address translation. If bit 0 of iotval2 is 1, and the implicit
     // access was a write then bit 1 of iotval2 is set to 1 else it is set to 0.
     *iotval2 = (a_gpa & ~0x3);
-    *iotval2 |= 1;
-    *iotval2 |= (SADE << 1);
+    *iotval2 |= is_implicit;
+    *iotval2 |= (is_implicit_write << 1);
     return 1;
 }
diff --git a/iommu_ref_model/test/test_app.c b/iommu_ref_model/test/test_app.c
index 432cdaaf..db12d91c 100644
--- a/iommu_ref_model/test/test_app.c
+++ b/iommu_ref_model/test/test_app.c
@@ -2497,7 +2497,7 @@ main(void) {
     send_translation_request(&iommu, 0x112233, 1, 0xBABEC, 0,
              0, 1, 0, ADDR_TYPE_UNTRANSLATED, 0xdeadbeef,
              1, WRITE, &req, &rsp);
-    fail_if( ( check_rsp_and_faults(&iommu, &req, &rsp, UNSUPPORTED_REQUEST, 274, 0) < 0 ) );
+    fail_if( ( check_rsp_and_faults(&iommu, &req, &rsp, UNSUPPORTED_REQUEST, 269, 0) < 0 ) );
     data_corruption_addr = -1;
 
     // Two stage translation
@@ -2759,7 +2759,7 @@ main(void) {
              0, 0, 0, ADDR_TYPE_UNTRANSLATED, gva,
              1, WRITE, &req, &rsp);
     fail_if( ( check_rsp_and_faults(&iommu, &req, &rsp, UNSUPPORTED_REQUEST, 23,
-               ((PC.fsc.iosatp.PPN * PAGESIZE) | 3)) < 0 ) );
+               ((PC.fsc.iosatp.PPN * PAGESIZE) | 1)) < 0 ) );
 
     gpte.V = 1;
     gpte.N = 1;
@@ -2768,7 +2768,7 @@ main(void) {
              0, 0, 0, ADDR_TYPE_UNTRANSLATED, gva,
              1, WRITE, &req, &rsp);
     fail_if( ( check_rsp_and_faults(&iommu, &req, &rsp, UNSUPPORTED_REQUEST, 23,
-               ((PC.fsc.iosatp.PPN * PAGESIZE) | 3)) < 0 ) );
+               ((PC.fsc.iosatp.PPN * PAGESIZE) | 1)) < 0 ) );
     gpte.N = 0;
     gpte.PBMT = 1;
     write_memory_test((char *)&gpte.raw, gpte_addr, 8);
@@ -2776,7 +2776,7 @@ main(void) {
              0, 0, 0, ADDR_TYPE_UNTRANSLATED, gva,
              1, WRITE, &req, &rsp);
     fail_if( ( check_rsp_and_faults(&iommu, &req, &rsp, UNSUPPORTED_REQUEST, 23,
-               ((PC.fsc.iosatp.PPN * PAGESIZE) | 3)) < 0 ) );
+               ((PC.fsc.iosatp.PPN * PAGESIZE) | 1)) < 0 ) );
     gpte.N = 0;
     gpte.PBMT = 0;
     write_memory_test((char *)&gpte.raw, gpte_addr, 8);
