diff --git a/tools/eclipse/udb-ls/build.gradle b/tools/eclipse/udb-ls/build.gradle
index 97de83f983..5ff51cade5 100644
--- a/tools/eclipse/udb-ls/build.gradle
+++ b/tools/eclipse/udb-ls/build.gradle
@@ -21,7 +21,7 @@ dependencies {
   implementation "org.eclipse.xtext:org.eclipse.xtext.xbase.ide:${xtextVersion}"
 
 
-  implementation "io.github.classgraph:classgraph:4.8.180"
+  implementation "io.github.classgraph:classgraph:4.8.184"
   implementation "aopalliance:aopalliance:1.0"
 }
 
