diff --git a/bin/clobber b/bin/clobber
index 3a9cc4f53d..f9a32d5cd0 100755
--- a/bin/clobber
+++ b/bin/clobber
@@ -5,6 +5,7 @@ ROOT=$(dirname "$(dirname "$(realpath "${BASH_SOURCE[0]}")")")
 rm -rf                   \
   "${ROOT}"/.stamps      \
   "${ROOT}"/.home        \
+  "${ROOT}"/.docker-home \
   "${ROOT}"/.bundle      \
   "${ROOT}"/.singularity \
   "${ROOT}"/.cache       \
diff --git a/bin/setup b/bin/setup
index 82f6ca9f7b..d179d22479 100755
--- a/bin/setup
+++ b/bin/setup
@@ -113,8 +113,11 @@ elif [ "${CONTAINER_TYPE}" == "docker" ] || [ "${CONTAINER_TYPE}" == "podman" ];
     fi
     if ! [ -v GITHUB_ACTIONS ]; then
       # create empty home directory (used as location to cache files by several tools)
+      if ! [ -d "${ROOT}"/.docker-home ]; then
+        mkdir "${ROOT}"/.docker-home
+      fi
       CONTAINER_RUN_FLAGS="${CONTAINER_RUN_FLAGS} -e HOME"
-      CONTAINER_RUN_FLAGS="${CONTAINER_RUN_FLAGS} --tmpfs ${HOME}"
+      CONTAINER_RUN_FLAGS="${CONTAINER_RUN_FLAGS} -v ${ROOT}/.docker-home:${HOME}"
     fi
 
     mkdir -p "${ROOT}/.cache"
