diff --git a/iopmp_ref_model/src/iopmp_error_capture.c b/iopmp_ref_model/src/iopmp_error_capture.c
index 869dcc1..a311c2e 100644
--- a/iopmp_ref_model/src/iopmp_error_capture.c
+++ b/iopmp_ref_model/src/iopmp_error_capture.c
@@ -25,17 +25,6 @@ static void setRridSv(iopmp_dev_t *iopmp, uint16_t rrid) {
     iopmp->err_svs.sv[rrid/16].svw |= (1 << (rrid % 16));
 }
 
-/**
-  * @brief Checks if the corresponding bit in the error subsequent violations (SV) structure is set for a given RRID.
-  *
-  * @param iopmp The IOPMP instance.
-  * @param rrid Resource Record ID (RRID) to check in the SV structure.
-  * @return int Returns 1 if the bit corresponding to the RRID is set, otherwise returns 0.
-  */
-static int checkRridSv(iopmp_dev_t *iopmp, uint16_t rrid) {
-    return (iopmp->err_svs.sv[rrid/16].svw >> (rrid % 16)) & 0x1;
-}
-
 /**
   * @brief Captures and logs error information for a transaction request.
   *
@@ -45,12 +34,29 @@ static int checkRridSv(iopmp_dev_t *iopmp, uint16_t rrid) {
   * @param rrid Requester ID associated with the transaction.
   * @param entry_id IOPMP entry ID where the error was encountered.
   * @param err_addr Address at which the error occurred.
-  * @param intrpt Pointer to an interrupt flag, which is set if an error is captured.
+  * @param intrpt Pointer to the variable to store wired interrupt flag.
+  *               This flag is set to 1 if the following conditions are true:
+  *                 - the transaction fails
+  *                 - a primary error capture occurs
+  *                 - the interrupts are not suppressed
+  *                 - IOPMP doesn't implement MSI extension, or MSI is not enabled
+  *               This flag is set to 0 if the following conditions are true:
+  *                 - this transaction fails
+  *                 - a primary error capture occurs
+  *                 - the interrupts are suppressed, or IOPMP implements MSI extension
+  *                   and triggers MSI instead of wired interrupt
  **/
 void errorCapture(iopmp_dev_t *iopmp, perm_type_e trans_type, uint8_t error_type, uint16_t rrid, uint16_t entry_id, uint64_t err_addr, uint8_t *intrpt) {
+
+    // Current value of ERR_INFO.v
     int err_reqinfo_v = iopmp->reg_file.err_info.v;
-    // If no error has been logged and interrupt and error both are not suppressed, capture error details
-    if (!iopmp->reg_file.err_info.v && (!iopmp->error_suppress | !iopmp->intrpt_suppress)) {
+
+    // First error capture occurs when the following conditions are true:
+    //   - An interrupt is triggered, or a bus error is returned
+    //   - There is no pending error capture, that is, current ERR_INFO.v = '0'
+    bool first_record = (!iopmp->error_suppress || !iopmp->intrpt_suppress) && !err_reqinfo_v;
+
+    if (first_record) {
         iopmp->reg_file.err_info.v = 1;               // Mark error as captured
         // Set error status and transaction details
         iopmp->reg_file.err_info.ttype = trans_type;  // Transaction type (read/write)
@@ -68,25 +74,23 @@ void errorCapture(iopmp_dev_t *iopmp, perm_type_e trans_type, uint8_t error_type
             // If IOPMP doesn't implement ERR_REQID.eid it won't be updated.
             iopmp->reg_file.err_reqid.eid = entry_id;
         }
-    // If an error was previously logged, handle a subsequent violation
-    } else if (iopmp->reg_file.hwcfg2.mfr_en) {
-        if (!checkRridSv(iopmp, rrid) && (!iopmp->error_suppress | !iopmp->intrpt_suppress)) {
-            // Update violation window
-            setRridSv(iopmp, rrid);
-        }
+
+        generate_interrupt(iopmp, intrpt);
     }
 
+    // If IOPMP implements Multi-Faults Record extension, IOPMP has ability to
+    // record subsequent violations when the primary error capture registers
+    // recorded last violation and software have not invalidated them yet.
     if (iopmp->reg_file.hwcfg2.mfr_en) {
-        // Check for any subsequent violation and set err_info.svc
-        for (int i = 0; i < NUM_SVW; i++) {
-            if (iopmp->err_svs.sv[i].svw) {
-                iopmp->reg_file.err_info.svc = 1;
-                break;
-            }
+        // Subsequent error capture occurs when the following conditions are true:
+        //   - An interrupt is triggered or a bus error is returned
+        //   - There is a pending error capture, that is, current ERR_INFO.v = '1'
+        bool subsq_record = (!iopmp->error_suppress || !iopmp->intrpt_suppress) && err_reqinfo_v;
+
+        if (subsq_record) {
+            // Update violation window
+            setRridSv(iopmp, rrid);
+            iopmp->reg_file.err_info.svc = 1;
         }
     }
-
-    // Generate Interrupt
-    if (!err_reqinfo_v)
-        generate_interrupt(iopmp, intrpt);
 }
diff --git a/iopmp_ref_model/src/iopmp_interrupt.c b/iopmp_ref_model/src/iopmp_interrupt.c
index 8973b9d..4977f0f 100644
--- a/iopmp_ref_model/src/iopmp_interrupt.c
+++ b/iopmp_ref_model/src/iopmp_interrupt.c
@@ -19,38 +19,60 @@
  * and data, writes them to memory, and handles any potential bus errors during the write operation.
  *
  * @param iopmp The IOPMP instance.
- * @param intrpt Pointer to a variable to store the interrupt flag.
+ * @param intrpt Pointer to the variable to store wired interrupt flag.
+ *               This flag is set to 1 if the following conditions are true:
+ *                 - the interrupts are not suppressed
+ *                 - IOPMP doesn't implement MSI extension, or MSI is not enabled
+ *               This flag is set to 0 if the following conditions are true:
+ *                 - the interrupts are suppressed, or IOPMP implements MSI extension
+ *                   and triggers MSI instead of wired interrupt
  */
 void generate_interrupt(iopmp_dev_t *iopmp, uint8_t *intrpt) {
-    // Extract configuration values for clarity
-    const uint8_t msi_enabled = iopmp->imp_msi && iopmp->reg_file.err_cfg.msi_en;
-    const uint8_t interrupt_enabled = iopmp->reg_file.err_cfg.ie;
-    *intrpt = 0;
-    // Check if interrupts are not enabled
-    if (interrupt_enabled) {
-        // Check if interrupts are not suppressed
-        if (!iopmp->intrpt_suppress) {
-            *intrpt = !msi_enabled;
-            if (iopmp->imp_msi) {
-                if (msi_enabled && !iopmp->reg_file.err_info.msi_werr) {
-                    // Construct MSI address and data for enabled MSI.
-                    // {MSI_ADDRH[64:34], MSI_ADDR[33:2], 2'b00}
-                    uint64_t msi_addr;
-                    if (iopmp->reg_file.hwcfg0.addrh_en) {
-                        msi_addr = CONCAT32(iopmp->reg_file.err_msiaddrh.raw, iopmp->reg_file.err_msiaddr.raw);
-                    } else {
-                        msi_addr = CONCAT32(0, iopmp->reg_file.err_msiaddr.raw << 2);
-                    }
-                    uint64_t msi_data = iopmp->reg_file.err_cfg.msidata;
-                    // Write MSI data to memory
-                    uint8_t status = write_memory(&msi_data, msi_addr, MSI_DATA_BYTE);
 
-                    // Handle bus errors during MSI write
-                    if (status == BUS_ERROR) {
-                        iopmp->reg_file.err_info.msi_werr = 1;
-                    }
-                }
-            }
+    if (!iopmp->imp_msi) {
+        // If IOPMP doesn't implement Message-Signaled Interrupts (MSI)
+        // extension, IOPMP only triggers wired interrupts.
+        // IOPMP triggers wired interrupt if the following conditions are true:
+        //   - IOPMP interrupts are enabled
+        //   - The interrupts are not suppressed
+        *intrpt = iopmp->reg_file.err_cfg.ie && !iopmp->intrpt_suppress;
+        return;
+    }
+
+    // IOPMP implements Message-Signaled Interrupts (MSI) extension.
+    // IOPMP triggers wired interrupt if the following conditions are true:
+    //   - IOPMP interrupts are enabled
+    //   - The interrupts are not suppressed
+    //   - IOPMP doesn't enable MSI
+    *intrpt = iopmp->reg_file.err_cfg.ie && !iopmp->intrpt_suppress &&
+              !iopmp->reg_file.err_cfg.msi_en;
+
+    // IOPMP implements Message-Signaled Interrupts (MSI) extension.
+    // IOPMP triggers MSI if the following conditions are true:
+    //   - IOPMP interrupts are enabled
+    //   - The interrupts are not suppressed
+    //   - MSI is enabled
+    //   - (implementation-specific) There are no pending MSI write error
+    bool msi = iopmp->reg_file.err_cfg.ie && !iopmp->intrpt_suppress &&
+               iopmp->reg_file.err_cfg.msi_en &&
+               !iopmp->reg_file.err_info.msi_werr;
+    if (msi) {
+        // Construct MSI address and data for enabled MSI.
+        uint64_t msi_addr;
+        if (iopmp->reg_file.hwcfg0.addrh_en) {
+            // {MSI_ADDRH[63:32], MSI_ADDR[31:0]}
+            msi_addr = CONCAT32(iopmp->reg_file.err_msiaddrh.raw, iopmp->reg_file.err_msiaddr.raw);
+        } else {
+            // {MSI_ADDR[33:2], 2'b00}
+            msi_addr = CONCAT32(0, iopmp->reg_file.err_msiaddr.raw << 2);
+        }
+        uint64_t msi_data = iopmp->reg_file.err_cfg.msidata;
+        // Write MSI data to memory
+        uint8_t status = write_memory(&msi_data, msi_addr, MSI_DATA_BYTE);
+
+        // Handle bus errors during MSI write
+        if (status == BUS_ERROR) {
+            iopmp->reg_file.err_info.msi_werr = 1;
         }
     }
 }
diff --git a/iopmp_ref_model/src/iopmp_validate.c b/iopmp_ref_model/src/iopmp_validate.c
index 0f782e2..65ea4b4 100644
--- a/iopmp_ref_model/src/iopmp_validate.c
+++ b/iopmp_ref_model/src/iopmp_validate.c
@@ -16,7 +16,17 @@
   *
   * @param iopmp The IOPMP instance.
   * @param trans_req The transaction request with required address, permissions, etc.
-  * @param intrpt Pointer to the interrupt flag.
+  * @param intrpt Pointer to the variable to store wired interrupt flag.
+  *               This flag is set to 1 if the following conditions are true:
+  *                 - the transaction fails
+  *                 - a primary error capture occurs
+  *                 - the interrupts are not suppressed
+  *                 - IOPMP doesn't implement MSI extension, or MSI is not enabled
+  *               This flag is set to 0 if the following conditions are true:
+  *                 - this transaction fails
+  *                 - a primary error capture occurs
+  *                 - the interrupts are suppressed, or IOPMP implements MSI extension
+  *                   and triggers MSI instead of wired interrupt
   * @return iopmp_trans_rsp_t Response structure with transaction status.
  **/
 void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iopmp_trans_rsp_t* iopmp_trans_rsp, uint8_t *intrpt) {
