diff --git a/tools/ruby-gems/idlc/lib/idlc/ast.rb b/tools/ruby-gems/idlc/lib/idlc/ast.rb
index 76760111b5..d67cdd45e0 100644
--- a/tools/ruby-gems/idlc/lib/idlc/ast.rb
+++ b/tools/ruby-gems/idlc/lib/idlc/ast.rb
@@ -7149,12 +7149,6 @@ def to_s
     end
   end
 
-  # TODO: move this into a unit test
-  tmp = UnknownLiteral.new(5, 4)
-  raise tmp.to_s unless tmp.to_s == "3'bx01"
-  tmp = UnknownLiteral.new(0x7fff_ffff, 0b1000_0000_0000)
-  raise tmp.to_s unless tmp.to_s == "31'b1111111111111111111x11111111111"
-
   # represents an integer literal
   class IntLiteralAst < AstNode
     include Rvalue
diff --git a/tools/ruby-gems/idlc/test/test_values.rb b/tools/ruby-gems/idlc/test/test_values.rb
index 8deb615422..c0ab36b3fc 100644
--- a/tools/ruby-gems/idlc/test/test_values.rb
+++ b/tools/ruby-gems/idlc/test/test_values.rb
@@ -57,3 +57,14 @@ def fields = [
     assert_equal 0, ast.value(symtab)
   end
 end
+
+# test UnknownLiteral
+class TestUnknownLiteral < Minitest::Test
+  def test_to_s
+    tmp = Idl::UnknownLiteral.new(5, 4)
+    assert_equal "3'bx01", tmp.to_s
+
+    tmp = Idl::UnknownLiteral.new(0x7fff_ffff, 0b1000_0000_0000)
+    assert_equal "31'b1111111111111111111x11111111111", tmp.to_s
+  end
+end
