diff --git a/spec/std/isa/inst/F/fdiv.s.yaml b/spec/std/isa/inst/F/fdiv.s.yaml
index bdab30c217..aedc732602 100644
--- a/spec/std/isa/inst/F/fdiv.s.yaml
+++ b/spec/std/isa/inst/F/fdiv.s.yaml
@@ -32,7 +32,10 @@ access:
   vu: always
 data_independent_timing: true
 operation(): |
-
+  check_f_ok($encoding);
+  RoundingMode mode = rm_to_mode(rm, $encoding);
+  f[fd] = f32_div(f[fs1], f[fs2], mode);
+  mark_f_state_dirty();
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
 # SPDX-License-Identifier: BSD-2-Clause
diff --git a/spec/std/isa/inst/F/fmadd.s.yaml b/spec/std/isa/inst/F/fmadd.s.yaml
index e846055995..2e7b8e32e7 100644
--- a/spec/std/isa/inst/F/fmadd.s.yaml
+++ b/spec/std/isa/inst/F/fmadd.s.yaml
@@ -33,6 +33,11 @@ access:
   vu: always
 data_independent_timing: true
 operation(): |
+  check_f_ok($encoding);
+  RoundingMode mode = rm_to_mode(rm, $encoding);
+  F32MulAddOp op = F32MulAddOp::Softfloat_mulAdd_addC;
+  f[fd] = f32_muladd(f[fs1], f[fs2], f[fs3], op, mode);
+  mark_f_state_dirty();
 
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
diff --git a/spec/std/isa/inst/F/fmsub.s.yaml b/spec/std/isa/inst/F/fmsub.s.yaml
index 1eb72b3433..45193772b2 100644
--- a/spec/std/isa/inst/F/fmsub.s.yaml
+++ b/spec/std/isa/inst/F/fmsub.s.yaml
@@ -33,6 +33,11 @@ access:
   vu: always
 data_independent_timing: true
 operation(): |
+  check_f_ok($encoding);
+  RoundingMode mode = rm_to_mode(rm, $encoding);
+  F32MulAddOp op = F32MulAddOp::Softfloat_mulAdd_subC;
+  f[fd] = f32_muladd(f[fs1], f[fs2], f[fs3], op, mode);
+  mark_f_state_dirty();
 
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
diff --git a/spec/std/isa/inst/F/fmul.s.yaml b/spec/std/isa/inst/F/fmul.s.yaml
index 4360f1f953..55f50b2c27 100644
--- a/spec/std/isa/inst/F/fmul.s.yaml
+++ b/spec/std/isa/inst/F/fmul.s.yaml
@@ -32,6 +32,10 @@ access:
   vu: always
 data_independent_timing: true
 operation(): |
+  check_f_ok($encoding);
+  RoundingMode mode = rm_to_mode(rm, $encoding);
+  f[fd] = f32_mul(f[fs1], f[fs2], mode);
+  mark_f_state_dirty();
 
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
diff --git a/spec/std/isa/inst/F/fnmadd.s.yaml b/spec/std/isa/inst/F/fnmadd.s.yaml
index b31ac89133..5c5e6994ca 100644
--- a/spec/std/isa/inst/F/fnmadd.s.yaml
+++ b/spec/std/isa/inst/F/fnmadd.s.yaml
@@ -34,6 +34,11 @@ access:
   vu: always
 data_independent_timing: true
 operation(): |
+  check_f_ok($encoding);
+  RoundingMode mode = rm_to_mode(rm, $encoding);
+  F32MulAddOp op = F32MulAddOp::Softfloat_mulAdd_subC;
+  f[fd] = f32_muladd(-f[fs1], f[fs2], -f[fs3], op, mode);
+  mark_f_state_dirty();
 
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
diff --git a/spec/std/isa/inst/F/fnmsub.s.yaml b/spec/std/isa/inst/F/fnmsub.s.yaml
index ae5e25a2a8..383c51e921 100644
--- a/spec/std/isa/inst/F/fnmsub.s.yaml
+++ b/spec/std/isa/inst/F/fnmsub.s.yaml
@@ -34,6 +34,11 @@ access:
   vu: always
 data_independent_timing: true
 operation(): |
+  check_f_ok($encoding);
+  RoundingMode mode = rm_to_mode(rm, $encoding);
+  F32MulAddOp op = F32MulAddOp::Softfloat_mulAdd_addC;
+  f[fd] = f32_muladd(-f[fs1], f[fs2], f[fs3], op, mode);
+  mark_f_state_dirty();
 
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
diff --git a/spec/std/isa/isa/fp.idl b/spec/std/isa/isa/fp.idl
index 1dcfe70b94..1397388d64 100644
--- a/spec/std/isa/isa/fp.idl
+++ b/spec/std/isa/isa/fp.idl
@@ -49,6 +49,12 @@ enum FpFlag {
   NV 0b10000 # Invalid Operation
 }
 
+enum F32MulAddOp {
+  Softfloat_mulAdd_addC    # will get value 0
+  Softfloat_mulAdd_subC    # will get value 1
+  Softfloat_mulAdd_subProd # will get value 2
+}
+
 function set_fp_flag {
   arguments
     FpFlag flag
@@ -313,6 +319,7 @@ function softfloat_shiftRightJam32 {
   }
 }
 
+
 function softfloat_shiftRightJam64 {
   returns Bits<64>
   arguments
@@ -329,7 +336,11 @@ function softfloat_shiftRightJam64 {
      is zero or nonzero.
   }
   body {
-    return (dist < 63) ? a>>dist | (((a<<(-dist & 63)) != 0) ? 1 : 0) : ((a != 0) ? 1 : 0);
+    # negate dist using 2's complement to avoid a truncation warning from the IDL compiler
+    # that would happen if you use -dist
+    # (because '-dist' is a signed type, which, when masked, loses information)
+    Bits<32> neg_dist = ~dist + 1;
+    return (dist < 63) ? a>>dist | (((a<<(neg_dist & 63)) != 0) ? 1 : 0) : ((a != 0) ? 1 : 0);
   }
 }
 
@@ -475,6 +486,19 @@ function softfloat_normSubnormalF16Sig {
   }
 }
 
+function softfloat_normSubnormalF32Sig {
+  returns Bits<8>, Bits<23>
+  arguments
+    Bits<32> sp_value
+  description {
+    normalize subnormal single-precision value
+  }
+  body {
+    Bits<8> shift_dist = count_leading_zeros<32>(sp_value) - 8;
+    return 1 - shift_dist, sp_value << shift_dist;
+  }
+}
+
 function softfloat_roundPackToF32 {
   returns Bits<32>     # single precision value
   arguments
@@ -759,7 +783,7 @@ function softfloat_addMagsF32 {
 
       U32 sigZ = 0x20000000 + sigA + sigB;
       if ( sigZ < 0x40000000 ) {
-            expZ = expZ - 1;
+            expZ = expZ `- 1'b1;
             sigZ =  sigZ << 1;
         }
     }
@@ -1236,3 +1260,368 @@ function round_f32_to_integral {
     return i32_to_f32_no_flag(intermediate, mode);
   }
 }
+
+function f32_mul {
+  returns U32
+  arguments
+    U32 a,
+    U32 b,
+    RoundingMode mode
+  description {
+    Returns product of 2 floating point numbers
+  }
+  body {
+    Bits<1> signA = signF32UI(a);
+    Bits<8> expA  = expF32UI(a);
+    Bits<23> sigA = fracF32UI(a);
+    Bits<1> signB = signF32UI(b);
+    Bits<8> expB  = expF32UI(b);
+    Bits<23> sigB = fracF32UI(b);
+
+    # declare a variable to store significand of product
+    U32 sigZ;
+
+    # declare a variable to store sign of product
+    Bits<1> signZ;
+
+    # declare a variable to store the exponent part of product
+    Bits<8> expZ;
+
+    # calculate sign of product
+    signZ = signA ^ signB;
+
+    U32 magBits;
+
+    if (expA == 0xFF) {
+      if ((sigA != 0) ||  ((expB == 0xFF) && (sigB != 0))) {
+        return softfloat_propagateNaNF32UI(a, b);
+        magBits = expB | sigB;
+        if (magBits == 0) {
+          set_fp_flag(FpFlag::NV);
+          return UI32_NAN;
+        } else {
+          return packToF32UI(signZ, 0xFF, 0);
+        }
+      }
+    }
+
+    if (expB == 0xFF) {
+      if (sigB != 0) {
+        return softfloat_propagateNaNF32UI(a, b);
+      }
+      magBits = expA | sigA;
+      if (magBits == 0) {
+          set_fp_flag(FpFlag::NV);
+          return UI32_NAN;
+      } else {
+          return packToF32UI(signZ, 0xFF, 0);
+      }
+    }
+
+    if (expA == 0) {
+      if (sigA == 0) {
+        return packToF32UI(signZ, 0, 0);
+      }
+      (expA, sigA) = softfloat_normSubnormalF32Sig(sigA);
+    }
+
+    if (expB == 0) {
+      if (sigB == 0) {
+        return packToF32UI(signZ, 0, 0);
+      }
+      (expB, sigB) = softfloat_normSubnormalF32Sig(sigB);
+    }
+
+    expZ = expA + expB - 0x7F;
+    sigA = (sigA | 0x00800000)<<7;
+    sigB = (sigB | 0x00800000)<<8;
+    sigZ = softfloat_shiftRightJam64(sigA `* sigB, 32);
+    if (sigZ < 0x40000000) {
+      expZ = expZ `- 1'b1;
+      sigZ = sigZ << 1;
+    }
+    return softfloat_roundPackToF32(signZ, expZ, sigZ, mode);
+  }
+}
+
+function f32_div {
+  returns U32
+  arguments
+    U32 a,
+    U32 b,
+    RoundingMode mode
+  description {
+    Returns quotient of 2 floating point numbers
+  }
+  body {
+    Bits<1> signA = signF32UI(a);
+    Bits<8> expA  = expF32UI(a);
+    Bits<23> sigA = fracF32UI(a);
+    Bits<1> signB = signF32UI(b);
+    Bits<8> expB  = expF32UI(b);
+    Bits<23> sigB = fracF32UI(b);
+    Bits<1> signZ = signA ^ signB;
+    Bits<8> expZ;
+    Bits<23> sigZ;
+
+    U64 sig64A;
+
+    if (expA == 0xFF) {
+      if (sigA != 0) {
+        return softfloat_propagateNaNF32UI(a, b);
+      }
+      if ((expB == 0xFF) && (sigB != 0)) {
+        return softfloat_propagateNaNF32UI(a, b);
+      }
+      if ((expB == 0xFF) && (sigB == 0)) {
+        set_fp_flag(FpFlag::NV);
+        return SP_CANONICAL_NAN;
+      }
+      return packToF32UI(signZ, 0xFF, 0);
+    }
+
+    if (expB == 0xFF) {
+      if (sigB != 0) {
+        return softfloat_propagateNaNF32UI(a, b);
+      }
+      return packToF32UI(signZ, 0, 0);
+    }
+
+    if (expB == 0) {
+      if (sigB == 0) {
+        if ((expA == 0) && (sigA == 0)) {
+          set_fp_flag(FpFlag::NV);
+          return SP_CANONICAL_NAN;
+        }
+        set_fp_flag(FpFlag::OF);
+        set_fp_flag(FpFlag::NX);
+        return packToF32UI(signZ, 0xFF, 0);
+      }
+      (expB, sigB) = softfloat_normSubnormalF32Sig(sigB);
+    }
+    if (expA == 0) {
+      if (sigA == 0) {
+        return packToF32UI(signZ, 0, 0);
+      }
+      (expA, sigA) = softfloat_normSubnormalF32Sig(sigA);
+    }
+
+    expZ = expA - expB + 0x7E;
+    sigA = (sigA | 0x800000);
+    sigB = (sigB | 0x800000);
+
+    if (sigA < sigB) {
+      expZ = expZ - 1;
+      sig64A = sigA `<< 31;
+    } else {
+      sig64A = sigA `<< 30;
+    }
+
+    sigZ = sig64A / sigB;
+    # Check if the lower 6 bits are zero, meaning the division was exact.
+    # This is used to determine if rounding/jamming is needed.
+    if ((sigZ & 0x3F) != 0) {
+      sigZ = sigZ | (((sigB `* sigZ) != sig64A) ? 0 : 1);
+    }
+
+    return softfloat_roundPackToF32(signZ, expZ, sigZ, mode);
+  }
+}
+
+function propagateNaN_ABC {
+  returns U32
+  arguments
+    U32 a,
+    U32 b,
+    U32 c
+  description {
+    Applies softfloat_propagateNaNF32UI to 3 numbers
+  }
+  body {
+    U32 z = softfloat_propagateNaNF32UI(a, b);
+    return softfloat_propagateNaNF32UI(z, c);
+  }
+}
+
+function f32_muladd {
+  returns U32
+  arguments
+    U32 a,
+    U32 b,
+    U32 c,
+    F32MulAddOp op,
+    RoundingMode mode
+  description {
+    Sum of the product of `a` and `b` with `c`. Sign of `c` depends on `op`
+  }
+  body {
+    Bits<1> signA = signF32UI(a);
+    Bits<8> expA  = expF32UI(a);
+    Bits<23> sigA = fracF32UI(a);
+    Bits<1> signB = signF32UI(b);
+    Bits<8> expB  = expF32UI(b);
+    Bits<23> sigB = fracF32UI(b);
+    Bits<1> intermediate_signC = (op == F32MulAddOp::Softfloat_mulAdd_subC) ? 1 : 0;
+    Bits<1> signC = signF32UI(c) ^ intermediate_signC;
+    Bits<8> expC  = expF32UI(c);
+    Bits<23> sigC = fracF32UI(c);
+    Bits<1> intermediate_signProd = (op == F32MulAddOp::Softfloat_mulAdd_subProd) ? 1 : 0;
+    Bits<1> signProd = (signA ^ signB) ^ intermediate_signProd;
+    Bits<8> expProd;
+    U64 sigProd;
+    Bits<1> signZ;
+    Bits<8> expZ;
+    Bits<23> sigZ;
+
+    Bits<32> z;
+    U64 sig64Z, sig64C;
+
+    Bits<32> expDiff;
+    U32 magBits;
+    Bits<8> shiftDist;
+
+    if (expA == 0xFF) {
+      if ((sigA != 0) || ((expB == 0xFF) && (sigB != 0))) {
+        propagateNaN_ABC(a,b,c);
+      }
+      magBits = expB | sigB;
+      if (magBits != 0) {
+        z = packToF32UI(signProd, 0xFF, 0);
+        if (expC != 0xFF) {
+          return z;
+        }
+        if (sigC != 0) {
+          softfloat_propagateNaNF32UI(z, c);
+        }
+        if (signProd == signC) {
+          return z;
+        }
+      }
+      set_fp_flag(FpFlag::NV);
+      z = UI32_NAN;
+      softfloat_propagateNaNF32UI(z, c);
+      return z;
+    }
+
+    if (expB == 0xFF) {
+      if (sigB != 0) {
+        propagateNaN_ABC(a,b,c);
+      }
+      magBits = expA | sigA;
+      if (magBits != 0) {
+        z = packToF32UI(signProd, 0xFF, 0);
+        if (expC != 0xFF) {
+          return z;
+        }
+        if (sigC != 0) {
+          softfloat_propagateNaNF32UI(z, c);
+        }
+        if (signProd == signC) {
+          return z;
+        }
+      }
+      set_fp_flag(FpFlag::NV);
+      z = UI32_NAN;
+      softfloat_propagateNaNF32UI(z, c);
+      return z;
+    }
+
+    if (expC == 0xFF) {
+      if (sigC != 0) {
+        z = 0;
+        softfloat_propagateNaNF32UI(z, c);
+      }
+      z = c;
+      return z;
+    }
+
+    if (expA == 0) {
+      if (sigA == 0) {
+        z = c;
+        if (((expC | sigC) == 0) && (signProd != signC)) {
+          z = packToF32UI((mode == RoundingMode::RDN) ? 1 : 0, 0, 0);
+        }
+        return z;
+      }
+      (expA, sigA) = softfloat_normSubnormalF32Sig(sigA);
+    }
+
+    if (expB == 0) {
+      if (sigB == 0) {
+        z = c;
+        if (((expC | sigC) == 0) && (signProd != signC)) {
+          z = packToF32UI((mode == RoundingMode::RDN) ? 1 : 0, 0, 0);
+        }
+        return z;
+      }
+      (expB, sigB) = softfloat_normSubnormalF32Sig(sigB);
+    }
+
+    expProd = expA + expB - 0x7E;
+    sigA = (sigA | 0x00800000)<<7;
+    sigB = (sigB | 0x00800000)<<7;
+    sigProd = sigA `* sigB;
+    if ( sigProd < 0x2000000000000000 ) {
+        expProd = expProd `- 1;
+        sigProd = sigProd << 1;
+    }
+    signZ = signProd;
+    if (expC == 0) {
+      if (sigC == 0) {
+        expZ = expProd `- 1;
+        sigZ = softfloat_shiftRightJam64(sigProd, 31);
+        return softfloat_roundPackToF32(signZ, expZ, sigZ, mode);
+      }
+      (expC, sigC) = softfloat_normSubnormalF32Sig( sigC );
+    }
+    sigC = (sigC | 0x00800000)<<6;
+    expDiff = expProd - expC;
+    if ( signProd == signC ) {
+        if ( expDiff <= 0 ) {
+            expZ = expC;
+            sigZ = sigC + softfloat_shiftRightJam64( sigProd, 32 - expDiff );
+        } else {
+            expZ = expProd;
+            sig64Z =
+                sigProd
+                    + softfloat_shiftRightJam64(
+                           sigC `<< 32, expDiff );
+            sigZ = softfloat_shiftRightJam64( sig64Z, 32 );
+        }
+        if ( sigZ < 0x40000000 ) {
+            expZ = expZ `- 1;
+            sigZ = sigZ `- 1;
+        }
+    } else {
+        sig64C = sigC `<< 32;
+        if ( expDiff < 0 ) {
+            signZ = signC;
+            expZ = expC;
+            sig64Z = sig64C - softfloat_shiftRightJam64( sigProd, -expDiff );
+        } else if (  expDiff == 0 ) {
+            expZ = expProd;
+            sig64Z = sigProd - sig64C;
+            if ( sig64Z == 0 ) {
+              return packToF32UI(
+                (mode == RoundingMode::RDN) ? 1 : 0, 0, 0 );
+            }
+            if ( (sig64Z &  0x8000000000000000) == 1  ) {
+                signZ = ! signZ;
+                sig64Z = -sig64Z;
+            }
+        } else {
+            expZ = expProd;
+            sig64Z = sigProd - softfloat_shiftRightJam64( sig64C, expDiff );
+        }
+        shiftDist = count_leading_zeros<64>( sig64Z ) - 1;
+        expZ = expZ - shiftDist;
+        shiftDist = shiftDist `- 32;
+        if ( shiftDist < 0 ) {
+            sigZ = softfloat_shiftRightJam64( sig64Z, -shiftDist );
+        } else {
+            sigZ = sig64Z<<shiftDist;
+        }
+
+    }
+  }
+}
