diff --git a/tools/scripts/pages.html.erb b/tools/scripts/pages.html.erb
index 7fa8fe5bb2..3999422e10 100644
--- a/tools/scripts/pages.html.erb
+++ b/tools/scripts/pages.html.erb
@@ -73,9 +73,9 @@
           Candidate replacement for <a href="https://docs.google.com/spreadsheets/d/1A40dfm0nnn2-tgKIhdi3UYQ1GBr8iRiV2edFowvgp7E/edit?gid=1157775000">Profiles & Bases & Extensions Google Sheet</a>
           using data in riscv-unified-db.
           <ul>
-            <li><a href="<%= ENV["PAGES_URL"] %>/isa_explorer/ext_table.html">Extensions</a></li>
-            <li><a href="<%= ENV["PAGES_URL"] %>/isa_explorer/inst_table.html">Instructions</a></li>
-            <li><a href="<%= ENV["PAGES_URL"] %>/isa_explorer/csr_table.html">CSRs</a></li>
+            <li><a href="<%= ENV["PAGES_URL"] %>/isa_explorer/ext-explorer.html">Extensions</a></li>
+            <li><a href="<%= ENV["PAGES_URL"] %>/isa_explorer/inst-explorer.html">Instructions</a></li>
+            <li><a href="<%= ENV["PAGES_URL"] %>/isa_explorer/csr-explorer.html">CSRs</a></li>
             <li><a href="<%= ENV["PAGES_URL"] %>/isa_explorer/isa_explorer.xlsx">Excel version (includes Extensions, Instructions, CSRs)</a></li>
           </ul>
         </div>
