diff --git a/iommu_ref_model/libiommu/include/iommu_req_rsp.h b/iommu_ref_model/libiommu/include/iommu_req_rsp.h
index 38513aa2..091c7004 100644
--- a/iommu_ref_model/libiommu/include/iommu_req_rsp.h
+++ b/iommu_ref_model/libiommu/include/iommu_req_rsp.h
@@ -44,6 +44,7 @@ typedef struct {
     uint8_t  exec_req;
     uint8_t  priv_req;
     uint8_t  is_cxl_dev;
+    uint8_t  orig_pid_valid;
     // Translation request
     iommu_trans_req_t tr;
 } hb_to_iommu_req_t;
diff --git a/iommu_ref_model/libiommu/src/iommu_translate.c b/iommu_ref_model/libiommu/src/iommu_translate.c
index 418aafeb..32496ed9 100644
--- a/iommu_ref_model/libiommu/src/iommu_translate.c
+++ b/iommu_ref_model/libiommu/src/iommu_translate.c
@@ -26,7 +26,7 @@ iommu_translate_iova(
     uint32_t DID, PID, GSCID, PSCID;
     pte_t vs_pte;
     gpte_t g_pte;
-    uint8_t ioatc_status, gst_fault, is_implicit, pasid_valid;
+    uint8_t ioatc_status, gst_fault, is_implicit;
     uint64_t napot_ppn, napot_iova, napot_gpa;
 
     // Classify transaction type
@@ -34,7 +34,7 @@ iommu_translate_iova(
     iotval = req->tr.iova;
     DTF = 0;
     PSCID = 0;
-    pasid_valid = req->pid_valid;
+    req->orig_pid_valid = req->pid_valid;
 
     // Count events
     if ( req->tr.at == ADDR_TYPE_UNTRANSLATED )
@@ -471,12 +471,12 @@ iommu_translate_iova(
         // in MRIF mode then a Success response is generated with R, W, and U bit set to
         // 1. The U bit being set to 1 in the response instructs the device that it must
         // only use Untranslated requests to access the implied 4 KiB memory range
-        rsp_msg->trsp.Priv   = (pasid_valid && req->priv_req) ? 1 : 0;
+        rsp_msg->trsp.Priv   = (req->orig_pid_valid && req->priv_req) ? 1 : 0;
         rsp_msg->trsp.CXL_IO = (req->is_cxl_dev &&
                                 ((vs_pte.PBMT != PMA) || (is_msi == 1) || (DC.tc.T2GPA == 1))) ? 1 : 0;
         rsp_msg->trsp.N      = 0;
         rsp_msg->trsp.AMA    = 0;
-        rsp_msg->trsp.Global = pasid_valid && (is_msi == 0) ? vs_pte.G : 0;
+        rsp_msg->trsp.Global = req->orig_pid_valid && (is_msi == 0) ? vs_pte.G : 0;
         rsp_msg->trsp.U      = (is_msi & is_mrif);
         rsp_msg->trsp.R      = (vs_pte.R & g_pte.R);
         rsp_msg->trsp.W      = (vs_pte.W & g_pte.W & vs_pte.D & g_pte.D);
@@ -548,7 +548,7 @@ iommu_translate_iova(
          (cause == 268) || (cause == 269) || (cause == 270) ||
          (cause == 271) || (cause == 272) || (cause == 274) ) {
         report_fault(iommu, cause, iotval, iotval2, TTYP, DTF,
-                     req->device_id, req->pid_valid, req->process_id, req->priv_req);
+                     req->device_id, req->orig_pid_valid, req->process_id, req->priv_req);
         goto return_completer_abort;
     }
 
@@ -563,7 +563,7 @@ iommu_translate_iova(
     if ( (cause == 256) || (cause == 257) || (cause == 258) || (cause == 259) ||
         (cause == 260) ) {
         report_fault(iommu, cause, iotval, iotval2, TTYP, DTF,
-                     req->device_id, req->pid_valid, req->process_id, req->priv_req);
+                     req->device_id, req->orig_pid_valid, req->process_id, req->priv_req);
         goto return_unsupported_request;
     }
 
