diff --git a/Makefile b/Makefile
index 4991f57..354cf3d 100644
--- a/Makefile
+++ b/Makefile
@@ -24,7 +24,7 @@ REVMARK ?= Stable
 DATE ?= $(shell date +%Y-%m-%d)
 ifneq ($(SKIP_DOCKER),true)
 	DOCKER_CMD := docker run --rm -v ${PWD}:/build -w /build \
-	riscvintl/riscv-docs-base-container-image:latest \
+	ghcr.io/riscv/riscv-docs-base-container-image:latest \
 	/bin/sh -c
 	DOCKER_QUOTE := "
 endif
