diff --git a/iommu_ref_model/libiommu/src/iommu_process_context.c b/iommu_ref_model/libiommu/src/iommu_process_context.c
index d89891c1..b8c543f0 100644
--- a/iommu_ref_model/libiommu/src/iommu_process_context.c
+++ b/iommu_ref_model/libiommu/src/iommu_process_context.c
@@ -106,7 +106,7 @@ locate_process_context(
             return 1;
         }
         if ( gst_fault == GST_DATA_CORRUPTION ) {
-            *cause = 274;           // Second-stage PT load data corruption fault
+            *cause = 269;           // PDT data corruption fault
             return 1;
         }
     }
