diff --git a/tools/ruby-gems/idlc/lib/idlc/ast.rb b/tools/ruby-gems/idlc/lib/idlc/ast.rb
index dc4e903969..d4d3f72130 100644
--- a/tools/ruby-gems/idlc/lib/idlc/ast.rb
+++ b/tools/ruby-gems/idlc/lib/idlc/ast.rb
@@ -4190,9 +4190,9 @@ def value(symtab)
         elsif op == "|"
           # if one side is all ones, we don't need to know the other side
           rhs_type = rhs.type(symtab)
-          value_error("Unknown width") if rhs_type.width == :unknown
+          value_error("Unknown width") if rhs_type.width == :unknown || rhs_type.width.nil?
           lhs_type = lhs.type(symtab)
-          value_error("unknown width") if lhs_type.width == :unknown
+          value_error("unknown width") if lhs_type.width == :unknown || lhs_type.width.nil?
 
           value_result = value_try do
             rhs_mask = ((1 << rhs_type.width) - 1)
diff --git a/tools/ruby-gems/idlc/lib/idlc/passes/prune.rb b/tools/ruby-gems/idlc/lib/idlc/passes/prune.rb
index b9d472a8c7..1359ec01e0 100644
--- a/tools/ruby-gems/idlc/lib/idlc/passes/prune.rb
+++ b/tools/ruby-gems/idlc/lib/idlc/passes/prune.rb
@@ -282,13 +282,13 @@ def prune(symtab)
         if lhs_value == 0
           # 0 & anything == 0
           create_literal(symtab, 0, type(symtab))
-        elsif (rhs.type(symtab).width != :unknown) && lhs_value == ((1 << rhs.type(symtab).width) - 1)
+        elsif (rhs.type(symtab).width.is_a?(Integer)) && lhs_value == ((1 << rhs.type(symtab).width) - 1)
           # rhs idenntity
           rhs.prune(symtab)
         elsif rhs_value == 0
           # anything & 0 == 0
           create_literal(symtab, 0, type(symtab))
-        elsif (lhs.type(symtab).width != :unknown) && rhs_value == ((1 << lhs.type(symtab).width) - 1)
+        elsif (lhs.type(symtab).width.is_a?(Integer)) && rhs_value == ((1 << lhs.type(symtab).width) - 1)
           # lhs identity
           lhs.prune(symtab)
         else
@@ -302,13 +302,13 @@ def prune(symtab)
         if lhs_value == 0
           # rhs idenntity
           rhs.prune(symtab)
-        elsif rhs_type.width != :unknown && lhs_value == ((1 << rhs.type(symtab).width) - 1)
+        elsif rhs_type.width.is_a?(Integer) && lhs_value == ((1 << rhs.type(symtab).width) - 1)
           # ~0 | anything == ~0
           create_literal(symtab, lhs_value, type(symtab))
         elsif rhs_value == 0
           # lhs identity
           lhs.prune(symtab)
-        elsif lhs_type.width != :unknown && rhs_value == ((1 << lhs.type(symtab).width) - 1)
+        elsif lhs_type.width.is_a?(Integer) && rhs_value == ((1 << lhs.type(symtab).width) - 1)
           # anything | ~0 == ~0
           create_literal(symtab, rhs_value, type(symtab))
         else
diff --git a/tools/ruby-gems/idlc/lib/idlc/type.rb b/tools/ruby-gems/idlc/lib/idlc/type.rb
index 1a97f07f44..aed80a7dca 100644
--- a/tools/ruby-gems/idlc/lib/idlc/type.rb
+++ b/tools/ruby-gems/idlc/lib/idlc/type.rb
@@ -82,7 +82,7 @@ def default
     sig { returns(T::Array[Symbol]) }
     attr_reader :qualifiers
 
-    sig { returns(T.any(Integer, Symbol)) }
+    sig { returns(T.nilable(T.any(Integer, Symbol))) }
     attr_reader :width
 
     sig { returns(T.nilable(AstNode)) }
