diff --git a/backends/cpp_hart_gen/tasks.rake b/backends/cpp_hart_gen/tasks.rake
index d6f789a669..b8bde94323 100644
--- a/backends/cpp_hart_gen/tasks.rake
+++ b/backends/cpp_hart_gen/tasks.rake
@@ -392,8 +392,14 @@ end
 task checkout_riscv_tests: "#{$root}/ext/riscv-tests/env/LICENSE"
 
 task build_riscv_tests: "checkout_riscv_tests" do
+  configs_name, build_name = configs_build_name
+
   Dir.chdir "#{$root}/tests/isa" do
-    sh "make"
+    if configs_name[0] == "rv32"
+      sh "make XLEN=32"
+    else
+      sh "make"
+    end
   end
 end
 
@@ -505,12 +511,13 @@ namespace :test do
   end
 
   task riscv_vector_tests: ["build_riscv_tests", "build:iss"] do
-    _, build_name = configs_build_name
+    configs_name, build_name = configs_build_name
 
     # These extensions to the riscv-tests suite have binaries under a different diretcory
-    rv32uvTests = ["vsetivli", "vsetvl", "vsetvli_rs1_eq_zero", "vsetvli_vl_lt_vlmax"]
-    rv32uvTests.each do |t|
-      sh "#{CPP_HART_GEN_DST}/#{build_name}/build/iss -m rv32 -c #{$root}/cfgs/rv32-vector.yaml tests/isa/rv32uv-p-#{t}"
+    # uvTests are common for rv32/64
+    uvTests = ["vsetivli", "vsetvl", "vsetvli_rs1_eq_zero", "vsetvli_vl_lt_vlmax"]
+    uvTests.each do |t|
+      sh "#{CPP_HART_GEN_DST}/#{build_name}/build/iss -m #{configs_name[0]} -c #{$root}/cfgs/#{configs_name[0]}-vector.yaml tests/isa/#{configs_name[0]}uv-p-#{t}"
     end
   end
 end
diff --git a/bin/clean b/bin/clean
index a2aa98a813..0a03ae6da9 100755
--- a/bin/clean
+++ b/bin/clean
@@ -3,3 +3,5 @@
 ROOT=$(dirname "$(dirname "$(realpath "${BASH_SOURCE[0]}")")")
 
 rm -rf "${ROOT}"/.stamps "${ROOT}"/gen
+make -C "${ROOT}"/tests/isa clean
+make -C "${ROOT}"/tests/isa XLEN=32 clean
diff --git a/ext/riscv-tests b/ext/riscv-tests
index 355f1fb9b8..f443f44860 160000
--- a/ext/riscv-tests
+++ b/ext/riscv-tests
@@ -1 +1 @@
-Subproject commit 355f1fb9b8a7142bee718e4123fea0aa8f49c319
+Subproject commit f443f4486085132552c9b43527fb0be5efa3cc0c
diff --git a/tests/isa/Makefile b/tests/isa/Makefile
index e81905276b..eb71a6d011 100644
--- a/tests/isa/Makefile
+++ b/tests/isa/Makefile
@@ -11,8 +11,9 @@ riscv_tests_dir := $(src_dir)/../../ext/riscv-tests
 
 ifeq ($(XLEN),64)
 include $(src_dir)/rv64uv/Makefrag
-endif
+else
 include $(src_dir)/rv32uv/Makefrag
+endif
 
 default: all
 
@@ -75,11 +76,11 @@ all: $(tests_dump) riscv_tests
 #------------------------------------------------------------
 # riscv-tests targets
 riscv_tests:
-	$(MAKE) -C $(riscv_tests_dir)/isa RISCV_GCC_OPTS='$(RISCV_GCC_OPTS)'
+	$(MAKE) -C $(riscv_tests_dir)/isa XLEN=$(XLEN) RISCV_GCC_OPTS='$(RISCV_GCC_OPTS)'
 
 #------------------------------------------------------------
 # Clean up
 
 clean:
-	$(MAKE) -C $(riscv_tests_dir)/isa clean
+	$(MAKE) -C $(riscv_tests_dir)/isa XLEN=$(XLEN) clean
 	rm -rf $(junk)
