diff --git a/backends/cpp_hart_gen/tasks.rake b/backends/cpp_hart_gen/tasks.rake
index 6d7ce8ae2f..9c25b1d31f 100644
--- a/backends/cpp_hart_gen/tasks.rake
+++ b/backends/cpp_hart_gen/tasks.rake
@@ -530,7 +530,7 @@ namespace :test do
     # These extensions to the riscv-tests suite have binaries under a different diretcory
     # uvTests are common for rv32/64
     uvTests = ["vsetivli", "vsetvl", "vsetvli_rs1_eq_zero", "vsetvli_vl_lt_vlmax",
-                "vle8", "vmv_v_i", "vadd.vv"]
+                "vle8", "vmv_v_i", "vadd.vv", "vnsra.wi"]
     uvTests.each do |t|
       sh "#{CPP_HART_GEN_DST}/#{build_name}/build/iss -m #{configs_name[0]} -c #{$root}/cfgs/#{configs_name[0]}-vector.yaml tests/isa/#{configs_name[0]}uv-p-#{t}"
     end
diff --git a/spec/std/isa/inst/V/vnsra.wv.yaml b/spec/std/isa/inst/V/vnsra.wv.yaml
index b0b14d80da..c7960f1e93 100644
--- a/spec/std/isa/inst/V/vnsra.wv.yaml
+++ b/spec/std/isa/inst/V/vnsra.wv.yaml
@@ -31,3 +31,41 @@ access:
   vu: always
 data_independent_timing: false
 operation(): |
+  VectorState state = vector_state();
+  VectorLmulType lmul_type = state.lmul_type;
+
+  XReg vlen = VLEN;
+  XReg vlmax;
+  XReg narrow_vs2_lmul;
+  XReg lmul_inc;
+
+  if (lmul_type == VectorLmulType::Multiply) {
+    vlmax = (vlen << state.log2_lmul) >> state.log2_sew;
+  } else {
+    vlmax = (vlen >> state.log2_lmul) >> state.log2_sew;
+  }
+
+  for (U32 i = CSR[vstart].VALUE; i < vlmax; i++) {
+    U32 start_bit_pos   = i * state.sew;
+    U32 end_bit_pos     = start_bit_pos + state.sew - 1;
+    U32 n_start_bit_pos = (i * state.sew *2) % vlen;
+    U32 n_end_bit_pos   = n_start_bit_pos + (state.sew*2) - 1;
+    narrow_vs2_lmul = (i * state.sew *2) / vlen;
+    lmul_inc = (i * state.sew) / vlen;
+    U32 shift_amount = (((v[vs1+lmul_inc])[end_bit_pos:start_bit_pos])%(state.sew*2));
+    if (CSR[vl].VALUE > i){
+      v[vd+lmul_inc] = {
+        v[vd+lmul_inc][VLEN-1:end_bit_pos + 1],
+        mask_elements(vm,(v[vs2+narrow_vs2_lmul])
+        [n_end_bit_pos:n_start_bit_pos] >>> shift_amount)[end_bit_pos:start_bit_pos],
+        v[vd+lmul_inc][start_bit_pos-1:0]
+      };
+    } else {
+        v[vd+lmul_inc] = {
+        v[vd+lmul_inc][VLEN-1:end_bit_pos + 1],
+        tail_elements((v[vd+lmul_inc])[end_bit_pos:start_bit_pos])[end_bit_pos:start_bit_pos],
+        v[vd+lmul_inc][start_bit_pos-1:0]
+      };
+    }
+  }
+  CSR[vstart].VALUE = 0;
\ No newline at end of file
diff --git a/spec/std/isa/inst/V/vnsrl.wi.yaml b/spec/std/isa/inst/V/vnsrl.wi.yaml
index 2f9550c8a8..c8023b351e 100644
--- a/spec/std/isa/inst/V/vnsrl.wi.yaml
+++ b/spec/std/isa/inst/V/vnsrl.wi.yaml
@@ -31,3 +31,41 @@ access:
   vu: always
 data_independent_timing: false
 operation(): |
+  VectorState state = vector_state();
+  VectorLmulType lmul_type = state.lmul_type;
+
+  XReg vlen = VLEN;
+  XReg vlmax;
+  XReg narrow_vs2_lmul;
+  XReg lmul_inc;
+
+  if (lmul_type == VectorLmulType::Multiply) {
+    vlmax = (vlen << state.log2_lmul) >> state.log2_sew;
+  } else {
+    vlmax = (vlen >> state.log2_lmul) >> state.log2_sew;
+  }
+
+  for (U32 i = CSR[vstart].VALUE; i < vlmax; i++) {
+    U32 start_bit_pos   = i * state.sew;
+    U32 end_bit_pos     = start_bit_pos + state.sew - 1;
+    U32 n_start_bit_pos = (i * state.sew *2) % vlen;
+    U32 n_end_bit_pos   = n_start_bit_pos + (state.sew*2) - 1;
+    narrow_vs2_lmul = (i * state.sew *2) / vlen;
+    lmul_inc = (i * state.sew) / vlen;
+    U32 shift_amount = ((imm)%(state.sew*2));
+    if (CSR[vl].VALUE > i){
+      v[vd+lmul_inc] = {
+        v[vd+lmul_inc][VLEN-1:end_bit_pos + 1],
+        mask_elements(vm,(v[vs2+narrow_vs2_lmul])
+        [n_end_bit_pos:n_start_bit_pos] >>> shift_amount),
+        v[vd+lmul_inc][start_bit_pos-1:0]
+      };
+    } else {
+        v[vd+lmul_inc] = {
+        v[vd+lmul_inc][VLEN-1:end_bit_pos + 1],
+        tail_elements((v[vd+lmul_inc])[end_bit_pos:start_bit_pos])[end_bit_pos:start_bit_pos],
+        v[vd+lmul_inc][start_bit_pos-1:0]
+      };
+    }
+  }
+  CSR[vstart].VALUE = 0;
\ No newline at end of file
diff --git a/spec/std/isa/isa/vec.idl b/spec/std/isa/isa/vec.idl
index 995d6de4d9..bea0c668d6 100644
--- a/spec/std/isa/isa/vec.idl
+++ b/spec/std/isa/isa/vec.idl
@@ -53,3 +53,40 @@ function vector_state {
     return state;
   }
 }
+
+
+function mask_elements {
+    returns Bits<64>
+    arguments
+        Bits<1> ins_mask_bit,
+        Bits<64> result
+    description {
+        Determines the value for masked elements based on mask policy.
+        Returns the original value if undisturbed, or all 1s if agnostic.
+    }
+    body {
+        if (CSR[vtype].VMA == 0 || ins_mask_bit==1) {
+            return result;
+        } else {
+            return 64'hffff_ffff_ffff_ffff;
+        }
+    }
+}
+
+function tail_elements {
+    returns Bits<64>
+    arguments
+        Bits<64> result
+    description {
+        Determines the value for tail elements based on the selected tail policy.
+        Returns the original value if undisturbed, or all 1s if agnostic.
+        Apply the tail policy when v[vstart] > VL.
+    }
+    body {
+        if (CSR[vtype].VTA==1) {
+            return 64'hffff_ffff_ffff_ffff;
+        } else {
+            return result;
+        }
+    }
+}
diff --git a/tests/isa/rv64uv/vnsra.wi.S b/tests/isa/rv64uv/vnsra.wi.S
new file mode 100644
index 0000000000..7000aad404
--- /dev/null
+++ b/tests/isa/rv64uv/vnsra.wi.S
@@ -0,0 +1,80 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+# See LICENSE for license details
+
+#*****************************************************************************
+# vnsra.wi.S
+#-----------------------------------------------------------------------------
+#
+# Test vnsra.wi instruction.
+#
+# vnsra.wi
+
+#include "riscv_test.h"
+#include "test_macros.h"
+
+RVTEST_RV64UV
+RVTEST_CODE_BEGIN
+
+# Write tests for vnsra.wi
+# vlen = 128
+# SEW=8, LMUL=1. x7 should contain VL == 10 (AVL, since its < VLMAX==128)
+# VLMAX=vlen*LMUL/sew, so VLMAX=16.
+# Use x5 to hold addresses
+
+# Testcase 1: SEW = 8, LMUL = 1,Mask Bit = 0, VL = 10
+    # IMM      = 4
+    # VS2 (v22)       = [3aef 5ff4 34b8 0fd4 f600 a3d1 195b 62bf fbba 7ae7 2c63 c847 0692 dadf 7ca6 60db]
+    # ELEMENTS          [--   --   --   --   --   --   ff   ff   ff   ff   ff   ff   ff   ff   ff   ff]
+    # VD (v28) BEFORE = [19   5b   62   bf   fb   ba   7a   e7   2c   63   c8   47   06   92   da   df]
+    # VD (v28) AFTER  = [19   5b   62   bf   fb   ba   95   2b   bb   ae   c6   84   69   ad   ca   0d]
+
+  TEST_CASE(1, x2, 0xbbaec68469adca0d, \
+# Set VL to VLMAX and load values into registers
+    li x7, -1; \
+    la x5, input1; \
+    li x6, 4; \
+    add x5, x5, x6; \
+    vsetvli x0, x7, e16,m2; \
+    vle16.v v22, (x5); \      # Load value into vs2
+    li x6, 4; \
+    add x5, x5, x6; \
+    vsetvli x0, x7, e16,m1; \
+    vle16.v v28, (x5); \       # Load value into vd
+    li x6, 4; \
+    add x5, x5, x6; \
+
+    li  x7, 10; \ # VL = 10
+    vsetvli x6, x7, e8,m1; \
+    vnsra.wi v28, v22, 4; \
+    la      x5, result; \
+    vse8.v  v28, (x5); \
+    lw      x2, (x5); \
+    )
+  TEST_PASSFAIL
+
+RVTEST_CODE_END
+
+# Input data section.
+# This section is optional, and this data is NOT saved in the output.
+.data
+  .balign 4
+input1:
+	.word 0xe54c8c1e
+	.word 0x7ca660db
+	.word 0x0692dadf
+	.word 0x2c63c847
+	.word 0xfbba7ae7
+	.word 0x195b62bf
+	.word 0xf600a3d1
+  .word 0x34b80fd4
+	.word 0x3aef5ff4
+
+# Output data section.
+# This section is optional, but this data IS saved in the output.
+RVTEST_DATA_BEGIN
+
+result:
+  .word -1
+
+RVTEST_DATA_END
