diff --git a/.devcontainer/Dockerfile b/.devcontainer/Dockerfile
index 484fffe5f5..14535c3c22 100644
--- a/.devcontainer/Dockerfile
+++ b/.devcontainer/Dockerfile
@@ -5,7 +5,7 @@ ENV DEBIAN_FRONTEND=noninteractive
 WORKDIR /workspace
 
 RUN export DEBIAN_FRONTEND=noninteractive
-RUN apt-get update
+RUN apt-get update --fix-missing
 
 # please keep pkgs sorted
 RUN apt-get install -y --no-install-recommends \
diff --git a/bin/.container-tag b/bin/.container-tag
index eb49d7c7fd..aec258df73 100644
--- a/bin/.container-tag
+++ b/bin/.container-tag
@@ -1 +1 @@
-0.7
+0.8
diff --git a/container.def b/container.def
index 6e77f80c0e..40ba3beed7 100644
--- a/container.def
+++ b/container.def
@@ -9,7 +9,7 @@ From: ubuntu:24.04
 %post
   export DEBIAN_FRONTEND=noninteractive
 
-  apt-get update
+  apt-get update --fix-missing
 
   # please keep pkgs sorted
   apt-get install -y --no-install-recommends \
