diff --git a/Gemfile.lock b/Gemfile.lock
index bf136d0007..9b8ec3cb68 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -18,7 +18,7 @@ PATH
 PATH
   remote: tools/ruby-gems/udb-gen
   specs:
-    udb-gen (0.1.3)
+    udb-gen (0.1.4)
       asciidoctor
       asciidoctor-diagram
       asciidoctor-pdf
@@ -39,7 +39,7 @@ PATH
 PATH
   remote: tools/ruby-gems/udb
   specs:
-    udb (0.1.5)
+    udb (0.1.6)
       activesupport
       asciidoctor
       awesome_print
diff --git a/cfgs/mc100-32-full-example.yaml b/cfgs/mc100-32-full-example.yaml
index 06dbc5c05d..5e702ac099 100644
--- a/cfgs/mc100-32-full-example.yaml
+++ b/cfgs/mc100-32-full-example.yaml
@@ -161,38 +161,3 @@ params:
       false,
       false,
     ]
-  SCOUNTENABLE_EN:
-    [
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-      false,
-    ]
diff --git a/cfgs/qc_iu.yaml b/cfgs/qc_iu.yaml
index dc9d15847a..893d442c83 100644
--- a/cfgs/qc_iu.yaml
+++ b/cfgs/qc_iu.yaml
@@ -115,39 +115,6 @@ params:
     - false # HPM29
     - false # HPM30
     - false # HPM31
-  SCOUNTENABLE_EN:
-    - false # CY
-    - false # empty
-    - false # IR
-    - false # HPM3
-    - false # HPM4
-    - false # HPM5
-    - false # HPM6
-    - false # HPM7
-    - false # HPM8
-    - false # HPM9
-    - false # HPM10
-    - false # HPM11
-    - false # HPM12
-    - false # HPM13
-    - false # HPM14
-    - false # HPM15
-    - false # HPM16
-    - false # HPM17
-    - false # HPM18
-    - false # HPM19
-    - false # HPM20
-    - false # HPM21
-    - false # HPM22
-    - false # HPM23
-    - false # HPM24
-    - false # HPM25
-    - false # HPM26
-    - false # HPM27
-    - false # HPM28
-    - false # HPM29
-    - false # HPM30
-    - false # HPM31
   COUNTINHIBIT_EN:
     - true # CY
     - false # empty
diff --git a/spec/std/isa/param/SCOUNTENABLE_EN.yaml b/spec/std/isa/param/SCOUNTENABLE_EN.yaml
index ab7bd0faee..46b4ce338d 100644
--- a/spec/std/isa/param/SCOUNTENABLE_EN.yaml
+++ b/spec/std/isa/param/SCOUNTENABLE_EN.yaml
@@ -39,4 +39,4 @@ requirements:
     When mhpmcounter[i] does not exist, it cannot be enabled.
 definedBy:
   extension:
-    name: Sm
+    name: S
diff --git a/tools/ruby-gems/udb-gen/Gemfile.lock b/tools/ruby-gems/udb-gen/Gemfile.lock
index 08d57ff5a7..0fb2115406 100644
--- a/tools/ruby-gems/udb-gen/Gemfile.lock
+++ b/tools/ruby-gems/udb-gen/Gemfile.lock
@@ -17,7 +17,7 @@ PATH
 PATH
   remote: ../udb
   specs:
-    udb (0.1.5)
+    udb (0.1.6)
       activesupport
       asciidoctor
       awesome_print
@@ -41,7 +41,7 @@ PATH
 PATH
   remote: .
   specs:
-    udb-gen (0.1.3)
+    udb-gen (0.1.4)
       asciidoctor
       asciidoctor-diagram
       asciidoctor-pdf
diff --git a/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb b/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb
index 44c23ae962..66abcb7e87 100644
--- a/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb
+++ b/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb
@@ -5,5 +5,5 @@
 # frozen_string_literal: true
 
 module UdbGen
-  def self.version = "0.1.3"
+  def self.version = "0.1.4"
 end
diff --git a/tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.5.rbi b/tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.6.rbi
similarity index 100%
rename from tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.5.rbi
rename to tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.6.rbi
diff --git a/tools/ruby-gems/udb/Gemfile.lock b/tools/ruby-gems/udb/Gemfile.lock
index 2c32d2bfa4..044cca16a2 100644
--- a/tools/ruby-gems/udb/Gemfile.lock
+++ b/tools/ruby-gems/udb/Gemfile.lock
@@ -17,7 +17,7 @@ PATH
 PATH
   remote: .
   specs:
-    udb (0.1.5)
+    udb (0.1.6)
       activesupport
       asciidoctor
       awesome_print
diff --git a/tools/ruby-gems/udb/lib/udb/version.rb b/tools/ruby-gems/udb/lib/udb/version.rb
index bca9263142..5a37006c58 100644
--- a/tools/ruby-gems/udb/lib/udb/version.rb
+++ b/tools/ruby-gems/udb/lib/udb/version.rb
@@ -5,5 +5,5 @@
 # frozen_string_literal: true
 
 module Udb
-  def self.version = "0.1.5"
+  def self.version = "0.1.6"
 end
