diff --git a/Makefile b/Makefile
index 3bddefa..c4e5762 100644
--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,7 @@ DATE ?= $(shell date +%Y-%m-%d)
 VERSION ?= v0.7.0
 REVMARK ?= Development
 DOCKER_RUN := docker run --rm -v ${PWD}:/build -w /build \
-riscvintl/riscv-docs-base-container-image:latest
+ghcr.io/riscv/riscv-docs-base-container-image:latest
 
 HEADER_SOURCE := zaamo-zalrsc.adoc
 PDF_RESULT := riscv-zaamo-zalrsc.pdf
