diff --git a/backends/manual/templates/playbook.yml.erb b/backends/manual/templates/playbook.yml.erb
index 7c5018fe02..4c1dd283c2 100644
--- a/backends/manual/templates/playbook.yml.erb
+++ b/backends/manual/templates/playbook.yml.erb
@@ -103,3 +103,7 @@ ui:
       .doc .admonitionblock td.icon i.icon-when::after {
         text-transform: none;
       }
+
+      .edit-this-page {
+         display: none !important;
+      }
