diff --git a/Gemfile.lock b/Gemfile.lock
index c9689d5c75..20fa72a92c 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -39,7 +39,7 @@ PATH
 PATH
   remote: tools/ruby-gems/udb
   specs:
-    udb (0.1.1)
+    udb (0.1.2)
       activesupport
       asciidoctor
       awesome_print
@@ -365,7 +365,7 @@ DEPENDENCIES
   yard-sorbet
 
 RUBY VERSION
-   ruby 3.4.8p72
+  ruby 3.4.8p72
 
 BUNDLED WITH
-   4.0.8
+  4.0.8
diff --git a/tests/data/golden/Xqci@0.13.0.adoc b/tests/data/golden/Xqci@0.13.0.adoc
index a044e33af6..bc99a8b401 100644
--- a/tests/data/golden/Xqci@0.13.0.adoc
+++ b/tests/data/golden/Xqci@0.13.0.adoc
@@ -1,6 +1,6 @@
 [[header]]
 :description: Qualcomm uC extensions (Xqci)
-:revdate: 2026-03-17
+:revdate: 2026-03-21
 :revnumber: 0.13.0
 :revmark: This document is in the http://riscv.org/spec-state[Frozen state]. + \
 + \
diff --git a/tools/ruby-gems/udb-gen/Gemfile.lock b/tools/ruby-gems/udb-gen/Gemfile.lock
index f1389ee138..48cf7a22e5 100644
--- a/tools/ruby-gems/udb-gen/Gemfile.lock
+++ b/tools/ruby-gems/udb-gen/Gemfile.lock
@@ -17,7 +17,7 @@ PATH
 PATH
   remote: ../udb
   specs:
-    udb (0.1.1)
+    udb (0.1.2)
       activesupport
       asciidoctor
       awesome_print
diff --git a/tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.1.rbi b/tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.2.rbi
similarity index 100%
rename from tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.1.rbi
rename to tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.2.rbi
diff --git a/tools/ruby-gems/udb/Gemfile.lock b/tools/ruby-gems/udb/Gemfile.lock
index 7c6e8cee3d..eb37ec8bdc 100644
--- a/tools/ruby-gems/udb/Gemfile.lock
+++ b/tools/ruby-gems/udb/Gemfile.lock
@@ -17,7 +17,7 @@ PATH
 PATH
   remote: .
   specs:
-    udb (0.1.1)
+    udb (0.1.2)
       activesupport
       asciidoctor
       awesome_print
diff --git a/tools/ruby-gems/udb/lib/udb/doc_link.rb b/tools/ruby-gems/udb/lib/udb/doc_link.rb
index f446bd1d59..9a22781641 100644
--- a/tools/ruby-gems/udb/lib/udb/doc_link.rb
+++ b/tools/ruby-gems/udb/lib/udb/doc_link.rb
@@ -1,3 +1,4 @@
+# typed: false
 # Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
 # SPDX-License-Identifier: BSD-3-Clause-Clear
 
@@ -29,21 +30,23 @@
 #                   where <id> is the ID of the normative rule
 #   IDL code      idl:code:inst:<inst-name>:<location>
 #                 TODO for CSR and CSR Fields
-class Udb::DocLink
-  # @param dst_link [String] The documentation link provided in the YAML
-  # @param db_obj [String] Database object
-  def initialize(dst_link, db_obj)
-    raise ArgumentError, "Need String but was passed a #{data.class}" unless dst_link.is_a?(String)
-    @dst_link = dst_link
+module Udb
+  class DocLink
+    # @param dst_link [String] The documentation link provided in the YAML
+    # @param db_obj [String] Database object
+    def initialize(dst_link, db_obj)
+      raise ArgumentError, "Need String but was passed a #{data.class}" unless dst_link.is_a?(String)
+      @dst_link = dst_link
 
-    raise ArgumentError, "Missing documentation link for #{db_obj.name} of kind #{db_obj.kind}" if @dst_link.nil?
-  end
+      raise ArgumentError, "Missing documentation link for #{db_obj.name} of kind #{db_obj.kind}" if @dst_link.nil?
+    end
 
-  # @return [String] Unique ID of the linked to normative rule
-  def dst_link = @dst_link
+    # @return [String] Unique ID of the linked to normative rule
+    def dst_link = @dst_link
 
-  # @return [String] Asciidoc to create desired link.
-  def to_adoc
-    "<<#{@dst_link},#{@dst_link}>>"
+    # @return [String] Asciidoc to create desired link.
+    def to_adoc
+      "<<#{@dst_link},#{@dst_link}>>"
+    end
   end
 end
diff --git a/tools/ruby-gems/udb/lib/udb/version.rb b/tools/ruby-gems/udb/lib/udb/version.rb
index 459f121485..6974fd6762 100644
--- a/tools/ruby-gems/udb/lib/udb/version.rb
+++ b/tools/ruby-gems/udb/lib/udb/version.rb
@@ -1,8 +1,9 @@
+# typed: false
 # Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
 # SPDX-License-Identifier: BSD-3-Clause-Clear
 
 # frozen_string_literal: true
 
 module Udb
-  def self.version = "0.1.1"
+  def self.version = "0.1.2"
 end
