diff --git a/iopmp_ref_model/include/iopmp_registers.h b/iopmp_ref_model/include/iopmp_registers.h
index a024ee9..b8a72e3 100644
--- a/iopmp_ref_model/include/iopmp_registers.h
+++ b/iopmp_ref_model/include/iopmp_registers.h
@@ -272,9 +272,9 @@ typedef union {
 typedef union {
     struct {
         uint32_t l   : 1;                   // ock bit to MDCFGLCK register.
-        uint32_t f   : 7;                   // Indicate the number of locked MDCFG entries
+        uint32_t f   : 6;                   // Indicate the number of locked MDCFG entries
                                             // MDCFG(i) is locked for i < f.
-        uint32_t rsv : 24;                  // Reserved for future use
+        uint32_t rsv : 25;                  // Reserved for future use
     };
     uint32_t raw;
 } mdcfglck_t;
