diff --git a/bin/asciidoctor b/bin/asciidoctor
index 82c3fbc4a2..ea43e71bc0 100755
--- a/bin/asciidoctor
+++ b/bin/asciidoctor
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 source $ROOT/setup
diff --git a/bin/bash b/bin/bash
index 3b0899f26c..10252c1155 100755
--- a/bin/bash
+++ b/bin/bash
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 source $ROOT/setup
diff --git a/bin/build_container b/bin/build_container
index 02588f5bad..da61bbb305 100755
--- a/bin/build_container
+++ b/bin/build_container
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(realpath $(dirname $(dirname $BASH_SOURCE[0])))
 
diff --git a/bin/bundle b/bin/bundle
index 30ccc96520..345dc47ee5 100755
--- a/bin/bundle
+++ b/bin/bundle
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 source $ROOT/setup
diff --git a/bin/clean b/bin/clean
index 22e665c5f6..8f9eaff2ff 100755
--- a/bin/clean
+++ b/bin/clean
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(dirname $(realpath $BASH_SOURCE[0])))
 
diff --git a/bin/node b/bin/node
index 64063b5d90..617416d976 100755
--- a/bin/node
+++ b/bin/node
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 source $ROOT/setup
diff --git a/bin/npx b/bin/npx
index 68138f347e..c4e156e2ca 100755
--- a/bin/npx
+++ b/bin/npx
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 source $ROOT/setup
diff --git a/bin/pip b/bin/pip
index fb39a98d5a..ba5dad6c90 100755
--- a/bin/pip
+++ b/bin/pip
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 source $ROOT/setup
diff --git a/bin/pre-commit b/bin/pre-commit
index bf5077bb17..53ee681aab 100755
--- a/bin/pre-commit
+++ b/bin/pre-commit
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(dirname $(realpath $BASH_SOURCE[0])))
 source $ROOT/bin/setup
diff --git a/bin/python b/bin/python
index 94bc144ead..3ae4083a5b 100755
--- a/bin/python
+++ b/bin/python
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 if [ ! -v NO_SOURCE ]; then
diff --git a/bin/ruby b/bin/ruby
index 669e2add7a..67e03fe5ad 100755
--- a/bin/ruby
+++ b/bin/ruby
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(realpath $BASH_SOURCE[0]))
 source $ROOT/setup
diff --git a/bin/setup b/bin/setup
index 57776bf342..5013effd1a 100755
--- a/bin/setup
+++ b/bin/setup
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
 ROOT=$(dirname $(dirname $(realpath $BASH_SOURCE[0])))
 OLDWD=$PWD
