diff --git a/iopmp_ref_model/README.md b/iopmp_ref_model/README.md
index 9b9d426..be27c53 100644
--- a/iopmp_ref_model/README.md
+++ b/iopmp_ref_model/README.md
@@ -24,7 +24,6 @@ The **IOPMP Reference Model** incorporates all IOPMP Models and each model can b
 | ------------------ | ------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | TOR_EN             | 0/1                       | Indicates if Top-Of-Range (TOR) addressing mode is supported.<br />**0**: TOR not supported. <br />**1**: TOR supported                                                                                                      |
 | SPS_EN             | 0/1                       | Indicates if Secondary Permission Settings (SPS) are supported.<br />**0**: SPS not supported. <br />**1**: SPS is supported                                                                                                    |
-| PRIENT_PROG        | 0/1                       | Indicates if the `prio_entry` field in HWCFG2 is programmable.<br /> **0**: Field is fixed. <br />**1**: Field is programmable                                                                                             |
 | RRID_TRANSL_EN     | 0/1                       | Indicates if tagging a new Requestor ID (RRID) on the initiator port is supported.<br />**0**: Tagging not supported. <br />**1**: New RRID tagging is supported.                                                            |
 | RRID_TRANSL_PROG   | 0/1                       | Indicates if the `rrid_transl` field is programmable. <br />**0**: Field is fixed. <br />**1**: Field is programmable                                                                                                      |
 | CHK_X              | 0/1                       | Indicates if the IOPMP implements check of an instruction fetches based on the instruction fetches fields<br />**0**: Fields of instruction fetches are ignored<br />**1**: Fields of instruction fetches are not ignored |
@@ -40,6 +39,8 @@ The **IOPMP Reference Model** incorporates all IOPMP Models and each model can b
 | RRID_NUM           | 0 - 65535                 | Indicates the maximum number of Requestor IDs (RRIDs) supported.                                                                                                                                                                         |
 | ENTRY_NUM          | 0 - IMP                   | Indicates the total number of entries supported.                                                                                                                                                                                         |
 | PRIO_ENTRY         | 1 - ENTRY_NUM             | Indicates the number of entries matched based on priority.                                                                                                                                                                               |
+| PRIO_ENT_PROG      | 0/1                       | Indicates if the `prio_entry` field in HWCFG2 is programmable.<br /> **0**: Field is fixed. <br />**1**: Field is programmable.    |
+| NON_PRIO_EN        | 0/1                       | Indicates whether the IOPMP supports non-priority entries.    |
 | ERROR_CAPTURE_EN   | 0/1                       | Indicates if the Error Capture Record feature is implemented.<br />**0**: Feature not supported. <br />**1**: Error details can be captured and logged.                                                                      |
 | IMP_ERROR_REQID    | 0/1                       | Indicates if `ERR_REQID` is implemented. <br />**0**: Feature not supported. <br />**1**: Errored RRID and Entry num is recorded.                                                                                          |
 | IMP_MDLCK          | 0/1                       | Indicates if the Memory Domain Lock (MDLCK) feature is implemented.<br />**0**: Feature is not implemented<br />**1**: Memory domains can be locked.                                                                         |
diff --git a/iopmp_ref_model/include/config.h b/iopmp_ref_model/include/config.h
index e972f8a..c3ed083 100644
--- a/iopmp_ref_model/include/config.h
+++ b/iopmp_ref_model/include/config.h
@@ -8,7 +8,6 @@
 
 #define IOPMP_TOR_EN            1
 #define IOPMP_SPS_EN            1
-#define IOPMP_PARIENT_PROG      0
 #define IOPMP_RRID_TRANSL_EN    1
 #define IOPMP_RRID_TRANSL_PROG  0
 #define IOPMP_CHK_X             1
@@ -23,7 +22,9 @@
 #define IOPMP_ADDRH_EN          1
 #define IOPMP_ENABLE            0
 #define IOPMP_ENTRY_NUM         512
-#define IOPMP_PRIO_ENTRY        16
+#define IOPMP_PRIO_ENTRY        16          // Depends on IOPMP_NON_PRIO_EN. If IOPMP_NON_PRIO_EN=0 the value should equal to IOPMP_ENTRY_NUM.
+#define IOPMP_PRIO_ENT_PROG     0           // Depends on IOPMP_NON_PRIO_EN.
+#define IOPMP_NON_PRIO_EN       1
 #define IOPMP_RRID_TRANSL       48
 
 #define USER                    0x80        // User-defined value for error suppression success responses.
diff --git a/iopmp_ref_model/include/iopmp_registers.h b/iopmp_ref_model/include/iopmp_registers.h
index dd72cc1..7d7b814 100644
--- a/iopmp_ref_model/include/iopmp_registers.h
+++ b/iopmp_ref_model/include/iopmp_registers.h
@@ -23,7 +23,8 @@
 #define HWCFG0_OFFSET         0x08
 #define HWCFG1_OFFSET         0x0C
 #define HWCFG2_OFFSET         0x10
-#define ENTRYOFFSET_OFFSET    0x14
+#define HWCFG3_OFFSET         0x14
+#define ENTRYOFFSET_OFFSET    0x2C
 #define MDSTALL_OFFSET        0x30
 #define MDSTALLH_OFFSET       0x34
 #define RRISCP_OFFSET         0x38
@@ -59,6 +60,14 @@
     typedef uint64_t reg_intf_dw;
 #endif
 
+// Indicate if HWCFG2 is implemented.
+#define IMP_HWCFG2              (IOPMP_NON_PRIO_EN | IOPMP_CHK_X | \
+                                 IOPMP_PEIS | IOPMP_PEES | IOPMP_SPS_EN | \
+                                 IOPMP_STALL_EN | IOPMP_MFR_EN)
+// Indicate if HWCFG3 is implemented.
+#define IMP_HWCFG3              ((MDCFG_FMT > 0) | (SRCMD_FMT > 0) | \
+                                 IOPMP_NO_X | IOPMP_NO_W | IOPMP_RRID_TRANSL_EN)
+
 extern int reset_iopmp(void);
 extern reg_intf_dw read_register(uint64_t offset, uint8_t num_bytes);
 extern void write_register(uint64_t offset, reg_intf_dw data, uint8_t num_bytes);
@@ -69,7 +78,6 @@ void rrid_stall_update(uint8_t exempt);
 // 1. vendor ID
 // 2. Specification Version
 typedef union {
-
     struct {
         uint32_t vendor  : 24;              // The vendor ID
         uint32_t specver : 8;               // The specification version
@@ -86,101 +94,97 @@ typedef union {
     uint32_t raw;
 } implementation_t;
 
-// HWCFG0 register is one of hrdware configuration registers reporting
+// HWCFG0 register is one of hardware configuration registers reporting
 // features supported by IOPMP. each bit if bot clear indicates
 // presence of that feature in IOPMP
 typedef union {
     struct {
-        uint32_t mdcfg_fmt          : 2;    // Indicate the MDCFG format
-                                            //  -> 0x0: Format 0. MDCFG table is implemented.
-                                            //  -> 0x1: Format 1. No MDCFG table. HWCFG.md_entry_num is fixed.
-                                            //  -> 0x2: Format 2. No MDCFG table. HWCFG.md_entry_num is programmable.
-                                            //  -> 0x3: reserved.
-
-        uint32_t srcmd_fmt          : 2;    // Indicate the MDCFG format
-                                            //  -> 0x0: Format 0. SRCMD_EN(s) and SRCMD_ENH(s) are available.
-                                            //  -> 0x1: Format 1. No SRCMD table.
-                                            //  -> 0x2: Format 2. SRCMD_PERM(m) and SRCMD_PERMH(m) are available.
-                                            //  -> 0x3: reserved.
-
-        uint32_t tor_en             : 1;    // Indicate if TOR is supported
-
-        uint32_t sps_en             : 1;    // Indicate secondary permission settings is supported; which are SRCMD_R/RH(i)
-                                            // and SRCMD_W/WH registers
-
-        uint32_t user_cfg_en        : 1;    // Indicate if user customized attributes is supported; which are
-                                            // ENTRY_USER_CFG(i) registers.
-
-        uint32_t prient_prog        : 1;    // A write-1-clear bit is sticky to 0 and indicates if HWCFG2.prio_entry
-                                            // is programmable. Reset to 1 if the implementation supports programmable
-                                            // prio_entry, otherwise, wired to 0.
-
-        uint32_t rrid_transl_en     : 1;    // Indicate the if tagging a new RRID on the requestor port is supported
-
-        uint32_t rrid_transl_prog   : 1;    // A write-1-set bit is sticky to 0 and indicate if the field sid_transl
-                                            // is programmable. Support only for rrid_transl_en=1, otherwise, wired to 0.
-
-        uint32_t chk_x              : 1;    // Indicate if the IOPMP implements the check of an
-                                            // instruction fetch. On chk_x=0, all fields of illegal
-                                            // instruction fetches are ignored, including
-                                            // HWCFG0.no_x, ENTRY_CFG(i).sixe, ENTRY_CFG(
-                                            // i).esxe, and ENTRY_CFG(i).x. It should be wired to
-                                            // zero if there is no indication for an instruction fetch
-                                            // otherwise, it should depend on x-bit in ENTRY_CFG(i).
-
-        uint32_t no_x               : 1;    // For chk_x=1, the IOPMP with no_x=1 always fails on
-                                            // an instruction fetch; otherwise, it should depend on
-                                            // x-bit in ENTRY_CFG(i). For chk_x=0, no_x has no
-                                            // effect
-
-        uint32_t no_w               : 1;    // Indicate if the IOPMP always fails write accesses
-
-        uint32_t stall_en           : 1;    // Indicate if the IOPMP implements stall-related features, which are MDSTALL,
-                                            // MDSTALLH, and RRIDSCP registers.
-
-        uint32_t peis               : 1;    // Indicate if the IOPMP implements interrupt suppression per entry
-
-        uint32_t pees               : 1;    // Indicate if the IOPMP implements the error suppression per entry
-
-        uint32_t mfr_en             : 1;    // Indicate if the IOPMP implements Multi Faults Record Extension
-
-        uint32_t md_entry_num       : 7;    // When HWCFG0.mdcfg_fmt =
-                                            //  -> 0x0: must be zero
-                                            //  -> 0x1 or 0x2: md_entry_num indicates each memory domain exactly has
-                                            //    (md_entry_num + 1) entries in a memory domain
-                                            // md_entry_num is locked if HWCFG0.enable is 1.
-
-        uint32_t md_num             : 6;    // Indicate the supported number of MD in the instance
-
-        uint32_t addrh_en           : 1;    // Indicate if the IOPMP implements ENTRY_ADDRH(i)
-
         uint32_t enable             : 1;    // Indicate if the IOPMP checks transactions by default.
                                             // If it is implemented, it should be initial to 0 and sticky to 1.
                                             // If it is not implemented, it should be wired to 1.
+        uint32_t HWCFG2_en          : 1;    // Indicate if HWCFG2 is implemented.
+        uint32_t HWCFG3_en          : 1;    // Indicate if HWCFG3 is implemented.
+        uint32_t rsv                : 21;   // Must be zero on write, reserved for future.
+        uint32_t md_num             : 6;    // Indicate the supported number of MD in the instance
+        uint32_t addrh_en           : 1;    // Indicate if ENTRY_ADDRH(i) and ERR_REQADDRH are available.
+        uint32_t tor_en             : 1;    // Indicate if TOR is supported
     };
     uint32_t raw;
 } hwcfg0_t;
 
-// HWCFG1 register is one of hrdware configuration
+// HWCFG1 register is one of hardware configuration
 // registers reporting features supported by IOPMP.
 typedef union {
     struct {
         uint32_t rrid_num  : 16;            // Indicate the supported number of RRID in the instance
-        uint32_t entry_num : 16;            // Indicate the supported number of entries in the instance
+        uint32_t entry_num : 16;            // Indicate the supported number of entries in the instance, which should be
+                                            // larger than zero.
     };
     uint32_t raw;
 } hwcfg1_t;
 
-// HWCFG2 register is one of hrdware configuration
-// registers reporting features supported by IOPMP.
+// HWCFG2 register is one of hardware configuration registers indicating the
+// extended configurations of current IOPMP instance for extension.
 typedef union {
     struct {
-        uint32_t prio_entry  : 16;          // Indicate the supported number of RRID in the instance
-        uint32_t rrid_transl : 16;          // Indicate the supported number of entries in the instance
+        uint32_t prio_entry    : 16;        // Indicate the number of entries matched with priority. These rules should
+                                            // be placed in the lowest order. Within these rules, the lower order has a
+                                            // higher priority.
+        uint32_t prio_ent_prog : 1;         // A write-1-clear bit is sticky to 0 and indicates if HWCFG2.prio_entry
+                                            // is programmable. Reset to 1 if the implementation supports programmable
+                                            // prio_entry, otherwise, wired to 0.
+        uint32_t non_prio_en   : 1;         // Indicates whether the IOPMP supports non-priority entries.
+        uint32_t rsv           : 8;         // Must be zero on write, reserved for future.
+        uint32_t chk_x         : 1;         // Indicate if the IOPMP implements the check of an instruction fetch.
+        uint32_t peis          : 1;         // Indicate if the IOPMP implements interrupt suppression per entry,
+                                            // including fields sire and siwe in ENTRY_CFG(i), i = 0…HWCFG1.entry_num-1.
+        uint32_t pees          : 1;         // Indicate if the IOPMP implements the error suppression per entry,
+                                            // including fields sere and sewe in ENTRY_CFG(i), i = 0…HWCFG1.entry_num-1.
+        uint32_t sps_en        : 1;         // Indicates secondary permission settings are supported; which are
+                                            // SRCMD_R/RH(s) and SRCMD_W/WH(s) registers, s = 0…HWCFG1.rrid_num - 1.
+        uint32_t stall_en      : 1;         // Indicate if the IOPMP implements stall-related features, which are MDSTALL,
+                                            // MDSTALLH, and RRIDSCP registers.
+        uint32_t mfr_en        : 1;         // Indicate if the IOPMP implements Multi Faults Record, that is ERR_MFR and
+                                            // ERR_INFO.svc.
     };
     uint32_t raw;
 } hwcfg2_t;
 
+// HWCFG3 register is one of hardware configuration registers indicating the
+// extended configurations of current IOPMP instance for application note.
+typedef union {
+    struct {
+        uint32_t mdcfg_fmt        : 2;      // Indicate the MDCFG format
+                                            //  -> 0x0: Format 0. MDCFG table is implemented.
+                                            //  -> 0x1: Format 1. No MDCFG table. HWCFG3.md_entry_num is fixed.
+                                            //  -> 0x2: Format 2. No MDCFG table. HWCFG3.md_entry_num is programmable.
+                                            //  -> 0x3: reserved.
+        uint32_t srcmd_fmt        : 2;      // Indicate the SRCMD Table format
+                                            //  -> 0x0: Format 0 (baseline). SRCMD_EN(s) and SRCMD_ENH(s) are available.
+                                            //  -> 0x1: Format 1. Exclusive Format. No SRCMD Table. RRID i associates
+                                            //     exclusively with memory domain i, i = 0…HWCFG0.md_num-1.
+                                            //  -> 0x2: Format 2. MD-indexed Format. SRCMD_PERM(m) and SRCMD_PERMH(m)
+                                            //     are available.
+                                            //  -> 0x3: reserved.
+        uint32_t md_entry_num     : 8;      // When HWCFG3.mdcfg_fmt =
+                                            //  -> 0x0: must be zero
+                                            //  -> 0x1 or 0x2: md_entry_num indicates each memory domain has exactly
+                                            //    (md_entry_num + 1) entries
+                                            // md_entry_num is locked if HWCFG0.enable is 1.
+        uint32_t no_x             : 1;      // For HWCFG2.chk_x=1, when no_x=1, the IOPMP denies all instruction fetch
+                                            // transactions; otherwise, it depends on the x-bit in ENTRY_CFG(i). For
+                                            // chk_x=0, no_x has no effect.
+        uint32_t no_w             : 1;      // Indicate if the IOPMP always denies write accesses as if no rule matched.
+        uint32_t rrid_transl_en   : 1;      // Indicate the if tagging a new RRID on the initiator port is supported
+        uint32_t rrid_transl_prog : 1;      // A write-1-clear bit that is sticky to 0. Indicates if the rrid_transl
+                                            // field is programmable. Supported only for rrid_transl_en=1, otherwise,
+                                            // wired to 0.
+        uint32_t rrid_transl      : 16;     // The RRID tagged to outgoing transactions. Supported only for
+                                            // rrid_transl_en=1. It is writable only when rrid_transl_prog=1.
+    };
+    uint32_t raw;
+} hwcfg3_t;
+
 // ENTRYOFFSET register ndicates the internal address
 // offsets of each table.
 typedef union {
@@ -341,7 +345,7 @@ typedef union {
                                             // 0x0E ~ 0x0F = user-defined error
 
         uint32_t svc   : 1;                 // Indicate there is a subsequent violation caught in ERR_MFR.
-                                            // Implemented only for HWCFG0.mfr_en=1,
+                                            // Implemented only for HWCFG2.mfr_en=1,
 
         uint32_t rsv   : 23;                // reserved for future use
     };
@@ -587,25 +591,24 @@ typedef union {
                                             // 0x2: NA4
                                             // 0x3: NAPOT
 
-        uint32_t sire : 1;                  // To suppress interrupt for an illegal read access caught by the entry
-        uint32_t siwe : 1;                  // Suppress interrupt for write violations caught by the entry
-        uint32_t sixe : 1;                  // Suppress interrupt on an illegal instruction fetch caught by the entry
+        uint32_t sire : 1;                  // Suppress interrupt for an illegal read access caught by the entry.
+        uint32_t siwe : 1;                  // Suppress interrupt for an illegal write access/AMO caught by the entry.
+        uint32_t sixe : 1;                  // Suppress interrupt on an illegal instruction fetch caught by the entry.
         uint32_t sere : 1;                  // Supress the (bus) error on an illegal read access caught by the entry
-                                            // 0x0: the response by ERR_CFG.rre
-                                            // 0x1: do not respond an error. User to define the behavior,
-                                            //      e.g., respond a success with an implementation-dependent
-                                            //      value to the requestor.
-
+                                            // * 0x0: respond an error if ERR_CFG.rs is 0x0.
+                                            // * 0x1: do not respond an error. User to define the behavior,
+                                            //        e.g., respond a success with an implementation-dependent value to
+                                            //        the initiator.
         uint32_t sewe : 1;                  // Supress the (bus) error on an illegal write access caught by the entry
-                                            //• 0x0: the response by ERR_CFG.rwe
-                                            //• 0x1: do not respond an error. User to define the behavior,
-                                            //       e.g., respond a success if response is needed
-        uint32_t sexe : 1;                  //       upress the (bus) error on an illegal instruction fetch
-                                            //       caught by the entry
-                                            // 0x0: the response by ERR_CFG.rxe
-                                            // 0x1: do not respond an error. User to define the behavior,
-                                            //      e.g., respond a success with an implementation-dependent
-                                            //      value to the requestor.
+                                            // * 0x0: respond an error if ERR_CFG.rs is 0x0.
+                                            // * 0x1: do not respond an error. User to define the behavior,
+                                            //        e.g., respond a success if response is needed
+        uint32_t sexe : 1;                  // Suppress the (bus) error on an illegal instruction fetch caught by the
+                                            // entry:
+                                            // * 0x0: the response by ERR_CFG.rxe
+                                            // * 0x1: do not respond an error. User to define the behavior,
+                                            //        e.g., respond a success with an implementation-dependent value to
+                                            //        the requestor.
         uint32_t rsv  : 21;                 // Must be zero on write, reserved for future
     };
     uint32_t raw;
@@ -636,9 +639,16 @@ typedef union {
         implementation_t implementation;
         hwcfg0_t         hwcfg0;
         hwcfg1_t         hwcfg1;
-        hwcfg2_t         hwcfg2;
+        union {
+            hwcfg2_t     hwcfg2;
+            uint32_t     reserved12;
+        };
+        union {
+            hwcfg3_t     hwcfg3;
+            uint32_t     reserved13;
+        };
+        uint32_t         reserved0[5];
         entryoffset_t    entryoffset;
-        uint32_t         reserved0[6];
         #if (IOPMP_STALL_EN)
         mdstall_t        mdstall;
         mdstallh_t       mdstallh;
diff --git a/iopmp_ref_model/src/iopmp_error_capture.c b/iopmp_ref_model/src/iopmp_error_capture.c
index 7b9565e..9735594 100644
--- a/iopmp_ref_model/src/iopmp_error_capture.c
+++ b/iopmp_ref_model/src/iopmp_error_capture.c
@@ -45,7 +45,7 @@ void errorCapture(perm_type_e trans_type, uint8_t error_type, uint16_t rrid, uin
     // If an error was previously logged, handle a subsequent violation
     }
 #if (IOPMP_MFR_EN)
-    else if (!checkRridSv(rrid) && g_reg_file.hwcfg0.mfr_en && (!error_suppress | !intrpt_suppress)) {
+    else if (!checkRridSv(rrid) && g_reg_file.hwcfg2.mfr_en && (!error_suppress | !intrpt_suppress)) {
         // Update violation window
         setRridSv(rrid);
     }
diff --git a/iopmp_ref_model/src/iopmp_reg.c b/iopmp_ref_model/src/iopmp_reg.c
index 17779f8..95e4b2b 100644
--- a/iopmp_ref_model/src/iopmp_reg.c
+++ b/iopmp_ref_model/src/iopmp_reg.c
@@ -51,50 +51,60 @@ int reset_iopmp() {
     g_reg_file.version.specver         = 1;     // Set IOPMP Specification version
     g_reg_file.implementation.raw      = 0;     // Set implementationID
 
+    // Hardware Configuration
+    g_reg_file.hwcfg0.enable           = IOPMP_ENABLE;
+    g_reg_file.hwcfg0.HWCFG2_en        = IMP_HWCFG2;
+    g_reg_file.hwcfg0.HWCFG3_en        = IMP_HWCFG3;
+    g_reg_file.hwcfg0.md_num           = IOPMP_MD_NUM;
+    g_reg_file.hwcfg0.addrh_en         = IOPMP_ADDRH_EN;
+    g_reg_file.hwcfg0.tor_en           = IOPMP_TOR_EN;
+    g_reg_file.hwcfg1.rrid_num         = IOPMP_RRID_NUM;
+    g_reg_file.hwcfg1.entry_num        = IOPMP_ENTRY_NUM;
+#if (IMP_HWCFG2)
+#if (IOPMP_NON_PRIO_EN)
+    g_reg_file.hwcfg2.prio_entry       = IOPMP_PRIO_ENTRY;
+    g_reg_file.hwcfg2.prio_ent_prog    = IOPMP_PRIO_ENT_PROG;
+#else
+    g_reg_file.hwcfg2.prio_entry       = 0;
+    g_reg_file.hwcfg2.prio_ent_prog    = 0;
+#endif
+    g_reg_file.hwcfg2.non_prio_en      = IOPMP_NON_PRIO_EN;
+    g_reg_file.hwcfg2.chk_x            = IOPMP_CHK_X;
+    g_reg_file.hwcfg2.peis             = IOPMP_PEIS;
+    g_reg_file.hwcfg2.pees             = IOPMP_PEES;
+    g_reg_file.hwcfg2.sps_en           = IOPMP_SPS_EN;
+    g_reg_file.hwcfg2.stall_en         = IOPMP_STALL_EN;
+    g_reg_file.hwcfg2.mfr_en           = IOPMP_MFR_EN;
+#else   // IMP_HWCFG2=0
+    g_reg_file.reserved12              = 0;
+#endif
+
+#if (IMP_HWCFG3)
     // Set the MDCFG Format - Based on compiled model
 #ifdef MDCFG_FMT
-    g_reg_file.hwcfg0.mdcfg_fmt        = MDCFG_FMT;
+    g_reg_file.hwcfg3.mdcfg_fmt        = MDCFG_FMT;
 #endif
-
     // Set the SRCMD Format - Based on compiled model
 #ifdef SRCMD_FMT
-    g_reg_file.hwcfg0.srcmd_fmt        = SRCMD_FMT;
+    g_reg_file.hwcfg3.srcmd_fmt        = SRCMD_FMT;
 #endif
-
-    // Hardware Configuration
-    g_reg_file.hwcfg0.tor_en           = IOPMP_TOR_EN;
-    g_reg_file.hwcfg0.sps_en           = IOPMP_SPS_EN;
-    g_reg_file.hwcfg0.user_cfg_en      = 0;
-    g_reg_file.hwcfg0.prient_prog      = IOPMP_PARIENT_PROG;
-    g_reg_file.hwcfg0.rrid_transl_en   = IOPMP_RRID_TRANSL_EN;
-    g_reg_file.hwcfg0.rrid_transl_prog = IOPMP_RRID_TRANSL_PROG;
-    g_reg_file.hwcfg0.chk_x            = IOPMP_CHK_X;
-    g_reg_file.hwcfg0.no_x             = IOPMP_NO_X;
-    g_reg_file.hwcfg0.no_w             = IOPMP_NO_W;
-    g_reg_file.hwcfg0.stall_en         = IOPMP_STALL_EN;
-    g_reg_file.hwcfg0.peis             = IOPMP_PEIS;
-    g_reg_file.hwcfg0.pees             = IOPMP_PEES;
-    g_reg_file.hwcfg0.mfr_en           = IOPMP_MFR_EN;
-
 #if (MDCFG_FMT == 0)
-    g_reg_file.hwcfg0.md_entry_num     = 0;
+    g_reg_file.hwcfg3.md_entry_num     = 0;
 #else
-    g_reg_file.hwcfg0.md_entry_num     = IOPMP_MD_ENTRY_NUM;
+    g_reg_file.hwcfg3.md_entry_num     = IOPMP_MD_ENTRY_NUM;
 #endif
-
-    g_reg_file.hwcfg0.md_num           = IOPMP_MD_NUM;
-    g_reg_file.hwcfg0.addrh_en         = IOPMP_ADDRH_EN;
-
-    g_reg_file.hwcfg0.enable           = IOPMP_ENABLE;
-
-    g_reg_file.hwcfg1.rrid_num         = IOPMP_RRID_NUM;
-    g_reg_file.hwcfg1.entry_num        = IOPMP_ENTRY_NUM;
-
-    g_reg_file.hwcfg2.prio_entry       = IOPMP_PRIO_ENTRY;
+    g_reg_file.hwcfg3.no_x             = IOPMP_NO_X;
+    g_reg_file.hwcfg3.no_w             = IOPMP_NO_W;
+    g_reg_file.hwcfg3.rrid_transl_en   = IOPMP_RRID_TRANSL_EN;
 #if (IOPMP_RRID_TRANSL_EN)
-    g_reg_file.hwcfg2.rrid_transl      = IOPMP_RRID_TRANSL;
+    g_reg_file.hwcfg3.rrid_transl_prog = IOPMP_RRID_TRANSL_PROG;
+    g_reg_file.hwcfg3.rrid_transl      = IOPMP_RRID_TRANSL;
 #else
-    g_reg_file.hwcfg2.rrid_transl      = 0;
+    g_reg_file.hwcfg3.rrid_transl_prog = 0;
+    g_reg_file.hwcfg3.rrid_transl      = 0;
+#endif
+#else   // IMP_HWCFG3=0
+    g_reg_file.reserved13              = 0;
 #endif
 
     g_reg_file.entryoffset.raw         = ENTRY_OFFSET;
@@ -392,7 +402,12 @@ void write_register(uint64_t offset, reg_intf_dw data, uint8_t num_bytes) {
 
   // Initialize temporary registers
     hwcfg0_t         hwcfg0_temp         = { .raw = lwr_data4 };
+#if (IMP_HWCFG2)
     hwcfg2_t         hwcfg2_temp         = { .raw = lwr_data4 };
+#endif
+#if (IMP_HWCFG3)
+    hwcfg3_t         hwcfg3_temp         = { .raw = lwr_data4 };
+#endif
     entrylck_t       entrylck_temp       = { .raw = upr_data4 };
     err_cfg_t        err_cfg_temp        = { .raw = lwr_data4 };
     entry_addr_t     entry_addr_temp     = { .raw = lwr_data4 };
@@ -469,13 +484,6 @@ void write_register(uint64_t offset, reg_intf_dw data, uint8_t num_bytes) {
       return;
 
     case HWCFG0_OFFSET:
-        g_reg_file.hwcfg0.prient_prog      &= ~hwcfg0_temp.prient_prog;
-        g_reg_file.hwcfg0.rrid_transl_prog &= ~hwcfg0_temp.rrid_transl_prog;
-        #if (MDCFG_FMT == 2)
-            if (!g_reg_file.hwcfg0.enable) {
-                g_reg_file.hwcfg0.md_entry_num = hwcfg0_temp.md_entry_num;
-            }
-        #endif
         g_reg_file.hwcfg0.enable           |= hwcfg0_temp.enable;
         break;
 
@@ -483,14 +491,32 @@ void write_register(uint64_t offset, reg_intf_dw data, uint8_t num_bytes) {
         // This register is read only
         return;
 
+#if (IMP_HWCFG2)
     case HWCFG2_OFFSET:
-        if (g_reg_file.hwcfg0.prient_prog) {
-            g_reg_file.hwcfg2.prio_entry  = hwcfg2_temp.prio_entry;
-        }
-        if (g_reg_file.hwcfg0.rrid_transl_en & g_reg_file.hwcfg0.rrid_transl_prog) {
-            g_reg_file.hwcfg2.rrid_transl = hwcfg2_temp.rrid_transl;
-        }
+        #if (IOPMP_NON_PRIO_EN)
+            if (g_reg_file.hwcfg2.prio_ent_prog) {
+                g_reg_file.hwcfg2.prio_entry = hwcfg2_temp.prio_entry;
+            }
+            g_reg_file.hwcfg2.prio_ent_prog &= ~hwcfg2_temp.prio_ent_prog;
+        #endif
+        break;
+#endif
+
+#if (IMP_HWCFG3)
+    case HWCFG3_OFFSET:
+        #if (MDCFG_FMT == 2)
+            if (!g_reg_file.hwcfg0.enable) {
+                g_reg_file.hwcfg3.md_entry_num = hwcfg3_temp.md_entry_num;
+            }
+        #endif
+        #if (IOPMP_RRID_TRANSL_EN)
+            if (g_reg_file.hwcfg3.rrid_transl_prog) {
+                g_reg_file.hwcfg3.rrid_transl = hwcfg3_temp.rrid_transl;
+            }
+            g_reg_file.hwcfg3.rrid_transl_prog &= ~hwcfg3_temp.rrid_transl_prog;
+        #endif
         break;
+#endif
 
     case ENTRYOFFSET_OFFSET:
         // This register is read only
@@ -789,14 +815,14 @@ void write_register(uint64_t offset, reg_intf_dw data, uint8_t num_bytes) {
                     }
 
                     // Interrupt suppression bits are writeable, only if interrupt suppression is supported
-                    if (g_reg_file.hwcfg0.peis){
+                    if (g_reg_file.hwcfg2.peis){
                         iopmp_entries.entry_table[ENTRY_TABLE_INDEX(offset)].entry_cfg.sire = entry_cfg_temp.sire;
                         iopmp_entries.entry_table[ENTRY_TABLE_INDEX(offset)].entry_cfg.siwe = entry_cfg_temp.siwe;
                         iopmp_entries.entry_table[ENTRY_TABLE_INDEX(offset)].entry_cfg.sixe = entry_cfg_temp.sixe;
                     }
 
                     // Error suppression bits are writeable, only if error suppression is supported
-                    if (g_reg_file.hwcfg0.pees) {
+                    if (g_reg_file.hwcfg2.pees) {
                         iopmp_entries.entry_table[ENTRY_TABLE_INDEX(offset)].entry_cfg.sere = entry_cfg_temp.sere;
                         iopmp_entries.entry_table[ENTRY_TABLE_INDEX(offset)].entry_cfg.sewe = entry_cfg_temp.sewe;
                         iopmp_entries.entry_table[ENTRY_TABLE_INDEX(offset)].entry_cfg.sexe = entry_cfg_temp.sexe;
diff --git a/iopmp_ref_model/src/iopmp_rule_analyzer.c b/iopmp_ref_model/src/iopmp_rule_analyzer.c
index 354d230..bc15608 100644
--- a/iopmp_ref_model/src/iopmp_rule_analyzer.c
+++ b/iopmp_ref_model/src/iopmp_rule_analyzer.c
@@ -118,9 +118,9 @@ iopmpMatchStatus_t iopmpCheckPerms(uint16_t rrid, perm_type_e req_perm, entry_cf
 
     // Extract hardware configuration flags
 #if (SRCMD_FMT == 0)
-    bool sps_en = g_reg_file.hwcfg0.sps_en;  // Software privilege separation enable
+    bool sps_en = g_reg_file.hwcfg2.sps_en; // Software privilege separation enable
 #endif
-    bool chk_x  = g_reg_file.hwcfg0.chk_x;    // Execute permission check enable
+    bool chk_x  = g_reg_file.hwcfg2.chk_x;  // Execute permission check enable
 
     // Common permission checks
     bool read_allowed    = false;
@@ -194,9 +194,9 @@ iopmpMatchStatus_t iopmpCheckPerms(uint16_t rrid, perm_type_e req_perm, entry_cf
 iopmpMatchStatus_t iopmpRuleAnalyzer(iopmp_trans_req_t trans_req, uint64_t prev_iopmpaddr, uint64_t iopmpaddr, entry_cfg_t iopmpcfg, uint8_t md, int is_priority) {
     iopmpMatchStatus_t match_status = ENTRY_MATCH;  // Default to full match
     uint64_t start_addr, end_addr;
-    bool no_w  = g_reg_file.hwcfg0.no_w;      // Write restriction
-    bool no_x  = g_reg_file.hwcfg0.no_x;      // Execute restriction
-    bool chk_x = g_reg_file.hwcfg0.chk_x;    // Execute permission check enable
+    bool no_w  = g_reg_file.hwcfg3.no_w;      // Write restriction
+    bool no_x  = g_reg_file.hwcfg3.no_x;      // Execute restriction
+    bool chk_x = g_reg_file.hwcfg2.chk_x;     // Execute permission check enable
     if ((no_w && (trans_req.perm == WRITE_ACCESS)) || (no_x && (trans_req.perm == INSTR_FETCH) && chk_x)) { return ENTRY_NOTMATCH; }
 
     // Set up the address range; if range setup fails, return not matched
diff --git a/iopmp_ref_model/src/iopmp_validate.c b/iopmp_ref_model/src/iopmp_validate.c
index 8863576..bdfc337 100644
--- a/iopmp_ref_model/src/iopmp_validate.c
+++ b/iopmp_ref_model/src/iopmp_validate.c
@@ -33,7 +33,7 @@ void iopmp_validate_access(iopmp_trans_req_t *trans_req, iopmp_trans_rsp_t* iopm
     iopmp_trans_rsp->user         = 0;
     iopmp_trans_rsp->status       = IOPMP_ERROR;
     #if (IOPMP_RRID_TRANSL_EN)
-        iopmp_trans_rsp->rrid_transl = g_reg_file.hwcfg2.rrid_transl;
+        iopmp_trans_rsp->rrid_transl = g_reg_file.hwcfg3.rrid_transl;
     #endif
 
     // Check to block invalid combination
@@ -125,15 +125,19 @@ void iopmp_validate_access(iopmp_trans_req_t *trans_req, iopmp_trans_rsp_t* iopm
             lwr_entry = (cur_md == 0) ? 0 : g_reg_file.mdcfg[cur_md - 1].t;
             upr_entry = g_reg_file.mdcfg[cur_md].t;
         #else
-            lwr_entry = cur_md * (g_reg_file.hwcfg0.md_entry_num + 1);
-            upr_entry = ((cur_md + 1) * (g_reg_file.hwcfg0.md_entry_num + 1));
+            lwr_entry = cur_md * (g_reg_file.hwcfg3.md_entry_num + 1);
+            upr_entry = ((cur_md + 1) * (g_reg_file.hwcfg3.md_entry_num + 1));
         #endif
 
         for (int cur_entry = lwr_entry; cur_entry <= upr_entry; cur_entry++) {
             uint64_t prev_addr     = (cur_entry == 0) ? 0 : CONCAT32(iopmp_entries.entry_table[cur_entry - 1].entry_addrh.addrh, iopmp_entries.entry_table[cur_entry - 1].entry_addr.addr);
             uint64_t curr_addr     = CONCAT32(iopmp_entries.entry_table[cur_entry].entry_addrh.addrh, iopmp_entries.entry_table[cur_entry].entry_addr.addr);
             entry_cfg_t entry_cfg  = iopmp_entries.entry_table[cur_entry].entry_cfg;
-            bool is_priority_entry = (cur_entry < g_reg_file.hwcfg2.prio_entry);
+            #if (IOPMP_NON_PRIO_EN)
+                bool is_priority_entry = (cur_entry < g_reg_file.hwcfg2.prio_entry);
+            #else
+                bool is_priority_entry = true;
+            #endif
 
             // Analyze entry for match
             iopmpMatchStatus = iopmpRuleAnalyzer(*trans_req, prev_addr, curr_addr, entry_cfg, cur_md, is_priority_entry);
