diff --git a/.devcontainer/Dockerfile b/.devcontainer/Dockerfile
index 324011d153..184aca6c44 100644
--- a/.devcontainer/Dockerfile
+++ b/.devcontainer/Dockerfile
@@ -8,6 +8,7 @@ WORKDIR /workspace
 # please keep pkgs sorted
 RUN apt-get update && \
   apt-get install -y --no-install-recommends --fix-missing \
+  bash-completion \
   build-essential \
   clang-format \
   clang-tidy \
@@ -29,6 +30,7 @@ RUN apt-get update && \
   libz3-dev \
   minisat \
   parallel \
+  ssh \
   sudo \
   zlib1g-dev \
   \
diff --git a/bin/.container-tag b/bin/.container-tag
index c74e8a041a..2037dfa6c4 100644
--- a/bin/.container-tag
+++ b/bin/.container-tag
@@ -1 +1 @@
-0.35
+0.36
