diff --git a/Gemfile.lock b/Gemfile.lock
index 9b8ec3cb68..87116ffdb8 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -18,7 +18,7 @@ PATH
 PATH
   remote: tools/ruby-gems/udb-gen
   specs:
-    udb-gen (0.1.4)
+    udb-gen (0.1.5)
       asciidoctor
       asciidoctor-diagram
       asciidoctor-pdf
@@ -39,7 +39,7 @@ PATH
 PATH
   remote: tools/ruby-gems/udb
   specs:
-    udb (0.1.6)
+    udb (0.1.7)
       activesupport
       asciidoctor
       awesome_print
diff --git a/cfgs/mc100-32-full-example.yaml b/cfgs/mc100-32-full-example.yaml
index 5e702ac099..449f27b86c 100644
--- a/cfgs/mc100-32-full-example.yaml
+++ b/cfgs/mc100-32-full-example.yaml
@@ -55,7 +55,6 @@ params:
   MUTABLE_MISA_M: false
   TIME_CSR_IMPLEMENTED: false
   NUM_PMP_ENTRIES: 0
-  PMP_GRANULARITY: 2
   COUNTINHIBIT_EN:
     [
       false,
diff --git a/spec/std/isa/param/PMP_GRANULARITY.yaml b/spec/std/isa/param/PMP_GRANULARITY.yaml
index d4991f872f..9723c9d438 100644
--- a/spec/std/isa/param/PMP_GRANULARITY.yaml
+++ b/spec/std/isa/param/PMP_GRANULARITY.yaml
@@ -21,5 +21,6 @@ schema:
   minimum: 2
   maximum: 66
 definedBy:
-  extension:
-    name: Sm
+  param:
+    name: NUM_PMP_ENTRIES
+    greaterThan: 0
diff --git a/tools/ruby-gems/udb-gen/Gemfile.lock b/tools/ruby-gems/udb-gen/Gemfile.lock
index 0fb2115406..3896bbe94c 100644
--- a/tools/ruby-gems/udb-gen/Gemfile.lock
+++ b/tools/ruby-gems/udb-gen/Gemfile.lock
@@ -17,7 +17,7 @@ PATH
 PATH
   remote: ../udb
   specs:
-    udb (0.1.6)
+    udb (0.1.7)
       activesupport
       asciidoctor
       awesome_print
@@ -41,7 +41,7 @@ PATH
 PATH
   remote: .
   specs:
-    udb-gen (0.1.4)
+    udb-gen (0.1.5)
       asciidoctor
       asciidoctor-diagram
       asciidoctor-pdf
diff --git a/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb b/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb
index 66abcb7e87..c2a55cce39 100644
--- a/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb
+++ b/tools/ruby-gems/udb-gen/lib/udb-gen/version.rb
@@ -5,5 +5,5 @@
 # frozen_string_literal: true
 
 module UdbGen
-  def self.version = "0.1.4"
+  def self.version = "0.1.5"
 end
diff --git a/tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.6.rbi b/tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.7.rbi
similarity index 100%
rename from tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.6.rbi
rename to tools/ruby-gems/udb-gen/sorbet/rbi/gems/udb@0.1.7.rbi
diff --git a/tools/ruby-gems/udb/Gemfile.lock b/tools/ruby-gems/udb/Gemfile.lock
index 044cca16a2..4fbdeea810 100644
--- a/tools/ruby-gems/udb/Gemfile.lock
+++ b/tools/ruby-gems/udb/Gemfile.lock
@@ -17,7 +17,7 @@ PATH
 PATH
   remote: .
   specs:
-    udb (0.1.6)
+    udb (0.1.7)
       activesupport
       asciidoctor
       awesome_print
diff --git a/tools/ruby-gems/udb/lib/udb/version.rb b/tools/ruby-gems/udb/lib/udb/version.rb
index 5a37006c58..e7d11317ac 100644
--- a/tools/ruby-gems/udb/lib/udb/version.rb
+++ b/tools/ruby-gems/udb/lib/udb/version.rb
@@ -5,5 +5,5 @@
 # frozen_string_literal: true
 
 module Udb
-  def self.version = "0.1.6"
+  def self.version = "0.1.7"
 end
