diff --git a/iopmp_ref_model/src/iopmp_reg.c b/iopmp_ref_model/src/iopmp_reg.c
index 107a1f1..844455f 100644
--- a/iopmp_ref_model/src/iopmp_reg.c
+++ b/iopmp_ref_model/src/iopmp_reg.c
@@ -651,30 +651,28 @@ void write_register(iopmp_dev_t *iopmp, uint64_t offset, reg_intf_dw data, uint8
 
         // MDCFG(m) is locked for m < MDCFGLCK.f
         if (mdcfg_idx >= iopmp->reg_file.mdcfglck.f) {
-            if (mdcfg_temp.t < iopmp->reg_file.hwcfg1.entry_num) {
-                iopmp->reg_file.mdcfg[mdcfg_idx].t = mdcfg_temp.t;
-                iopmp->reg_file.mdcfg[mdcfg_idx].rsv = 0;
-
-            #if (MDCFG_TABLE_IMPROPER_SETTING_BEHAVIOR == 0)
-                /*
-                 * MDCFG table must be monotonically incremental. Some reference
-                 * behaviors for an improper setting are given in the specification,
-                 * e.g., "correct the values to make the table have a proper setting".
-                 * The reference model authomatically fixes it if current MDCFG table
-                 * violates the monotonically incremental rule. Programmer can check
-                 * this register after programming done.
-                 *
-                 * The MDCFG look up table is implemented in the following way:
-                 * - For any m >= 1, if (MDCFG(m).t < MDCFG(m-1).t):
-                 *                       MDCFG(m).t = MDCFG(m-1).t
-                 */
-                for (int m = 1; m < iopmp->reg_file.hwcfg0.md_num; m++) {
-                    if (iopmp->reg_file.mdcfg[m].t < iopmp->reg_file.mdcfg[m - 1].t) {
-                        iopmp->reg_file.mdcfg[m].t = iopmp->reg_file.mdcfg[m - 1].t;
-                    }
+            iopmp->reg_file.mdcfg[mdcfg_idx].t = mdcfg_temp.t;
+            iopmp->reg_file.mdcfg[mdcfg_idx].rsv = 0;
+
+        #if (MDCFG_TABLE_IMPROPER_SETTING_BEHAVIOR == 0)
+            /*
+             * MDCFG table must be monotonically incremental. Some reference
+             * behaviors for an improper setting are given in the specification,
+             * e.g., "correct the values to make the table have a proper setting".
+             * The reference model authomatically fixes it if current MDCFG table
+             * violates the monotonically incremental rule. Programmer can check
+             * this register after programming done.
+             *
+             * The MDCFG look up table is implemented in the following way:
+             * - For any m >= 1, if (MDCFG(m).t < MDCFG(m-1).t):
+             *                       MDCFG(m).t = MDCFG(m-1).t
+             */
+            for (int m = 1; m < iopmp->reg_file.hwcfg0.md_num; m++) {
+                if (iopmp->reg_file.mdcfg[m].t < iopmp->reg_file.mdcfg[m - 1].t) {
+                    iopmp->reg_file.mdcfg[m].t = iopmp->reg_file.mdcfg[m - 1].t;
                 }
-            #endif
             }
+        #endif
         }
     }
 
diff --git a/iopmp_ref_model/src/iopmp_validate.c b/iopmp_ref_model/src/iopmp_validate.c
index 2117a39..bfbc89c 100644
--- a/iopmp_ref_model/src/iopmp_validate.c
+++ b/iopmp_ref_model/src/iopmp_validate.c
@@ -199,6 +199,10 @@ void iopmp_validate_access(iopmp_dev_t *iopmp, iopmp_trans_req_t *trans_req, iop
         }
 
         for (int cur_entry = lwr_entry; cur_entry < upr_entry; cur_entry++) {
+            /* Any entry with index >= HWCFG1.entry_num is not available */
+            if (cur_entry >= iopmp->reg_file.hwcfg1.entry_num)
+                break;
+
             /* Assign necessary input information */
             rule_analyzer_i.prev_iopmpaddr =
                 (cur_entry == 0) ? 0 : CONCAT32(iopmp->iopmp_entries.entry_table[cur_entry - 1].entry_addrh.addrh,
