diff --git a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb.tests/src/org/xtext/example/udb/tests/UdbParsingTest.xtend b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb.tests/src/org/xtext/example/udb/tests/UdbParsingTest.xtend
index ee8fbdac39..885c3a1aee 100644
--- a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb.tests/src/org/xtext/example/udb/tests/UdbParsingTest.xtend
+++ b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb.tests/src/org/xtext/example/udb/tests/UdbParsingTest.xtend
@@ -36,7 +36,7 @@ class UdbParsingTest {
 					location: 2-1
 					description: "See vxrm."
 					type: RW-RH
-					alias: vxrm.VALUE[1:0]
+					alias: "vxrm.VALUE[1:0]"
 					sw_write(csr_value): "|
 					  CSR[vxrm].VALUE = csr_value.VXRM;
 					  return csr_value.VXRM;"
@@ -45,7 +45,7 @@ class UdbParsingTest {
 					location: 0
 					description: "See vxsat."
 					type: RW-RH
-					alias: vxsat.VALUE[0]
+					alias: "vxsat.VALUE[0]"
 					sw_write(csr_value): "|
 					  CSR[vxsat].VALUE = csr_value.VXSAT;
 					  return csr_value.VXSAT;"
@@ -57,30 +57,32 @@ class UdbParsingTest {
 
 
 		// check basic inputs
-		var schema = result.getSchema().getSchema();
+		var csr = result.getCsr();
+		
+		var schema = csr.getSchema().getSchema();
 		Assertions.assertEquals("csr_schema.json#", schema as String);
-		var k = result.getKind().getKind().getType();
+		var k = csr.getCsrKind().getKind();
 		Assertions.assertEquals("csr", k as String);
-		var n = result.getCsrName().getName().getType();
+		var n = csr.getCsrName().getName();
 		Assertions.assertEquals("vcsr", n as String);
-		var ln = result.getLongName().getLongName();
+		var ln = csr.getLongName().getLongName();
 		Assertions.assertEquals("Vector Control and Status Register", ln);
-		var add = result.getAddress().getAddress().getValue();
+		var add = csr.getAddress().getAddress().getValue();
 		Assertions.assertEquals(0x00F, add);
-		var writ = result.getWritable().isWritable();
+		var writ = csr.getWritable().isWritable();
 		Assertions.assertTrue(writ);
-		var priv = result.getPrivmode().getPrivMode().getType();
+		var priv = csr.getPrivmode().getPrivMode();
 		Assertions.assertEquals("U", priv);
-		var len = result.getLength().getLength().getParmType().getParmName();
+		var len = csr.getLength().getLength().getParmType().getParmName();
 		Assertions.assertEquals("MXLEN", len as String);
-		var desc = result.getDescription().getDescription();
+		var desc = csr.getDescription().getDescription();
 		Assertions.assertEquals("Contains aliases to vxrm and vxsat CSRs", desc);
-		var def = result.getDefinedBy().getExtensionName();
+		var def = csr.getDefinedBy().getExtensionName();
 		Assertions.assertEquals("V", def);
 
 		// test fields
-		var vxrm = result.getFields().getFields().get(0);
-		var vxsat = result.getFields().getFields().get(1);
+		var vxrm = csr.getCsrFields().getFields.get(0);
+		var vxsat = csr.getCsrFields().getFields().get(1);
 		Assertions.assertEquals("VXRM", vxrm.getName());
 		Assertions.assertEquals("VXSAT", vxsat.getName());
 
@@ -116,12 +118,67 @@ class UdbParsingTest {
 		Assertions.assertEquals("RW-RH", sattype);
 
 		// testing alias
-		var vxalias = vxrm.getAlias().getAlias().getAlias().get(0);
-		var satalias = vxsat.getAlias().getAlias().getAlias().get(0);
+		var vxalias = vxrm.getAlias().getAliasName().getName();
+		var satalias = vxsat.getAlias().getAliasName().getName();
 		Assertions.assertEquals("vxrm.VALUE[1:0]", vxalias);
 		Assertions.assertEquals("vxsat.VALUE[0]", satalias);
 
 
 	}
+	
+	@Test
+	def void parsesValidExtension() {
+		val result = parseHelper.parse('''
+		$schema: "ext_schema.json#"
+		kind: extension
+		name: V
+		type: unprivileged
+		long_name: "Vector Operations"
+		versions:
+		  - version: "1.0.0"
+		    state: ratified
+		    ratification_date: 2021-11
+		description: "|
+		  General support for data-parallel execution."
+		requirements: '
+		  extension:
+		    allOf:
+		      - name: Zve64d
+		      - name: Zvl128b'
+		''')
+		Assertions.assertNotNull(result)
+		val errors = result.eResource.errors
+		Assertions.assertTrue(errors.isEmpty, '''Unexpected errors: «errors.join(", ")»''')
+		
+		var ext = result.getExt();
+		
+		var schema = ext.getSchema().getSchema();
+		Assertions.assertEquals("ext_schema.json#", schema as String);
+		var kind = ext.getExtKind().getKind();
+		Assertions.assertEquals("extension", kind);
+		var name = ext.getExtName().getName();
+		Assertions.assertEquals("V", name);
+		var type = ext.getType().getPerms();
+		Assertions.assertEquals("unprivileged", type);
+		var longname = ext.getLongName().getLongName();
+		Assertions.assertEquals("Vector Operations", longname);
+		var description = ext.getDescription().getDescription();
+		Assertions.assertEquals("|\n  General support for data-parallel execution.", description);
+		
+		// version testing
+		var version = ext.getExtVersions().getElements().get(0);
+		Assertions.assertEquals("1.0.0", version.getVersion());
+		Assertions.assertEquals("ratified", version.getVersionState().getState());
+		Assertions.assertEquals("2021-11", version.getRatificationDate().getDate());
+		
+		// requirements testing
+		var reqs = ext.getRequirements();
+		System.out.println(reqs.getRequirements());
+		
+		
+		
+	}
+	
+	
 
 }
diff --git a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/GenerateUdb.mwe2 b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/GenerateUdb.mwe2
index 2a089f6e8e..95a3fe5cb2 100644
--- a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/GenerateUdb.mwe2
+++ b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/GenerateUdb.mwe2
@@ -55,6 +55,12 @@ Workflow {
 			junitSupport = {
 				junitVersion = "5"
 			}
+			parserGenerator = {
+				options = {
+					backtrack = true
+					skipUnusedRules = false
+				}
+			}
 		}
 	}
 }
diff --git a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/Udb.xtext b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/Udb.xtext
index 5ee4834bdf..3fe4ffe199 100644
--- a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/Udb.xtext
+++ b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/Udb.xtext
@@ -5,138 +5,442 @@ generate udb "http://www.xtext.org/example/udb/Udb"
 import "http://www.eclipse.org/emf/2002/Ecore" as ecore
 
 Model:
-	schema = Schema
-	kind = Kind &
-	csrName = Name &
-	base = Base? &
+	  csr  = CsrModel
+	| inst = InstModel
+	| ext  = ExtModel
+;
+
+
+CsrModel:
+	schema = Schema // must be "csr_schema.json#"
+	csrKind = CsrKind
+	csrName = CsrName &
 	longName = LongName &
 	description = Description &
-	definedBy = DefinedBy &
-	address = Address &
-	indirectAddress = IndirectAddress? &
-	indirectSlot = IndirectSlot? &
-	writable = Writable &
-	privmode = PrivMode &
-	virtualAddress = VirtualAddress? &
-	length = Length &
-	requires = Requires? &
-	fields = Fields? &
-	swRead = SwRead? &
+	definedBy = CsrDefinedBy &
+	address = CsrAddress &
+	indirectAddress = CsrIndirectAddress? &
+	indirectSlot = CsrIndirectSlot? &
+	writable = CsrWritable &
+	privmode = CsrPrivMode &
+	virtualAddress = CsrVirtualAddress? &
+	length = CsrLength &
+	requires = CsrRequires? &
+	csrFields = CsrFields? &
+	swRead = CsrSwRead? &
 	source = Source? &
 	certNormativeRules = CertNormativeRules?
 ;
 
-// This is so we can directly copy+paste from a UDB .yaml file
-Schema: "$schema" ":" schema=STRING;
-
-/* Required fields for the CSR schema */
-Kind: 'kind' ':' kind=KindType ;
-Name: 'name' ':' name=NameType ;
-LongName: 'long_name' ':' longName=STRING ;
-Address: 'address' ':' address=Hex ;
-Writable: 'writable' ':' writable=Boolean ;
-PrivMode:'priv_mode' ':' privMode=PrivType ;
-Length: 'length' ':' length=LengthType ;
-Description: 'description' ':' description=STRING ;
-DefinedBy:'definedBy' ':'extensionName=STRING ;
 
-/* Optional fields for the CSR schema */
-Base: 'base' ':' base=INT ;
-IndirectAddress:'indirect_address' ':' indirectAddress=Hex;
-IndirectSlot:'indirect_slot' ':' indirectSlot=INT ; /* TODO: validator: min:1 max:6 */
+InstModel: 
+	schema = Schema // must be "inst_schema.json#"
+	kind = InstKind
+	instName = InstName &
+	longName = LongName &
+	description = Description &
+	definedBy = InstDefinedBy &
+	hints = Hints? &
+	base = Base? &
+	access = Access &
+	accessDetail = AccessDetail? &
+	operation = Operation? &
+	operationAst = OperationAst? &
+	sail = Sail? &
+	certNormativeRules = CertNormativeRules? &
+	assembly = Assembly &
+	dataIndependentTiming = DataIndependentTiming? &
+	pseudoInstructions = PseudoInstructions? &
+	source = Source? &
+	format = Format? &
+	encoding = Encoding?
+;
 
 
-VirtualAddress:'virtual_address' ':' virtualAddress=Hex ;
-CertNormativeRules: 'cert_normative_rules' ':' certNormativeRules=CNRType ;
 
-// implement in validator?
-Requires:'requires' ':' requires=STRING ;
-SwRead:'sw_read()' ':' swRead=STRING ;
-Source:'$source' ':' source=STRING ;
+ExtModel: 
+	schema = Schema // must be "ext_schema.json#"
+	extKind = ExtKind
+	extName = ExtName &
+	longName = LongName &
+	description = Description &
+	rviJiraIssue = ExtRviJiraIssue? &
+	company = ExtCompany? &
+	docLicense = DocLicense? &
+	type = ExtType &
+	requirements = Requirements? &
+	extVersions = ExtVersions &
+	source = Source? &
+	certNormativeRules = CertNormativeRules?
+;
 
 
-/* CSR Schema Types */
-KindType: type=('csr'|'extension');
-NameType: type=(CSR_NAME);
-PrivType: type=('M'|'S'|'U'|'VS'|'D');
 
-LengthType: intType = IntType | parmType = ParmType;
-IntType: intVal=INT; // must be either 32 or 64
-ParmType: parmName=('MXLEN'|'SXLEN'|'VSXLEN'|'XLEN');
+/* 
+ * Common stuff
+ */
+Schema: "$schema" ":" schema=STRING;
+LongName: 'long_name' ":" longName=STRING;
+Description: 'description' ':' description=STRING; 
+Requirements: "requirements" ":" requirements=Condition;
+Source:'$source' ':' source=STRING;
+
+UdbId: 'id' ':' id=STRING;
+Url: 'url' ':' url=STRING;
+Email: 'email' ':' email=STRING;
+
+// see "cert_normative_rules" in schema_defs.json
+// TODO: not tested
+CertNormativeRules: 
+	'cert_normative_rules' ':' 
+	INDENT	
+		rules += CNRuleArrayElement+
+	DEDENT
+;
+CNRuleArrayElement:
+	'-' id = UdbId // TODO: makes id the first element (do we want this?)
+	INDENT (
+		('name' ':' name = STRING) &
+		docLinks = DocLinks &
+		description = Description
+	) DEDENT
+;
+DocLinks:	
+	"doc_links" ":"
+	INDENT
+		links += DocLinksArrayElement+
+	DEDENT
+;
+DocLinksArrayElement:
+	'-' link = STRING	// TODO: this must follow the URI format
+;
 
-CNRType: '[' id=STRING ',' name=STRING ',' docLinks=StringArray ','
-			 description=STRING ']';
-CTPType: '[' id=STRING ',' name=STRING ',' description=STRING ','
-			 normativeRules=StringArray ',' list=STRING?']';
+// see "license" in schema_defs.json 
+DocLicense:
+	'doc_license' ':'
+	INDENT (
+		('name' ':' name=STRING) & 
+		id=UdbId? &
+		url=Url? &
+		textUrl=Url?
+	) DEDENT
+;
 
 
 
-/* Grammar for fields */
-Fields:
-	{Fields} 'fields' ':'
+/*
+ * CSR stuff (see csr_schema.json)
+ */
+CsrKind: 'kind' ':' kind="csr";
+CsrName: 'name' ':' name=ID;
+CsrAddress: 'address' ':' address=HEX;
+CsrWritable: 'writable' ':' writable=Boolean;
+CsrDefinedBy: 'definedBy' ':' extensionName=STRING;
+CsrIndirectAddress:'indirect_address' ':' indirectAddress=HEX;
+CsrIndirectSlot:'indirect_slot' ':' indirectSlot=INT ; 
+CsrVirtualAddress:'virtual_address' ':' virtualAddress=HEX;
+CsrSwRead:'sw_read()' ':' swRead=IDL;
+CsrPrivMode:'priv_mode' ':' privMode=ID;
+CsrLength: 'length' ':' length=CsrLengthType;
+CsrLengthType: intType = CsrIntType | parmType = CsrParmType;
+CsrIntType: intVal=INT; 
+CsrParmType: parmName=('MXLEN'|'SXLEN'|'VSXLEN'|'XLEN');
+
+// extension that must be implemented for this CSR to exist
+CsrRequires:'requires' ':' requires=STRING; 
+
+// grammar for csr_field
+CsrFields:
+	{CsrFields} 'fields' ':'
 	INDENT
-    	fields += FieldDef+
+    	fields += CsrFieldDef+
     DEDENT
 ;
 
-FieldDef:
-	name=FIELD_NAME ':'
-	INDENT
-		('name' ':' fieldName = STRING ';')? &
+CsrFieldDef:
+	name=ID ':'
+	INDENT (
+		displayName = CsrFieldDisplayName? &
 		longName = LongName? &
-		location = LocationType &
-		resetValue = ResetValueType &
-		swWriteFunc = SWWriteFunc? &
-		legalFunc = LegalFunc? &
-		base = Base? &
+		location = CsrFieldLocationType &
+		resetValue = CsrFieldResetValueType &
+		swWriteFunc = CsrFieldSWWriteFunc? &
+		legalFunc = CsrFieldLegalFunc? &
 		description = Description &
-		type = TypeType &
-		alias = Alias? &
-		definedBy = DefinedBy? &
-		affectedBy = AffectedBy? &
+		type = CsrFieldTypeType &
+		alias = CsrFieldAlias? &
+		definedBy = CsrDefinedBy? &
+		affectedBy = CsrFieldAffectedBy? &
 		certNormativeRules = CertNormativeRules?
+	) DEDENT
+;
+
+CsrFieldDisplayName: 'name' ':' fieldName=STRING ;
+
+CsrFieldLocationType: staticLoc = CsrStaticLocation | funcLoc = CsrVarLocation;
+CsrStaticLocation: 'location' ':' locValue=CsrFieldLocation;
+CsrVarLocation:
+	locationRV32=LocationRV32 &
+	locationRV64=LocationRV64
+;
+LocationRV32: "location_rv32" ":" locationRV32=CsrFieldLocation;
+LocationRV64: "location_rv64" ":" locationRV64=CsrFieldLocation;
+CsrFieldLocation: value=INT | range=CsrFieldLocationRange;
+CsrFieldLocationRange: upper=INT '-' lower=INT;
+
+CsrFieldResetValueType: value=CsrFieldResetValue | function=CsrFieldResetValueFunc;
+CsrFieldResetValue: "reset_value" ":" resetValue=CsrFieldResetValueValue;
+CsrFieldResetValueValue: value=INT | undefinedLegal=UndefinedLegal;
+CsrFieldResetValueFunc: "reset_value()" ":" resetValueFunc=IDL;
+
+CsrFieldSWWriteFunc: "sw_write(csr_value)" ":" swWriteFunc=IDL;
+CsrFieldLegalFunc: "legal?(csr_value)" ":" legalFunc=IDL;
+
+CsrFieldTypeType: typeVal=CsrFieldType | typeFunc=CsrFieldTypeFunc;
+CsrFieldType: "type" ":" perms=("RO"|"RO-H"|"RW"|"RW-R"|"RW-H"|"RW-RH");
+CsrFieldTypeFunc: "type()" ":" idl=IDL;
+
+CsrFieldAlias: "alias" ":" (aliasName=CsrFieldAliasName | aliasArray=CsrFieldAliasArray);
+CsrFieldAliasName: name=STRING;
+CsrFieldAliasArray:
+	"items" ":"
+	INDENT
+		aliases += CsrAliasArrayElement+
 	DEDENT
 ;
+CsrAliasArrayElement:
+	"-" alias = CsrFieldAliasName 
+;
+
+// TODO: cross-referencing
+CsrFieldAffectedBy: "affectedBy" ":" affectedBy=CsrAffectedByType; 
+CsrAffectedByType: affectedByName=ID | affectedByArray=Array;
+
 
-FieldDisplayName: 'name' ':' fieldName=STRING ;
 
 
+/*
+ * Instruction stuff (see inst_schema.json)
+ * NOTE: implementation started in PRIDE1099-INST branch
+ */
+InstKind: 'kind' ':' kind="instruction";
+InstName: 'name' ':' name=ID;
+InstDefinedBy: 'definedBy' ':' dbname=STRING;
+Base: 'base' ':' base=INT; // 32 or 64
+Access:
+	'access' ':'
+	INDENT
+	('m' ':' m=AccessType)? &
+	's' ':' s=AccessType &
+	'u' ':' u=AccessType &
+	'vs' ':' vs=AccessType &
+	'vu' ':' vu=AccessType
+	DEDENT
+	;
+AccessType:
+	accessLevel=('always'|'sometimes'|'never')
+;
+Assembly: 'assembly' ':' assembly=STRING;
+/* Optional fields for the INST schema */
+OperationAst: 'operation_ast' ':' operationAst=STRING;
+Hints:
+	'hints' ':'
+	INDENT
+		hints += HintElement+
+	DEDENT
+;
+HintElement: '-' '{' hint = STRING '}'; // must match "^inst/.+\\.yaml#.*$"
+
+AccessDetail: 'access_detail' ':' accessDetail=STRING;
+Operation: 'operation()' ':' operation=STRING; // string for now, IDL later TODO
+Sail: 'sail()' ':' sail=STRING; // string for now, IDL later TODO
+DataIndependentTiming: 'data_independent_timing' ':' dataIndependentTiming=Boolean;
+PseudoInstructions: 'pseudoinstructions' ':' pseudoInstructions=Array; // type TODO
+Format: 'format' ':' 
+	INDENT
+		'$inherits' ':' inherits=InheritType
+		'opcodes' ':' opcodes=OpcodeType 
+	DEDENT
+;
+InheritType:
+	INDENT
+		('-' reference += STRING)+ //regex must fit yaml path
+	DEDENT
+;
+
+OpcodeType:
+	INDENT
+		opcode += OpcodeEntry+
+	DEDENT
+;
+OpcodeEntry:
+	OpcodeField
+	| OpcodeInherits
+;
 
-/* fields for Fields */
-StaticLocation: 'location' ':' locValue=FieldLocation ;
-LocationRV32: "location_rv32" ":" locationRV32=FieldLocation ;
-LocationRV64: "location_rv64" ":" locationRV64=FieldLocation ;
+OpcodeField: // ex) funct7 
+	name=ID ':' // must fit regex ('a'..'z') ('a'..'z'|'0'..'9')*, example of opcode name: "function7"
+	INDENT
+		(opcodeProperties += OpcodeProps)+
+	DEDENT
+	
+;
 
+OpcodeProps: 
+	'display_name' ':' displayName=ID 
+	| 'value' ':' value=JsonIntLike
+;
+JsonIntLike:
+	intVal=INT
+	| dec=DEC
+	| bin=BIN
+	| hex=HEX
+;
+OpcodeInherits: 
+	'opcode' ':' '{' '$inherits' ':' inheritsAddress=STRING '}'// must fit regex 'inst_opcode/' INST_OPCODE_FILE '.yaml#/data'
+;
 
-ResetValue: "reset_value" ":" resetValue=ResetValueValue ;
-ResetValueFunc: "reset_value()" ":" resetValueFunc=IDL ;
-SWWriteFunc: "sw_write(csr_value)" ":" swWriteFunc=IDL ;
-LegalFunc: "legal?(csr_value)" ":" legalFunc=IDL ;
-Type: "type" ":" perms=("RO"|"RO-H"|"RW"|"RW-R"|"RW-H"|"RW-RH") ;
-TypeFunc: "type()" ":" idl=IDL ;
-Alias: "alias" ":" alias=AliasType ;
-AffectedBy: "affectedBy" ":" affectedBy=AffectedByType ;
 
-/* Types for Fields */
+Encoding: 
+	'encoding' ':'
+	(
+		RvPairEncoding
+		| OldEncoding
+	)
+;
 
-LocationType: staticLoc = StaticLocation | funcLoc = VarLocation;
-VarLocation:
-	locationRV32=LocationRV32
-	locationRV64=LocationRV64
+OldEncoding:
+	{OldEncoding} 
+	INDENT
+	    match=EncodingMatch
+	    variables=EncodingVariables?
+  	DEDENT
+;
+EncodingMatch:
+	'match' ':' pattern=STRING
+;
+
+EncodingVariables:
+	'variables' ':'
+	  INDENT
+	    vars += EncodingVar+
+	  DEDENT
+;
+
+EncodingVar: // TODO Check if its ok name must always be first
+    '-' 'name' ':' name=ID
+    (INDENT 
+    	'location' ':' location=BitRange
+    DEDENT)
+
+;
+
+BitRange:
+  hi=INT '-' lo=INT
+;
+RvPairEncoding:
+	{RvPairEncoding}
+	INDENT
+	'RV32' ':' rv32=OldEncoding
+	'RV64' ':' rv64=OldEncoding
+	DEDENT
+; 
+
+
+/* 
+ * Extension stuff (see ext_schema.json) 
+ */
+ExtKind: 'kind' ':' kind="extension";
+ExtName: 'name' ':' name=ID;
+ExtType: 'type' ':' perms=('unprivileged' | 'privileged');
+ExtRviJiraIssue: 'rvi_jira_issue' ':' issue=STRING;
+ExtCompany: 
+	'company' ':' 
+	INDENT (
+		('name' ':' name=STRING) &
+		 url=Url?
+	) DEDENT 
 ;
-FieldLocation: value=INT | range=Range;
-Range: upper=INT '-' lower=INT;
-ResetValueType: value=ResetValue | function=ResetValueFunc;
-ResetValueValue: value=INT | undefinedLegal=UndefinedLegal;
-TypeType: typeVal=Type | typeFunc=TypeFunc;
-AliasType: alias+=(CSR_FIELD | CSR_FIELD_BITS)+;
 
-// TODO: we probably will want a cross ref thing for this
-AffectedByType: affectedByName=STRING | affectedByArray=StringArray;
+// versions grammar (see ext_schema.json)
+ExtVersions:
+	"versions" ":" 
+	INDENT
+		elements += ExtVersionArrayElement+
+	DEDENT
+;
+
+ExtVersionArrayElement:
+	'-' "version" ":" version=STRING // TODO: this has to be first, is that okay?
+	INDENT (
+		versionState = ExtVersionState &
+		repositories = ExtVersionRepositories? &
+		ratificationDate = ExtVersionRatificationDate? &
+		versionChanges = ExtVersionVersionChanges? &
+		url = Url? &
+		versionContributors = ExtVersionContributors? &
+		requirements = Requirements?
+	) DEDENT
+;
+
+ExtVersionState: 
+	"state" ":" 
+	state=("development" | "frozen" | "public-review" | 
+		   "ratification-ready" | "ratified" | "nonstandard-released")
+;
+
+ExtVersionRepositories:
+	"repositories" ":"
+	INDENT
+		repositories += ExtVersionRepoArrayElement+
+	DEDENT
+;
+ExtVersionRepoArrayElement: // TODO: this is ordered, that okay?
+	'-' url=Url
+	(INDENT
+		"branch" ":" branch=STRING
+	DEDENT)?
+;
+
+ExtVersionRatificationDate:
+	"ratification_date" ":"
+	date = (
+	  YEAR_MONTH			// when ratification date is known
+	| "unknown"				// when ratification date is unknown
+	| "null")				// when version isn't ratified			
+;
+
+ExtVersionVersionChanges:
+	"changes" ":"
+	INDENT
+		changes += ExtVersionChangesArrayElement+
+	DEDENT
+;
+ExtVersionChangesArrayElement:
+	'-' change=STRING
+;
+
+ExtVersionContributors:
+	"contributors" ":"
+	INDENT
+		contributors += ExtVersionContributorsArrayElement+ 
+	DEDENT
+;
+ExtVersionContributorsArrayElement:
+	'-' name=ExtVersionContributorName // TODO: is it okay that name must be the first element?
+	(INDENT (
+		("company" ":" company=STRING)? & // company contributor worked for or 'Individual'
+		email=Email?
+	) DEDENT)?
+;
+ExtVersionContributorName:
+	"name" ":" firstName = ID lastName = ID?
+;
 
 
 
 /* Conditions */
+Condition: cond=STRING; // placeholder
 //Condition: yamlCond=YAMLCondition | idlCond=IDLCondition;
 
 //YAMLCondition:
@@ -146,22 +450,22 @@ AffectedByType: affectedByName=STRING | affectedByArray=StringArray;
 //	| xlenCond = XLenCondition
 //;
 
-IDLCondition:
-	"idl()" ":" idl=IDL
-	reason?=Reason
-;
-
-ExtensionCondition:
-	STRING /* TODO */
-;
-
-ParamCondition:
-	STRING /* TODO */
-;
+//IDLCondition:
+//	"idl()" ":" idl=IDL
+//	reason?=Reason
+//;
+//
+//ExtensionCondition:
+//	STRING /* TODO */
+//;
+//
+//ParamCondition:
+//	STRING /* TODO */
+//;
 
-XLenCondition: xlen=INT; /* can be 32 or 64 (TODO: implement in validator) */
+//XLenCondition: xlen=INT; /* can be 32 or 64 (TODO: implement in validator) */
 
-Reason: "reason" ":" reasonStr=STRING;
+//Reason: "reason" ":" reasonStr=STRING;
 
 
 
@@ -172,38 +476,47 @@ IDL: idl=STRING;
 
 
 
-/* Keywords */
+/*
+ * Keywords
+ */
 UndefinedLegal: "UNDEFINED_LEGAL";
 Boolean returns ecore::EBoolean: "true" | "false";
 
-/* Helpers*/
-StringArray: '[' names+=STRING (',' names+=STRING)+ ']';
 
-/* Terminals */
-terminal CSR_NAME:
-	('a'..'z')('a'..'z' | '0'..'9' | '_' | '.')+
-;
-terminal FIELD_NAME:
-	('a'..'z' | 'A'..'Z')+
-;
-terminal CSR_FIELD:
-	('a'..'z')('a'..'z' | '0'..'9' | '_' | '.')+
-	('.')('A'..'Z' | '0'..'9')+
-;
-terminal CSR_FIELD_BITS:
-	('a'..'z')('a'..'z' | '0'..'9' | '_' | '.')+
-	'.'('A'..'Z' | '0'..'9')+
-	'[' ('0'..'9')+(':'('0'..'9')+)? ']'
-;
+
+/*
+ * Helpers
+ */
+Array: '[' names+=ID (',' names+=ID)+ ']';
+
 
 
-Hex: value=HEX_VALUE;
-terminal HEX_VALUE returns ecore::EInt:
+/*
+ * Hex
+ */
+
+terminal HEX returns ecore::EInt:
 	('0x' | '0X')
 	('0'..'9' | 'A'..'F' | 'a'..'f')
 	(('0'..'9' | 'A'..'F' | 'a'..'f' | '_')*('0'..'9' | 'A'..'F' | 'a'..'f'))?
 ;
 
+terminal BIN: 
+	'0b' ('0'|'1')+
+;
+
+
+
+terminal DEC: '0' ('0'..'9')+;  // "0111", "00010", etc.
+/*
+ * Terminals
+ */
+terminal YEAR_MONTH:
+	(('20')('0'..'9')('0'..'9')'-'    
+	  ('0'('1'..'9') | '1'('0'..'2')))
+;
+
+// For whitespace awareness
 terminal INDENT: 'synthetic:INDENT';
 terminal DEDENT: 'synthetic:DEDENT';
 
@@ -212,3 +525,8 @@ terminal SL_COMMENT: '#' !('\n'|'\r')*;
 
 @Override
 terminal WS: ('\r'|'\n'|'\t'|' ')+;
+
+@Override
+terminal ID: '^'?('a'..'z'|'A'..'Z'|'_') ('a'..'z'|'A'..'Z'|'_'|'0'..'9'|'.')*;
+
+
diff --git a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/validation/UdbValidator.java b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/validation/UdbValidator.java
index 4f49d4ddc4..d553992ef3 100644
--- a/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/validation/UdbValidator.java
+++ b/tools/eclipse/dev/org.xtext.udb.parent/org.xtext.udb/src/org/xtext/example/udb/validation/UdbValidator.java
@@ -5,14 +5,29 @@
 
 import org.eclipse.emf.common.util.EList;
 import org.eclipse.xtext.validation.Check;
-import org.xtext.example.udb.parser.antlr.UdbParser;
-import org.xtext.example.udb.udb.FieldDef;
-import org.xtext.example.udb.udb.IntType;
-import org.xtext.example.udb.udb.LengthType;
-import org.xtext.example.udb.udb.Model;
-import org.xtext.example.udb.udb.ParmType;
 import org.xtext.example.udb.udb.UdbPackage;
 
+import org.xtext.example.udb.udb.Url;
+import org.xtext.example.udb.udb.Email;
+
+import org.xtext.example.udb.udb.CsrModel;
+import org.xtext.example.udb.udb.CsrFieldDef;
+import org.xtext.example.udb.udb.CsrIntType;
+import org.xtext.example.udb.udb.CsrLengthType;
+import org.xtext.example.udb.udb.CsrName;
+import org.xtext.example.udb.udb.CsrFieldAliasName;
+import org.xtext.example.udb.udb.CsrAffectedByType;
+import org.xtext.example.udb.udb.CsrFieldAffectedBy;
+
+import org.xtext.example.udb.udb.ExtModel;
+import org.xtext.example.udb.udb.ExtVersionArrayElement;
+import org.xtext.example.udb.udb.ExtVersionRepoArrayElement;
+import org.xtext.example.udb.udb.ExtVersionContributorsArrayElement;
+
+import org.xtext.example.udb.udb.InstModel;
+
+
+
 
 /**
  * This class contains custom validation rules.
@@ -20,62 +35,71 @@
  * See https://www.eclipse.org/Xtext/documentation/303_runtime_concepts.html#validation
  */
 public class UdbValidator extends AbstractUdbValidator {
+	
+	// Regex's found in schema_defs.json
+	String rviVersionRegex = "^[0-9]+(\\.[0-9]+(\\.[0-9]+(-pre)?)?)?$";
+	String csrFieldRegex = "^[a-z][a-z0-9_.]+\\.[A-Z0-9]+$";
+    String csrFieldBitsRegex = "^[a-z][a-z0-9_.]+\\.[A-Z0-9]+\\[[0-9]+(:[0-9]+)?\\]$";
+    String csrNameRegex = "^[a-z][a-z0-9_.]+$";
+    String csrAffectedByRegex = "^(RV64)|([A-WY]|(Z[a-z]+)|(S[a-z]+))$";
+    String extensionNameRegex = "^(([A-WY])|([SXZ][a-z0-9]+))$";
+    String instructionNameRegex= "[a-z0-9.]+";
+    String instructionHintsRegex="^\\\\$ref:\\\\s*inst/.+\\\\.yaml#.*$";
+    // Extra regex's for validation
+    String urlRegex = "^https?:\\/\\/[^\\s/$.?#].[^\\s]*$";
+    String emailRegex = "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$";
 
+	
+    
+    /*
+     * CSR Validation -- rules found in csr_schema.json
+     */
 	@Check
-	public void checkAddress(Model m) {
-		int address = m.getAddress() != null ? m.getAddress().getAddress().getValue() : null;
+	public void checkAddress(CsrModel csr) {
+		// Address must be between 0 and 12 bits
+		int address = csr.getAddress() != null ? csr.getAddress().getAddress() : null;
 
 		if(address < 0 || address > 4096) {
-			error("Address must be between 0 and 12 bits.", UdbPackage.Literals.MODEL__ADDRESS);
+			error("Address must be between 0 and 12 bits.", UdbPackage.Literals.CSR_MODEL__ADDRESS);
 		}
 	}
 
 	@Check
-	public void checkVirtualAddressValue(Model m) {
+	public void checkVirtualAddressValue(CsrModel csr) {
 		/* Ensure virtual address is in required range 0-4095 */
-		int vaddress = m.getVirtualAddress() != null ?
-					   m.getVirtualAddress().getVirtualAddress().getValue() : null;
+		int vaddress = csr.getVirtualAddress() != null ?
+					   csr.getVirtualAddress().getVirtualAddress() : null;
 
 		if (vaddress < 0 || vaddress > 4095) {
-			error("Virtual address must be between 0 and 12 bits.", UdbPackage.Literals.MODEL__VIRTUAL_ADDRESS);
+			error("Virtual address must be between 0 and 12 bits.", UdbPackage.Literals.CSR_MODEL__VIRTUAL_ADDRESS);
 		}
 	}
 
 	@Check
-	public void checkIndirectAddressValue(Model m) {
+	public void checkIndirectAddressValue(CsrModel csr) {
 		/* Ensure indirect address is in required range */
-		int iaddress = m.getIndirectAddress() != null ?
-					   m.getIndirectAddress().getIndirectAddress().getValue() : null;
+		int iaddress = csr.getIndirectAddress() != null ?
+					   csr.getIndirectAddress().getIndirectAddress() : null;
 
 		if(iaddress < 0 || iaddress > (2^64)) {
-			error("Indirect address must be between 0 and 64 bits.", UdbPackage.Literals.MODEL__INDIRECT_ADDRESS);
+			error("Indirect address must be between 0 and 64 bits.", UdbPackage.Literals.CSR_MODEL__INDIRECT_ADDRESS);
 		}
 	}
 
 	@Check
-	public void checkBaseValue(Model m) {
-		/* Ensure base value is either 32 or 64 */
-		int base = m.getBase() != null ? m.getBase().getBase() : null;
-
-		if (base != 32 && base != 64) {
-			error("Base must have value of 32 or 64.", UdbPackage.Literals.MODEL__BASE);
-		}
-	}
-
-	@Check
-	public void checkLengthValue(Model m) {
+	public void checkLengthValue(CsrModel csr) {
 		// Containment reference is always instantiated
-		LengthType length_t = m.getLength().getLength();
+		CsrLengthType length_t = csr.getLength().getLength();
 		if (length_t == null) {
 			error("length should not be null",
-					UdbPackage.Literals.MODEL__LENGTH);
+					UdbPackage.Literals.CSR_MODEL__LENGTH);
 		}
 		// If length is an integer, value is either 32 or 64
-		if (length_t instanceof IntType) {
-			Integer length = ((IntType) m.getLength().getLength()).getIntVal();
+		if (length_t instanceof CsrIntType) {
+			Integer length = ((CsrIntType) csr.getLength().getLength()).getIntVal();
 			if (length != 32 && length != 64) {
 				error("length if specified as integer, should be 32 or 64",
-						UdbPackage.Literals.MODEL__LENGTH);
+						UdbPackage.Literals.CSR_MODEL__LENGTH);
 			}
 		}
 //		if (length != null && length != "MXLEN" && length != "SXLEN" && length != "VSXLEN" && length != "XLEN") {
@@ -87,25 +111,273 @@ public void checkLengthValue(Model m) {
 	}
 
 	@Check
-	public void checkIndirectSlotValue(Model m) {
+	public void checkIndirectSlotValue(CsrModel csr) {
 		/* Ensure indirect_slot is between 1 and 6 */
-		int slot = m.getIndirectSlot() != null ? m.getIndirectSlot().getIndirectSlot() : null;
+		int slot = csr.getIndirectSlot() != null ? csr.getIndirectSlot().getIndirectSlot() : null;
 
 		if (slot < 1 || slot > 6) {
-			error("Indirect slot value must be between 1 and 6.", UdbPackage.Literals.MODEL__INDIRECT_SLOT);
+			error("Indirect slot value must be between 1 and 6.", UdbPackage.Literals.CSR_MODEL__INDIRECT_SLOT);
+		}
+	}
+	
+	@Check
+	public void checkCsrPrivMode(CsrModel csr) {
+		String mode = csr.getPrivmode() != null ? csr.getPrivmode().getPrivMode() : null;
+		java.util.Set<String> ALLOWED_TYPES = java.util.Set.of("M","S","U","VS","D");
+		
+		if (mode != null && !ALLOWED_TYPES.contains(mode)) {
+		    error(
+		        "Invalid privilege mode: " + mode,
+		        UdbPackage.Literals.CSR_MODEL__PRIVMODE
+		    );
 		}
 	}
 
 	@Check
-	public void checkVirtualAddress(Model m) {
-		String mode = m.getPrivmode() != null ? m.getPrivmode().getPrivMode().getType() : null;
+	public void checkVirtualAddress(CsrModel csr) {
+		String mode = csr.getPrivmode() != null ? csr.getPrivmode().getPrivMode() : null;
 
 		if (mode.equals("VS")) {
-			if (m.getVirtualAddress() == null) {
-				error("VS mode requires a virtual address.", UdbPackage.Literals.MODEL__PRIVMODE);
+			if (csr.getVirtualAddress() == null) {
+				error("VS mode requires a virtual address.", UdbPackage.Literals.CSR_MODEL__PRIVMODE);
+			}
+		}
+
+	}
+	
+	@Check
+	public void checkCsrName(CsrName name) {
+	    String value = name.getName();
+	    if (!value.matches(csrNameRegex)) {
+	        error("Invalid csr name", 
+	              UdbPackage.Literals.CSR_NAME__NAME);
+	    }
+	}
+	
+	@Check
+	public void checkCsrFieldName(CsrFieldDef field) {
+		String value = field.getName();
+		if (!value.matches("^[a-zA-Z].*$")) {
+			error("Invalid field name",
+					UdbPackage.Literals.CSR_FIELD_DEF__NAME);
+		}
+	}
+	
+	@Check
+	public void checkCsrFieldAlias(CsrFieldAliasName alias) {
+		String value = alias.getName();
+
+	    if (!value.matches(csrFieldRegex) &&
+	        !value.matches(csrFieldBitsRegex)) {
+
+	        error(
+	            "Alias must match CSR_FIELD or CSR_FIELD_BITS format",
+	            UdbPackage.Literals.CSR_FIELD_ALIAS_NAME__NAME
+	        );
+	    }
+	}
+	
+	@Check
+	public void checkCsrFieldAffectedBy(CsrFieldAffectedBy affectedBy) {
+		CsrAffectedByType type = affectedBy.getAffectedBy(); 
+		String name = type.getAffectedByName();
+		
+		if (name != null) {
+			if (!name.matches(csrAffectedByRegex)) {
+				error("Invalid extension name", // TODO: is this an okay error message?
+						UdbPackage.Literals.CSR_FIELD_AFFECTED_BY__AFFECTED_BY);
+			}
+		} else {
+			EList<String> exts = type.getAffectedByArray().getNames();
+			for (int i = 0; i < exts.size(); i++) {
+				String ext = exts.get(i);
+				if (!ext.matches(csrAffectedByRegex)) {
+					error("Invalid extension name",
+							UdbPackage.Literals.CSR_FIELD_AFFECTED_BY__AFFECTED_BY);
+				}
 			}
 		}
+	}
+	
+	
+	
+	/*
+	 * Instruction Validation -- rules found in inst_schema.json
+	 */
+	@Check
+	public void checkBaseValue(InstModel inst) {
+		/* Ensure base value is either 32 or 64 */
+		int base = inst.getBase() != null ?inst.getBase().getBase() : null;
 
+		if (base != 32 && base != 64) {
+			error("Base must have value of 32 or 64.", UdbPackage.Literals.INST_MODEL__BASE);
+		}
+	}
+
+
+	@Check
+	public void checkInstructionModel(InstModel inst) {
+		String schema = inst.getSchema().getSchema();
+		if (!schema.equals("inst_schema.json#")) {
+			error("Schema incompatible with kind", inst.getSchema(), UdbPackage.Literals.SCHEMA__SCHEMA);
+		}
+		
+		// check that the extension name is valid
+	    String name = inst.getInstName().getName();
+	    if (!name.matches(instructionNameRegex)) {
+	        error("Invalid instruction name", inst.getInstName(),
+	              UdbPackage.Literals.INST_NAME__NAME);
+	    }
+	}
+	
+	/*
+	 *  Check access detail field is present when at least one mode 
+	 *  within access field is sometimes
+	 */
+	@Check
+	public void checkInstAccessDetail(InstModel inst) {
+		String accessdetail = inst.getAccessDetail() != null ? inst.getAccessDetail().getAccessDetail(): null;
+
+		String m = inst.getAccess().getM() != null ? inst.getAccess().getM().getAccessLevel(): null;
+		String s = inst.getAccess().getS() != null ? inst.getAccess().getS().getAccessLevel(): null;
+		String u = inst.getAccess().getU() != null ? inst.getAccess().getU().getAccessLevel(): null;
+		String vs = inst.getAccess().getVs() != null ? inst.getAccess().getVs().getAccessLevel(): null;
+		String vu = inst.getAccess().getVu() != null ? inst.getAccess().getVu().getAccessLevel(): null;
+
+		boolean sometimes =
+			    "sometimes".equals(m) ||
+			    "sometimes".equals(s) ||
+			    "sometimes".equals(u) ||
+			    "sometimes".equals(vs) ||
+			    "sometimes".equals(vu);
+		if (sometimes == true && (accessdetail == null || accessdetail == "")) {
+			error("Must provide access_detail field when at least one access type is sometimes", UdbPackage.Literals.INST_MODEL__ACCESS);
+		}
+	}
+	
+
+	// Check that hints (strings within array) are of format ^inst/.+\\.yaml#.*$
+	public void checkHints(InstModel inst) {
+		EList<org.xtext.example.udb.udb.HintElement> hints = inst.getHints() != null ? inst.getHints().getHints(): null;
+		
+		error("not entering loop", UdbPackage.Literals.INST_MODEL__HINTS);
+		for (org.xtext.example.udb.udb.HintElement hint : hints) {
+			String hintString = hint.getHint();
+			error("hint: " + hintString, UdbPackage.Literals.INST_MODEL__HINTS);
+			if (!(hintString.matches(instructionHintsRegex))) {
+				error("hints must be of format ^\\$ref:\\s*inst/.+\\.yaml#.*$", UdbPackage.Literals.INST_MODEL__HINTS);
+			}
+		}
+	}
+	
+	
+	
+	
+	
+	/*
+	 * Extension Validation -- rules found in ext_schema.json
+	 */
+	@Check
+	public void checkExtensionModel(ExtModel ext) {
+		String schema = ext.getSchema().getSchema();
+		if (!schema.equals("ext_schema.json#")) {
+			error("Schema incompatible with kind", ext.getSchema(), UdbPackage.Literals.SCHEMA__SCHEMA);
+		}
+		
+		// check that the extension name is valid
+	    String name = ext.getExtName().getName();
+	    if (!name.matches(extensionNameRegex)) {
+	        error("Invalid extension name", ext.getExtName(),
+	              UdbPackage.Literals.EXT_NAME__NAME);
+	    }
+	    
+	    EList<ExtVersionArrayElement> versions = ext.getExtVersions().getElements();
+	    
+	    for (int i=0; i < versions.size(); i++) {
+	    	ExtVersionArrayElement elem = versions.get(i);
+	    	checkExtVersionArrayElement(elem);
+	    }
+	    
+	    // Check that url's follow uri format
+	    Url companyUrl = ext.getCompany() != null ? ext.getCompany().getUrl() : null;
+	    checkUrlFormat(companyUrl);
+	    Url docLicenseUrl = ext.getDocLicense() != null ? ext.getDocLicense().getUrl() : null;
+	    checkUrlFormat(docLicenseUrl);
+	    Url docLicenseTextUrl = ext.getDocLicense() != null ? ext.getDocLicense().getTextUrl() : null;
+	    checkUrlFormat(docLicenseTextUrl);
+	}
+	
+	public void checkExtVersionArrayElement(ExtVersionArrayElement elem) {
+		// Validate elements in the versions array
+		
+		// check that the string representation of the version is valid
+    	String versionString = elem.getVersion();
+    	if (!versionString.matches(rviVersionRegex)) {
+			error("Invalid version", elem, UdbPackage.Literals.EXT_VERSION_ARRAY_ELEMENT__VERSION);
+		}
+    	
+    	// if state is ratified, a ratification date must be given
+    	String versionState = elem.getVersionState().getState();
+		if (versionState.equals("ratified")) {
+			if (elem.getRatificationDate() == null) {
+				error("Ratified states require a ratification date.", elem,
+						UdbPackage.Literals.EXT_VERSION_ARRAY_ELEMENT__VERSION_STATE);
+			}
+		}
+		
+		// check that url follows uri format
+		Url versionUrl = elem.getUrl();
+		checkUrlFormat(versionUrl);
+		
+		// check that url's in repositories follow uri format
+		EList<ExtVersionRepoArrayElement> repos = 
+				elem.getRepositories() != null ? elem.getRepositories().getRepositories() : null;
+		if (!(repos == null)) {
+			for (int j = 0; j < repos.size(); j++) {
+				ExtVersionRepoArrayElement r = repos.get(j);
+				
+				Url repoUrl = r.getUrl();
+				checkUrlFormat(repoUrl);
+			}
+		}
+		
+		// check that contributor emails follow email format
+		EList<ExtVersionContributorsArrayElement> contributors = 
+				elem.getVersionContributors() != null ? elem.getVersionContributors().getContributors() : null;
+		if (!(contributors == null)) {
+			for (int j = 0; j < contributors.size(); j++) {
+				ExtVersionContributorsArrayElement c = contributors.get(j);
+				
+				Email contributorEmail = c.getEmail();
+				checkEmailFormat(contributorEmail);
+			}
+		}	
+	}
+	
+	
+	
+	/*
+	 *  Helper functions to validate more general fields (e.g. url, email, etc.)
+	 */
+	public void checkUrlFormat(Url url) {
+		// Check that URLs follow the URI format
+		if (!(url == null)) {
+			String urlString = url.getUrl();
+			if (!urlString.matches(urlRegex)) {
+				error("URL not in URI format", url, UdbPackage.Literals.URL__URL);
+			}
+		}
+
+	}
+	
+	public void checkEmailFormat(Email email) {
+		// Check that emails follow email format
+		if (!(email == null)) {
+			String emailString = email.getEmail();
+			if (!emailString.matches(emailRegex)) {
+				error("Email not in formatted correctly", email, UdbPackage.Literals.EMAIL__EMAIL);
+			}
+		}
 	}
 
 }
diff --git a/tools/eclipse/run/testIdl/src/csr/vcsr.udb b/tools/eclipse/run/testIdl/src/csr/vcsr.udb
new file mode 100644
index 0000000000..f6be19577b
--- /dev/null
+++ b/tools/eclipse/run/testIdl/src/csr/vcsr.udb
@@ -0,0 +1,36 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# yaml-language-server: $schema=../../schemas/csr_schema.json
+
+$schema: "csr_schema.json#"
+kind: csr
+name: vcsr
+long_name: "Vector Control and Status Register"
+address: 0x00F
+writable: true
+priv_mode: U
+length: MXLEN
+description: "Contains aliases to vxrm and vxsat CSRs"
+definedBy:'
+  extension:
+    name: Zvl32b'
+fields:
+  VXRM:
+    location: 2-1
+    description: "See vxrm."
+    type: RW-RH
+    alias: "vxrm.VALUE[1:0]"
+    sw_write(csr_value): '|
+      CSR[vxrm].VALUE = csr_value.VXRM;
+      return csr_value.VXRM;'
+    reset_value: UNDEFINED_LEGAL
+  VXSAT:
+    location: 0
+    description: "See vxsat."
+    type: RW-RH
+    alias: "vxsat.VALUE[0]"
+    sw_write(csr_value): "|
+      CSR[vxsat].VALUE = csr_value.VXSAT;
+      return csr_value.VXSAT;"
+    reset_value: UNDEFINED_LEGAL
\ No newline at end of file
diff --git a/tools/eclipse/run/testIdl/src/csr/vsstatus.udb b/tools/eclipse/run/testIdl/src/csr/vsstatus.udb
new file mode 100644
index 0000000000..c41000e6bb
--- /dev/null
+++ b/tools/eclipse/run/testIdl/src/csr/vsstatus.udb
@@ -0,0 +1,232 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# yaml-language-server: $schema=../../../schemas/csr_schema.json
+
+$schema: "csr_schema.json#"
+kind: csr
+name: vsstatus
+long_name: "Virtual Supervisor Status"
+address: 0x200
+writable: true
+virtual_address: 0x100
+priv_mode: VS
+length: VSXLEN
+description: "|
+  The vsstatus register tracks and controls the hart's current operating state.
+
+  It is VS-mode's version of `sstatus`, and substitutes for it when in VS-mode
+  (_i.e._, in VS-mode CSR address 0x100 is `vsstatus`, not `sstatus`).
+
+  Unlike the relationship between `sstatus` and `mstatus`, none of the bits in `vsstatus` are
+  aliases of another field."
+definedBy:'
+  extension:
+    name: H'
+fields:
+  SD:
+    location_rv64: 63
+    location_rv32: 31
+    description: "|
+      *State Dirty*
+
+      Read-only bit that summarizes whether any of the
+      `vsstatus.FS`, <% if ext?(:V) %> `vsstatus.VS`, <% end %> or `vsstatus.XS`
+      fields signal the presence of some dirty state
+      (_i.e._, any of them hold the value `11`).
+
+      This bit is _not_ an alias of `mstatus.SD` since
+      it only reflects the state visible to VS-mode
+      (_e.g._, `status.FS` does not affect `vsstatus.SD`)."
+    type: RO-H
+    reset_value: UNDEFINED_LEGAL
+    affectedBy: [F, D, V]
+  UXL:
+    location: 33-32
+    definedBy:'
+      xlen: 64'
+    description: '|
+      *VU-mode XLEN*
+
+      Sets the effective XLEN for VU-mode (0 = 32-bit, 1 = 64-bit, 2 = 128-bit).
+
+      [when,"VUXLEN == 32"]
+      Since the hart only supports VUXLEN==32, this is hardwired to 0.
+
+      [when,"VUXLEN == 64"]
+      Since the hart only supports VUXLEN==64, this is hardwired to 1.'
+
+    type(): "|
+      return ($array_size(VUXLEN) > 1) ? CsrFieldType::RW : CsrFieldType::RO;"
+    reset_value(): "|
+      return ($array_size(VUXLEN) > 1) ? UNDEFINED_LEGAL : VUXLEN[0];"
+  MXR:
+    alias: "mstatus.MXR"
+    location: 19
+    description: "|
+      *Make eXecutable Readable*
+
+      Makes it possible to read executable pages when loading from effective VU-mode or VS-mode
+      (normally, executable pages are not readable).
+
+      * When 1, load in effective VU-mode or VS-mode from pages marked readable *or executable*
+        are allowed as long as the page is marked readable (or executable and `status.MXR` is set) in the
+        G-stage translation.
+      * When 0, load in effective VU-mode or VS-mode from pages marked executable raise a
+        Page Fault exception (unless `sstatus.MXR` is also set, in which case the above applies).
+
+      'vsstatus.MXR' affects all loads that operate under effective VU-mode or VS-mode:
+
+      * All loads in VU-mode
+      * All loads in VS-mode
+      * Loads in M-mode when `mstatus.MPRV` == 1, `mstatus.MPV` == 1, and `mstatus.MPP`[1] == 0
+      * Loads generated by one of the `hlv.*` instructions."
+
+    type: RW
+    definedBy: '
+      extension:
+        name: S'
+    reset_value: UNDEFINED_LEGAL
+  SUM:
+    location: 18
+    description: "|
+      *permit Supervisor Memory Access*
+
+      Allows VS-mode to read user pages.
+
+      Applies to the following loads and stores:
+
+      * All loads and stores in VS-mode.
+      * All loads and stores in M-mode when `mstatus.MPRV` == 1, `mstatus.MPP` == 1, and `mstatus.MPV` == 1
+      * Loads and stores generated by one of the `hlv.*`, `hlvx.*`, or `hsv.*` instructions.
+
+      When `vsstatus.SUM` is 0, the loads and stores from the above categories cause an
+      `Illegal Instruction` exception if they access a user page during VS-level translation.
+      Otherwise, a load or store from the above categories is permitted to access a user page
+      during VS-level translation."
+
+    type: RW
+    definedBy:'
+      extension:
+        name: S'
+    reset_value: UNDEFINED_LEGAL
+  XS:
+    alias: "mstatus.XS"
+    location: 16-15
+    description: "|
+      *Custom (X) extension context Status*
+
+      Summarizes the current state of any custom extension state.
+      Either 0 - Off, 1 - Initial, 2 - Clean, 3 - Dirty.
+      Since there are no custom extensions, this field is read-only 0."
+    type: RO
+    reset_value: 0
+  FS:
+    location: 14-13
+    description: "|
+      *Floating point context status*
+
+      When 0, floating point instructions (from F and D extensions) in VS-mode or VU-mode are disabled,
+      and cause ILLEGAL INSTRUCTION exceptions.
+      Floating point instructions in all modes, including VS-mode and VU-mode,
+      are similarly disabled when `mstatus.FS` is clear.
+
+      When a floating point register, or the `fcsr` register is written in VS-mode or VU-mode,
+      `vsstatus.FS` is written with the value 3.
+
+      Values 1 and 2 are valid write values for software, but are not interpreted by hardware
+      other than to possibly enable a previously-disabled floating point unit."
+    type: RW-H
+    definedBy:'
+      extension:
+        name: F'
+    reset_value: UNDEFINED_LEGAL
+  VS:
+    location: 10-9
+    description: "|
+      *Vector context status*
+
+      When 0, vector instructions (from the V extension) are disabled, and cause ILLEGAL INSTRUCTION exceptions.
+      When a vector register or vector CSR is written, VS obtains the value 3.
+      Values 1 and 2 are valid write values for software, but are not interpreted by hardware
+      other than to possibly enable a previously-disabled vector unit."
+    type: RW-H
+    reset_value: UNDEFINED_LEGAL
+    definedBy: '
+      extension:
+        name: Zvl32b'
+  SPP:
+    location: 8
+    description: "|
+      *VS-mode Previous Privilege*
+
+      Written with the prior nominal privilege level (_i.e._, 0 for VU-mode and 1 for VS-mode)
+      when entering VS-mode from an exception/interrupt.
+      Can also be written by software without immediate side-effect.
+
+      On a return from an exception from VS-mode, the machine will enter the nominal privilege level
+      stored in `vsstatus.SPP`."
+    type: RW-H
+    reset_value: UNDEFINED_LEGAL
+  UBE:
+    location: 6
+    description: "|
+      *VU-mode Big Endian*
+
+      Controls the endianness of VU-mode (0 = little, 1 = big).
+
+      [when,\"VU_MODE_ENDIANNESS == 'little'\"]
+      Since the CPU does not support big endian, this is hardwired to 0.
+
+      [when,\"VU_MODE_ENDIANNESS == 'big'\"]
+      Since the CPU does not support big endian, this is hardwired to 1."
+    type(): '|
+      return (VU_MODE_ENDIANNESS == "dynamic") ? CsrFieldType::RW : CsrFieldType::RO;'
+    definedBy:'
+      extension:
+        name: S'
+    reset_value(): '|
+      if (VU_MODE_ENDIANNESS == "little") {
+        # little endian
+        return 0;
+      } else if (VU_MODE_ENDIANNESS == "big") {
+        # big endian
+        return 1;
+      } else {
+        # mutable
+        return UNDEFINED_LEGAL;
+      }'
+  SPIE:
+    location: 5
+    description: "|
+      *VS-mode Previous Interrupt Enable*
+
+      Written by hardware in two cases:
+
+      * Written with prior value of `vsstatus.SIE` when entering VS-mode from an exception/interrupt.
+      * Written with the value 1 when returning from an exception in VS-mode (via the `sret` instruction).
+
+      Can also be written by software without immediate side effect.
+
+      Other than serving as a record of nested traps as described above, `vsstatus.SPIE` does not affect execution."
+    type: RW-H
+    definedBy: '
+      extension:
+        name: S'
+    reset_value: UNDEFINED_LEGAL
+  SIE:
+    location: 1
+    description: "|
+      *VS-mode Interrupt Enable*
+
+      Written by hardware in two cases:
+
+      * Written with the value 0 when entering VS-mode from an exception/interrupt.
+      * Written with the prior value of `vsstatus.SPIE` when returning from an exception in VS-mode (via `sret`).
+
+      Affects execution by:
+
+      * When 0, all VS-mode interrupts are disabled when the current privilege level is VS ((H)S-mode and M-mode interrupts are still enabled).
+      * When 1, VS-mode interrupts that are not otherwise disabled with a field in `vsie` are enabled."
+    type: RW-H
+    reset_value: UNDEFINED_LEGAL
\ No newline at end of file
diff --git a/tools/eclipse/run/testIdl/src/ext/A.udb b/tools/eclipse/run/testIdl/src/ext/A.udb
new file mode 100644
index 0000000000..edf5930428
--- /dev/null
+++ b/tools/eclipse/run/testIdl/src/ext/A.udb
@@ -0,0 +1,78 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# yaml-language-server: $schema=../../../schemas/ext_schema.json
+
+$schema: "ext_schema.json#"
+kind: extension
+name: A
+type: unprivileged
+long_name: "Atomic instructions"
+company:
+  name: "RISC-V International"
+  url: "https://riscv.org"
+versions:
+  - version: "2.1.0"
+    state: ratified
+    ratification_date: 2019-12
+    contributors:
+      - name: Unknown
+        email: "unknown@void.segfault"
+        company: "Unknown"
+    requirements: 'extension:
+        allOf:
+          - name: Zaamo
+            version: "= 1.0.0"
+          - name: Zalrsc
+            version: "= 1.0.0"'
+description: "|
+
+  The atomic-instruction extension, named `A`, contains
+  instructions that atomically read-modify-write memory to support
+  synchronization between multiple RISC-V harts running in the same memory
+  space. The two forms of atomic instruction provided are
+  load-reserved/store-conditional instructions and atomic fetch-and-op
+  memory instructions. Both types of atomic instruction support various
+  memory consistency orderings including unordered, acquire, release, and
+  sequentially consistent semantics. These instructions allow RISC-V to
+  support the RCsc memory consistency model. cite:[Gharachorloo90memoryconsistency]
+
+  [NOTE]
+  ====
+  After much debate, the language community and architecture community
+  appear to have finally settled on release consistency as the standard
+  memory consistency model and so the RISC-V atomic support is built
+  around this model.
+  ====
+
+  The `A` extension comprises instructions provided by the `Zaamo` and `Zalrsc`
+  extensions.
+
+  = Specifying Ordering of Atomic Instructions
+
+  The base RISC-V ISA has a relaxed memory model, with the `FENCE`
+  instruction used to impose additional ordering constraints. The address
+  space is divided by the execution environment into memory and I/O
+  domains, and the `FENCE` instruction provides options to order accesses to
+  one or both of these two address domains.
+
+  To provide more efficient support for release consistency cite:[Gharachorloo90memoryconsistency], each atomic
+  instruction has two bits, _aq_ and _rl_, used to specify additional
+  memory ordering constraints as viewed by other RISC-V harts. The bits
+  order accesses to one of the two address domains, memory or I/O,
+  depending on which address domain the atomic instruction is accessing.
+  No ordering constraint is implied to accesses to the other domain, and a
+  FENCE instruction should be used to order across both domains.
+
+  If both bits are clear, no additional ordering constraints are imposed
+  on the atomic memory operation. If only the _aq_ bit is set, the atomic
+  memory operation is treated as an _acquire_ access, i.e., no following
+  memory operations on this RISC-V hart can be observed to take place
+  before the acquire memory operation. If only the _rl_ bit is set, the
+  atomic memory operation is treated as a _release_ access, i.e., the
+  release memory operation cannot be observed to take place before any
+  earlier memory operations on this RISC-V hart. If both the _aq_ and _rl_
+  bits are set, the atomic memory operation is _sequentially consistent_
+  and cannot be observed to happen before any earlier memory operations or
+  after any later memory operations in the same RISC-V hart and to the
+  same address domain."
\ No newline at end of file
diff --git a/tools/eclipse/run/testIdl/src/ext/B.udb b/tools/eclipse/run/testIdl/src/ext/B.udb
new file mode 100644
index 0000000000..f8d07d4d3b
--- /dev/null
+++ b/tools/eclipse/run/testIdl/src/ext/B.udb
@@ -0,0 +1,40 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# yaml-language-server: $schema=../../../schemas/ext_schema.json
+
+$schema: "ext_schema.json#"
+kind: extension
+name: B
+type: unprivileged
+long_name: "Bit Manipulation"
+company:
+  name: "RISC-V International"
+  url: "https://riscv.org"
+doc_license:
+  name: "Creative Commons Attribution 4.0 International License"
+  url: "https://creativecommons.org/licenses/by/4.0/"
+versions:
+  - version: "1.0.0"
+    state: ratified
+    ratification_date: 2024-04
+    contributors:
+      - name: Ved Shanbhogue
+        email: "ved@rivosinc.com"
+        company: "Rivos, Inc."
+    url: "https://drive.google.com/file/d/1SgLoasaBjs5WboQMaU3wpHkjUwV71UZn/view"
+    requirements: 'extension:
+        allOf:
+          - name: Zba
+            version: "= 1.0.0"
+          - name: Zbb
+            version: "= 1.0.0"
+          - name: Zbs
+            version: "= 1.0.0"'
+description: "|
+  The B standard extension comprises instructions provided by the `Zba`, `Zbb`, and `Zbs` extensions.
+
+  Bit 1 of the `misa` register encodes the presence of the B standard extension. When `misa.B` is 1,
+  the implementation supports the instructions provided by the `Zba`, `Zbb`, and `Zbs` extensions.
+  When `misa.B` is 0, it indicates that the implementation may not support one or more of the
+  `Zba`, `Zbb`, or `Zbs` extensions."
\ No newline at end of file
diff --git a/tools/eclipse/run/testIdl/src/ext/Sha.udb b/tools/eclipse/run/testIdl/src/ext/Sha.udb
new file mode 100644
index 0000000000..12d1f9bf94
--- /dev/null
+++ b/tools/eclipse/run/testIdl/src/ext/Sha.udb
@@ -0,0 +1,69 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# yaml-language-server: $schema=../../../schemas/ext_schema.json
+
+$schema: "ext_schema.json#"
+kind: extension
+name: Sha
+type: privileged
+long_name: "Augmented hypervisor"
+description: "|
+  *Sha* comprises the following extensions:
+
+  ** *H* The hypervisor extension.
+
+  ** *Ssstateen* Supervisor-mode view of the state-enable extension.  The
+     supervisor-mode (`sstateen0-3`) and hypervisor-mode (`hstateen0-3`)
+     state-enable registers must be provided.
+
+  ** *Shcounterenw* For any `hpmcounter` that is not read-only zero, the corresponding bit in `hcounteren` must be writable.
+
+  ** *Shvstvala* `vstval` must be written in all cases described above for `stval`.
+
+  ** *Shtvala* `htval` must be written with the faulting guest physical
+     address in all circumstances permitted by the ISA.
+
+  ** *Shvstvecd* `vstvec.MODE` must be capable of holding the value 0 (Direct).
+    When `vstvec.MODE`=Direct, `vstvec.BASE` must be capable of holding
+    any valid four-byte-aligned address.
+
+  ** *Shvsatpa* All translation modes supported in `satp` must be supported in `vsatp`.
+
+  ** *Shgatpa* For each supported virtual memory scheme SvNN supported in
+    `satp`, the corresponding hgatp SvNNx4 mode must be supported.  The
+    `hgatp` mode Bare must also be supported.
+
+  [NOTE]
+  Sha is a new profile-defined extension that captures the
+  full set of features that are mandated to be supported along with the
+  H extension.  There is no change to the features added by including
+  the hypervisor extension in a profile--the new name is solely to
+  simplify the text of the profiles.  The definition has been added to
+  the RVA22 profile text, where the hypervisor extension was first
+  added, and will be added to the hypervisor section of the combined ISA
+  manual."
+
+versions:
+  - version: "1.0.0"
+    state: ratified
+    ratification_date: 2024-10
+    requirements: '
+      extension:
+        allOf:
+          - name: H
+            version: "= 1.0.0"
+          - name: Ssstateen
+            version: "= 1.0.0"
+          - name: Shcounterenw
+            version: "= 1.0.0"
+          - name: Shvstvala
+            version: "= 1.0.0"
+          - name: Shtvala
+            version: "= 1.0.0"
+          - name: Shvstvecd
+            version: "= 1.0.0"
+          - name: Shvsatpa
+            version: "= 1.0.0"
+          - name: Shgatpa
+            version: "= 1.0.0"'
\ No newline at end of file
diff --git a/tools/eclipse/run/testIdl/src/vcsr.udb b/tools/eclipse/run/testIdl/src/vcsr.udb
deleted file mode 100644
index b7175a3c97..0000000000
--- a/tools/eclipse/run/testIdl/src/vcsr.udb
+++ /dev/null
@@ -1,30 +0,0 @@
-kind: csr
-name: vcsr
-address: 0x99F # validator works
-definedBy: "V"
-long_name: "Vector Control and Status Register"
-description: "Contains aliases to vxrm and vxsat CSRs"
-writable: true
-priv_mode: VS
-virtual_address: 0x00
-length: RNDLEN
-# single line comment works
-fields:
-	VXRM:
-		location: 2-1
-		reset_value: 0
-		description: "See vxrm."
-		sw_write(csr_value): "|
-	      CSR[vxrm].VALUE = csr_value.VXRM;
-	      return csr_value.VXRM;"
-		type: RW-RH
-		alias: vxrm.VALUE[1:0]
-	VXSAT:  # syntax error if vxrm is dedented TODO: better error message
-		location: 0
-		reset_value: 0
-		sw_write(csr_value): "|
-	      CSR[vxsat].VALUE = csr_value.VXSAT
-	      return csr_value.VXSAT;"
-		description: "See vxsat."
-		type: RW-RH
-		alias: vxsat.VALUE[0]
diff --git a/tools/eclipse/run/testIdl/src/vsstatus.udb b/tools/eclipse/run/testIdl/src/vsstatus.udb
deleted file mode 100644
index 14a9541dc4..0000000000
--- a/tools/eclipse/run/testIdl/src/vsstatus.udb
+++ /dev/null
@@ -1,225 +0,0 @@
-# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
-# SPDX-License-Identifier: BSD-3-Clause-Clear
-
-
-kind: csr
-name: vsstatus
-long_name: "Virtual Supervisor Status"
-description: "|
-  The vsstatus register tracks and controls the hart's current operating state.
-
-  It is VS-mode's version of `sstatus`, and substitutes for it when in VS-mode
-  (_i.e._, in VS-mode CSR address 0x100 is `vsstatus`, not `sstatus`).
-
-  Unlike the relationship between `sstatus` and `mstatus`, none of the bits in `vsstatus` are
-  aliases of another field."
-definedBy: "H"
-address: 0x200
-writable: true
-priv_mode: VS
-virtual_address: 0x100
-length: VSXLEN
-fields:
-	SD:
-		location_rv32: 31
-		location_rv64: 63
-		reset_value: UNDEFINED_LEGAL
-		description: "|
-	      *State Dirty*
-
-	      Read-only bit that summarizes whether any of the
-	      `vsstatus.FS`, <% if ext?(:V) %> `vsstatus.VS`, <% end %> or `vsstatus.XS`
-	      fields signal the presence of some dirty state
-	      (_i.e._, any of them hold the value `11`).
-
-	      This bit is _not_ an alias of `mstatus.SD` since
-	      it only reflects the state visible to VS-mode
-	      (_e.g._, `status.FS` does not affect `vsstatus.SD`)."
-		type: RO-H
-		# affectedBy: [F, D, V]; TODO: doesn't work yet
-
-	UXL:
-		location: 33-32
-		reset_value(): "|
-      		return (VUXLEN == 3264) ? UNDEFINED_LEGAL : VUXLEN;"
-		base: 64
-		description: '|
-		      *VU-mode XLEN*
-
-		      Sets the effective XLEN for VU-mode (0 = 32-bit, 1 = 64-bit, 2 = 128-bit).
-
-		      [when,"VUXLEN == 32"]
-		      Since the hart only supports VUXLEN==32, this is hardwired to 0.
-
-		      [when,"VUXLEN == 64"]
-		      Since the hart only supports VUXLEN==64, this is hardwired to 1.'
-
-		type(): "|
-      		return (VUXLEN == 3264) ? CsrFieldType::RW : CsrFieldType::RO;"
-
-	MXR:
-		location: 19
-		reset_value: UNDEFINED_LEGAL
-		description: "|
-	      *Make eXecutable Readable*
-
-	      Makes it possible to read executable pages when loading from effective VU-mode or VS-mode
-	      (normally, executable pages are not readable).
-
-	      * When 1, load in effective VU-mode or VS-mode from pages marked readable *or executable*
-	        are allowed as long as the page is marked readable (or executable and `status.MXR` is set) in the
-	        G-stage translation.
-	      * When 0, load in effective VU-mode or VS-mode from pages marked executable raise a
-	        Page Fault exception (unless `sstatus.MXR` is also set, in which case the above applies).
-
-	      'vsstatus.MXR' affects all loads that operate under effective VU-mode or VS-mode:
-
-	      * All loads in VU-mode
-	      * All loads in VS-mode
-	      * Loads in M-mode when `mstatus.MPRV` == 1, `mstatus.MPV` == 1, and `mstatus.MPP`[1] == 0
-	      * Loads generated by one of the `hlv.*` instructions."
-
-		type: RW
-		alias: mstatus.MXR
-		definedBy: "S"
-
-	SUM:
-		location: 18
-		reset_value: UNDEFINED_LEGAL
-		description: "|
-	      *permit Supervisor Memory Access*
-
-	      Allows VS-mode to read user pages.
-
-	      Applies to the following loads and stores:
-
-	      * All loads and stores in VS-mode.
-	      * All loads and stores in M-mode when `mstatus.MPRV` == 1, `mstatus.MPP` == 1, and `mstatus.MPV` == 1
-	      * Loads and stores generated by one of the `hlv.*`, `hlvx.*`, or `hsv.*` instructions.
-
-	      When `vsstatus.SUM` is 0, the loads and stores from the above categories cause an
-	      `Illegal Instruction` exception if they access a user page during VS-level translation.
-	      Otherwise, a load or store from the above categories is permitted to access a user page
-	      during VS-level translation."
-
-		type: RW
-		definedBy: "S"
-
-	XS:
-		location: 16-15
-		reset_value: 0
-		description: "|
-	      *Custom (X) extension context Status*
-
-	      Summarizes the current state of any custom extension state.
-	      Either 0 - Off, 1 - Initial, 2 - Clean, 3 - Dirty.
-	      Since there are no custom extensions, this field is read-only 0."
-		type: RO
-		alias: mstatus.XS
-
-	FS:
-		location: 14-13
-		reset_value: UNDEFINED_LEGAL
-		description: "|
-	      *Floating point context status*
-
-	      When 0, floating point instructions (from F and D extensions) in VS-mode or VU-mode are disabled,
-	      and cause ILLEGAL INSTRUCTION exceptions.
-	      Floating point instructions in all modes, including VS-mode and VU-mode,
-	      are similarly disabled when `mstatus.FS` is clear.
-
-	      When a floating point register, or the `fcsr` register is written in VS-mode or VU-mode,
-	      `vsstatus.FS` is written with the value 3.
-
-	      Values 1 and 2 are valid write values for software, but are not interpreted by hardware
-	      other than to possibly enable a previously-disabled floating point unit."
-		type: RW-H
-		definedBy: "F"
-
-	VSD: # I think there's an issue with this being a keyword
-	    location: 10-9
-	    reset_value: UNDEFINED_LEGAL
-	    description: "|
-	      *Vector context status*
-
-	      When 0, vector instructions (from the V extension) are disabled, and cause ILLEGAL INSTRUCTION exceptions.
-	      When a vector register or vector CSR is written, VS obtains the value 3.
-	      Values 1 and 2 are valid write values for software, but are not interpreted by hardware
-	      other than to possibly enable a previously-disabled vector unit."
-	    type: RW-H
-	    definedBy: "V"
-
-	SPP:
-	    location: 8
-	    reset_value: UNDEFINED_LEGAL
-	    description: "|
-	      *VS-mode Previous Privilege*
-
-	      Written with the prior nominal privilege level (_i.e._, 0 for VU-mode and 1 for VS-mode)
-	      when entering VS-mode from an exception/interrupt.
-	      Can also be written by software without immediate side-effect.
-
-	      On a return from an exception from VS-mode, the machine will enter the nominal privilege level
-	      stored in `vsstatus.SPP`."
-	    type: RW-H
-
-	UBE:
-		location: 6
-		reset_value(): '|
-	      if (VU_MODE_ENDIANNESS == "little") {
-	        # little endian
-	        return 0;
-	      } else if (VU_MODE_ENDIANNESS == "big") {
-	        # big endian
-	        return 1;
-	      } else {
-	        # mutable
-	        return UNDEFINED_LEGAL;
-	      }'
-		description: "|
-	      *VU-mode Big Endian*
-
-	      Controls the endianness of VU-mode (0 = little, 1 = big).
-
-	      [when,\"VU_MODE_ENDIANNESS == 'little'\"]
-	      Since the CPU does not support big endian, this is hardwired to 0.
-
-	      [when,\"VU_MODE_ENDIANNESS == 'big'\"]
-	      Since the CPU does not support big endian, this is hardwired to 1."
-		type(): '|
-	      return (VU_MODE_ENDIANNESS == "dynamic") ? CsrFieldType::RW : CsrFieldType::RO;'
-		definedBy: "S"
-
-	SPIE:
-		location: 5
-		reset_value: UNDEFINED_LEGAL
-		description: "|
-      		*VS-mode Previous Interrupt Enable*
-
-			 Written by hardware in two cases:
-
-			 * Written with prior value of `vsstatus.SIE` when entering VS-mode from an exception/interrupt.
-			 * Written with the value 1 when returning from an exception in VS-mode (via the `sret` instruction).
-
-			 Can also be written by software without immediate side effect.
-
-			 Other than serving as a record of nested traps as described above, `vsstatus.SPIE` does not affect execution."
-		type: RW-H
-		definedBy: "S"
-
-	SIE:
-		location: 1
-		reset_value: UNDEFINED_LEGAL
-		description: "|
-      		*VS-mode Interrupt Enable*
-
-      		Written by hardware in two cases:
-
-	      	* Written with the value 0 when entering VS-mode from an exception/interrupt.
-	      	* Written with the prior value of `vsstatus.SPIE` when returning from an exception in VS-mode (via `sret`).
-
-	      	Affects execution by:
-
-	      	* When 0, all VS-mode interrupts are disabled when the current privilege level is VS ((H)S-mode and M-mode interrupts are still enabled).
-	      	* When 1, VS-mode interrupts that are not otherwise disabled with a field in `vsie` are enabled."
-		type: RW-H
diff --git a/tools/eclipse/run/testIdl/.classpath b/tools/eclipse/run/testudb/.classpath
similarity index 100%
rename from tools/eclipse/run/testIdl/.classpath
rename to tools/eclipse/run/testudb/.classpath
diff --git a/tools/eclipse/run/testIdl/src/module-info.java b/tools/eclipse/run/testudb/src/module-info.java
similarity index 100%
rename from tools/eclipse/run/testIdl/src/module-info.java
rename to tools/eclipse/run/testudb/src/module-info.java
diff --git a/tools/eclipse/run/testudb/src/vcsr.udb b/tools/eclipse/run/testudb/src/vcsr.udb
new file mode 100644
index 0000000000..f6be19577b
--- /dev/null
+++ b/tools/eclipse/run/testudb/src/vcsr.udb
@@ -0,0 +1,36 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# yaml-language-server: $schema=../../schemas/csr_schema.json
+
+$schema: "csr_schema.json#"
+kind: csr
+name: vcsr
+long_name: "Vector Control and Status Register"
+address: 0x00F
+writable: true
+priv_mode: U
+length: MXLEN
+description: "Contains aliases to vxrm and vxsat CSRs"
+definedBy:'
+  extension:
+    name: Zvl32b'
+fields:
+  VXRM:
+    location: 2-1
+    description: "See vxrm."
+    type: RW-RH
+    alias: "vxrm.VALUE[1:0]"
+    sw_write(csr_value): '|
+      CSR[vxrm].VALUE = csr_value.VXRM;
+      return csr_value.VXRM;'
+    reset_value: UNDEFINED_LEGAL
+  VXSAT:
+    location: 0
+    description: "See vxsat."
+    type: RW-RH
+    alias: "vxsat.VALUE[0]"
+    sw_write(csr_value): "|
+      CSR[vxsat].VALUE = csr_value.VXSAT;
+      return csr_value.VXSAT;"
+    reset_value: UNDEFINED_LEGAL
\ No newline at end of file
diff --git a/tools/eclipse/run/testudb/src/vsstatus.udb b/tools/eclipse/run/testudb/src/vsstatus.udb
new file mode 100644
index 0000000000..c41000e6bb
--- /dev/null
+++ b/tools/eclipse/run/testudb/src/vsstatus.udb
@@ -0,0 +1,232 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+# yaml-language-server: $schema=../../../schemas/csr_schema.json
+
+$schema: "csr_schema.json#"
+kind: csr
+name: vsstatus
+long_name: "Virtual Supervisor Status"
+address: 0x200
+writable: true
+virtual_address: 0x100
+priv_mode: VS
+length: VSXLEN
+description: "|
+  The vsstatus register tracks and controls the hart's current operating state.
+
+  It is VS-mode's version of `sstatus`, and substitutes for it when in VS-mode
+  (_i.e._, in VS-mode CSR address 0x100 is `vsstatus`, not `sstatus`).
+
+  Unlike the relationship between `sstatus` and `mstatus`, none of the bits in `vsstatus` are
+  aliases of another field."
+definedBy:'
+  extension:
+    name: H'
+fields:
+  SD:
+    location_rv64: 63
+    location_rv32: 31
+    description: "|
+      *State Dirty*
+
+      Read-only bit that summarizes whether any of the
+      `vsstatus.FS`, <% if ext?(:V) %> `vsstatus.VS`, <% end %> or `vsstatus.XS`
+      fields signal the presence of some dirty state
+      (_i.e._, any of them hold the value `11`).
+
+      This bit is _not_ an alias of `mstatus.SD` since
+      it only reflects the state visible to VS-mode
+      (_e.g._, `status.FS` does not affect `vsstatus.SD`)."
+    type: RO-H
+    reset_value: UNDEFINED_LEGAL
+    affectedBy: [F, D, V]
+  UXL:
+    location: 33-32
+    definedBy:'
+      xlen: 64'
+    description: '|
+      *VU-mode XLEN*
+
+      Sets the effective XLEN for VU-mode (0 = 32-bit, 1 = 64-bit, 2 = 128-bit).
+
+      [when,"VUXLEN == 32"]
+      Since the hart only supports VUXLEN==32, this is hardwired to 0.
+
+      [when,"VUXLEN == 64"]
+      Since the hart only supports VUXLEN==64, this is hardwired to 1.'
+
+    type(): "|
+      return ($array_size(VUXLEN) > 1) ? CsrFieldType::RW : CsrFieldType::RO;"
+    reset_value(): "|
+      return ($array_size(VUXLEN) > 1) ? UNDEFINED_LEGAL : VUXLEN[0];"
+  MXR:
+    alias: "mstatus.MXR"
+    location: 19
+    description: "|
+      *Make eXecutable Readable*
+
+      Makes it possible to read executable pages when loading from effective VU-mode or VS-mode
+      (normally, executable pages are not readable).
+
+      * When 1, load in effective VU-mode or VS-mode from pages marked readable *or executable*
+        are allowed as long as the page is marked readable (or executable and `status.MXR` is set) in the
+        G-stage translation.
+      * When 0, load in effective VU-mode or VS-mode from pages marked executable raise a
+        Page Fault exception (unless `sstatus.MXR` is also set, in which case the above applies).
+
+      'vsstatus.MXR' affects all loads that operate under effective VU-mode or VS-mode:
+
+      * All loads in VU-mode
+      * All loads in VS-mode
+      * Loads in M-mode when `mstatus.MPRV` == 1, `mstatus.MPV` == 1, and `mstatus.MPP`[1] == 0
+      * Loads generated by one of the `hlv.*` instructions."
+
+    type: RW
+    definedBy: '
+      extension:
+        name: S'
+    reset_value: UNDEFINED_LEGAL
+  SUM:
+    location: 18
+    description: "|
+      *permit Supervisor Memory Access*
+
+      Allows VS-mode to read user pages.
+
+      Applies to the following loads and stores:
+
+      * All loads and stores in VS-mode.
+      * All loads and stores in M-mode when `mstatus.MPRV` == 1, `mstatus.MPP` == 1, and `mstatus.MPV` == 1
+      * Loads and stores generated by one of the `hlv.*`, `hlvx.*`, or `hsv.*` instructions.
+
+      When `vsstatus.SUM` is 0, the loads and stores from the above categories cause an
+      `Illegal Instruction` exception if they access a user page during VS-level translation.
+      Otherwise, a load or store from the above categories is permitted to access a user page
+      during VS-level translation."
+
+    type: RW
+    definedBy:'
+      extension:
+        name: S'
+    reset_value: UNDEFINED_LEGAL
+  XS:
+    alias: "mstatus.XS"
+    location: 16-15
+    description: "|
+      *Custom (X) extension context Status*
+
+      Summarizes the current state of any custom extension state.
+      Either 0 - Off, 1 - Initial, 2 - Clean, 3 - Dirty.
+      Since there are no custom extensions, this field is read-only 0."
+    type: RO
+    reset_value: 0
+  FS:
+    location: 14-13
+    description: "|
+      *Floating point context status*
+
+      When 0, floating point instructions (from F and D extensions) in VS-mode or VU-mode are disabled,
+      and cause ILLEGAL INSTRUCTION exceptions.
+      Floating point instructions in all modes, including VS-mode and VU-mode,
+      are similarly disabled when `mstatus.FS` is clear.
+
+      When a floating point register, or the `fcsr` register is written in VS-mode or VU-mode,
+      `vsstatus.FS` is written with the value 3.
+
+      Values 1 and 2 are valid write values for software, but are not interpreted by hardware
+      other than to possibly enable a previously-disabled floating point unit."
+    type: RW-H
+    definedBy:'
+      extension:
+        name: F'
+    reset_value: UNDEFINED_LEGAL
+  VS:
+    location: 10-9
+    description: "|
+      *Vector context status*
+
+      When 0, vector instructions (from the V extension) are disabled, and cause ILLEGAL INSTRUCTION exceptions.
+      When a vector register or vector CSR is written, VS obtains the value 3.
+      Values 1 and 2 are valid write values for software, but are not interpreted by hardware
+      other than to possibly enable a previously-disabled vector unit."
+    type: RW-H
+    reset_value: UNDEFINED_LEGAL
+    definedBy: '
+      extension:
+        name: Zvl32b'
+  SPP:
+    location: 8
+    description: "|
+      *VS-mode Previous Privilege*
+
+      Written with the prior nominal privilege level (_i.e._, 0 for VU-mode and 1 for VS-mode)
+      when entering VS-mode from an exception/interrupt.
+      Can also be written by software without immediate side-effect.
+
+      On a return from an exception from VS-mode, the machine will enter the nominal privilege level
+      stored in `vsstatus.SPP`."
+    type: RW-H
+    reset_value: UNDEFINED_LEGAL
+  UBE:
+    location: 6
+    description: "|
+      *VU-mode Big Endian*
+
+      Controls the endianness of VU-mode (0 = little, 1 = big).
+
+      [when,\"VU_MODE_ENDIANNESS == 'little'\"]
+      Since the CPU does not support big endian, this is hardwired to 0.
+
+      [when,\"VU_MODE_ENDIANNESS == 'big'\"]
+      Since the CPU does not support big endian, this is hardwired to 1."
+    type(): '|
+      return (VU_MODE_ENDIANNESS == "dynamic") ? CsrFieldType::RW : CsrFieldType::RO;'
+    definedBy:'
+      extension:
+        name: S'
+    reset_value(): '|
+      if (VU_MODE_ENDIANNESS == "little") {
+        # little endian
+        return 0;
+      } else if (VU_MODE_ENDIANNESS == "big") {
+        # big endian
+        return 1;
+      } else {
+        # mutable
+        return UNDEFINED_LEGAL;
+      }'
+  SPIE:
+    location: 5
+    description: "|
+      *VS-mode Previous Interrupt Enable*
+
+      Written by hardware in two cases:
+
+      * Written with prior value of `vsstatus.SIE` when entering VS-mode from an exception/interrupt.
+      * Written with the value 1 when returning from an exception in VS-mode (via the `sret` instruction).
+
+      Can also be written by software without immediate side effect.
+
+      Other than serving as a record of nested traps as described above, `vsstatus.SPIE` does not affect execution."
+    type: RW-H
+    definedBy: '
+      extension:
+        name: S'
+    reset_value: UNDEFINED_LEGAL
+  SIE:
+    location: 1
+    description: "|
+      *VS-mode Interrupt Enable*
+
+      Written by hardware in two cases:
+
+      * Written with the value 0 when entering VS-mode from an exception/interrupt.
+      * Written with the prior value of `vsstatus.SPIE` when returning from an exception in VS-mode (via `sret`).
+
+      Affects execution by:
+
+      * When 0, all VS-mode interrupts are disabled when the current privilege level is VS ((H)S-mode and M-mode interrupts are still enabled).
+      * When 1, VS-mode interrupts that are not otherwise disabled with a field in `vsie` are enabled."
+    type: RW-H
+    reset_value: UNDEFINED_LEGAL
\ No newline at end of file
