diff --git a/backends/instructions_appendix/all_instructions.golden.adoc b/backends/instructions_appendix/all_instructions.golden.adoc
index 9e3e7ec94e..e2a8b138dc 100644
--- a/backends/instructions_appendix/all_instructions.golden.adoc
+++ b/backends/instructions_appendix/all_instructions.golden.adoc
@@ -42941,7 +42941,7 @@ Included in::
 == vle8.v
 
 Synopsis::
-Vector load 8-bit elements
+Vector 8-bit unit-stride load
 
 Base::
 
@@ -58519,7 +58519,7 @@ Included in::
 == vse8.v
 
 Synopsis::
-Vector store 8-bit elements
+Vector 8-bit unit-stride store
 
 Base::
 
diff --git a/ext/docs-resources b/ext/docs-resources
index 7a9931f276..be002f1c6d 160000
--- a/ext/docs-resources
+++ b/ext/docs-resources
@@ -1 +1 @@
-Subproject commit 7a9931f276a59ad24730c92e068cb5506bce0d7a
+Subproject commit be002f1c6d7346a471506f69782f2d6bac4eff97
diff --git a/ext/riscv-isa-manual b/ext/riscv-isa-manual
index 2a857ebba2..b9f1438191 160000
--- a/ext/riscv-isa-manual
+++ b/ext/riscv-isa-manual
@@ -1 +1 @@
-Subproject commit 2a857ebba262ca65b9c83f40d1523dd9e1ddf8a6
+Subproject commit b9f1438191c25c95e877e521431f949990f8aa85
diff --git a/spec/std/isa/inst/V/vadd.vv.yaml b/spec/std/isa/inst/V/vadd.vv.yaml
index 761389bb80..4d309ba4e6 100644
--- a/spec/std/isa/inst/V/vadd.vv.yaml
+++ b/spec/std/isa/inst/V/vadd.vv.yaml
@@ -44,7 +44,7 @@ operation(): |
     vlmax = (vlen >> state.log2_lmul) >> state.log2_sew;
   }
 
-  for (U32 i = CSR[vstart].VALUE; i < vlmax; i++) {
+  for (U32 i = CSR[vstart].VALUE; i < CSR[vl].VALUE; i++) {
     U32 start_bit_pos = i * state.sew;
     U32 end_bit_pos = start_bit_pos + state.sew - 1;
     v[vd] = {
diff --git a/spec/std/isa/inst/V/vle8.v.yaml b/spec/std/isa/inst/V/vle8.v.yaml
index fb6fdab3ca..b20b7a3d59 100644
--- a/spec/std/isa/inst/V/vle8.v.yaml
+++ b/spec/std/isa/inst/V/vle8.v.yaml
@@ -6,7 +6,7 @@
 $schema: "inst_schema.json#"
 kind: instruction
 name: vle8.v
-long_name: Vector load 8-bit elements
+long_name: Vector 8-bit unit-stride load
 description: |
   Loads 8-bit elements from memory starting at address in xs1 register, using vector mask vm, and stores results in vector register vd. The effective element width is 8 bits regardless of SEW setting. Supports vector masking and follows vector tail policy.
 definedBy:
@@ -29,6 +29,18 @@ access:
   vu: always
 data_independent_timing: false
 operation(): |
+  XReg virtual_address = X[xs1];
+  # "A simple in-order implementation can ignore the settings and simply execute all vector instructions using the undisturbed policy"
+  for (U32 i = 0; i < CSR[vl].VALUE; i++) {
+    U32 start_bit_pos = (CSR[vstart].VALUE + i) * 8;
+    U32 end_bit_pos = start_bit_pos + 7;
+    v[vd] = {
+      v[vd][VLEN-1:end_bit_pos + 1],
+      sext(read_memory<8>(virtual_address, $encoding), 8),
+      v[vd][start_bit_pos-1:0]
+    };
+    virtual_address = virtual_address + 1;
+  }
 
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
diff --git a/spec/std/isa/inst/V/vse8.v.yaml b/spec/std/isa/inst/V/vse8.v.yaml
index 8284104f91..71e82b194d 100644
--- a/spec/std/isa/inst/V/vse8.v.yaml
+++ b/spec/std/isa/inst/V/vse8.v.yaml
@@ -6,7 +6,7 @@
 $schema: "inst_schema.json#"
 kind: instruction
 name: vse8.v
-long_name: Vector store 8-bit elements
+long_name: Vector 8-bit unit-stride store
 description: |
   Stores 8-bit elements from vector register vs3 to memory starting at address in xs1 register, using vector mask vm. The effective element width is 8 bits regardless of SEW setting.
 definedBy:
@@ -29,6 +29,14 @@ access:
   vu: always
 data_independent_timing: false
 operation(): |
+  XReg virtual_address = X[xs1];
+  # "A simple in-order implementation can ignore the settings and simply execute all vector instructions using the undisturbed policy"
+  for (U32 i = 0; i < CSR[vl].VALUE; i++) {
+    U32 start_bit_pos = (CSR[vstart].VALUE + i) * 8;
+    U32 end_bit_pos = start_bit_pos + 7;
+    write_memory<8>(virtual_address, v[vs3][end_bit_pos:start_bit_pos], $encoding);
+    virtual_address = virtual_address + 1;
+  }
 
 # SPDX-SnippetBegin
 # SPDX-FileCopyrightText: 2017-2025 Contributors to the RISCV Sail Model <https://github.com/riscv/sail-riscv/blob/master/LICENCE>
diff --git a/tests/isa/rv32uv/Makefrag b/tests/isa/rv32uv/Makefrag
index c5b2f51355..7ac1c77f89 100644
--- a/tests/isa/rv32uv/Makefrag
+++ b/tests/isa/rv32uv/Makefrag
@@ -2,7 +2,7 @@
 # Makefrag for rv32uv tests
 #-----------------------------------------------------------------------
 
-rv32uv_sc_tests = \
+rv32uv_sc_tests = vadd.vv vle8 \
     vsetivli \
     vsetvl \
     vsetvli_vl_lt_vlmax \
diff --git a/tests/isa/rv32uv/vadd.vv.S b/tests/isa/rv32uv/vadd.vv.S
new file mode 100644
index 0000000000..9b65d4285e
--- /dev/null
+++ b/tests/isa/rv32uv/vadd.vv.S
@@ -0,0 +1,9 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+# See LICENSE for license details
+
+#include "riscv_test.h"
+#undef RVTEST_RV64UV
+#define RVTEST_RV64UV RVTEST_RV32UV
+
+#include "../rv64uv/vadd.vv.S"
diff --git a/tests/isa/rv32uv/vle8.S b/tests/isa/rv32uv/vle8.S
new file mode 100644
index 0000000000..f93495dedf
--- /dev/null
+++ b/tests/isa/rv32uv/vle8.S
@@ -0,0 +1,9 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+# See LICENSE for license details
+
+#include "riscv_test.h"
+#undef RVTEST_RV64UV
+#define RVTEST_RV64UV RVTEST_RV32UV
+
+#include "../rv64uv/vle8.S"
diff --git a/tests/isa/rv32uv/vsetivli.S b/tests/isa/rv32uv/vsetivli.S
index e698d285fb..534f9c444d 100644
--- a/tests/isa/rv32uv/vsetivli.S
+++ b/tests/isa/rv32uv/vsetivli.S
@@ -3,7 +3,7 @@
 # See LICENSE for license details
 
 #include "riscv_test.h"
-#undef RVTEST_RV64U
-#define RVTEST_RV64U RVTEST_RV32U
+#undef RVTEST_RV64UV
+#define RVTEST_RV64UV RVTEST_RV32UV
 
 #include "../rv64uv/vsetivli.S"
diff --git a/tests/isa/rv32uv/vsetvl.S b/tests/isa/rv32uv/vsetvl.S
index 2871bf5c8c..10d500bdce 100644
--- a/tests/isa/rv32uv/vsetvl.S
+++ b/tests/isa/rv32uv/vsetvl.S
@@ -3,7 +3,7 @@
 # See LICENSE for license details
 
 #include "riscv_test.h"
-#undef RVTEST_RV64U
-#define RVTEST_RV64U RVTEST_RV32U
+#undef RVTEST_RV64UV
+#define RVTEST_RV64UV RVTEST_RV32UV
 
 #include "../rv64uv/vsetvl.S"
diff --git a/tests/isa/rv32uv/vsetvli_rs1_eq_zero.S b/tests/isa/rv32uv/vsetvli_rs1_eq_zero.S
index 806d6b8d0e..4eefa01a9e 100644
--- a/tests/isa/rv32uv/vsetvli_rs1_eq_zero.S
+++ b/tests/isa/rv32uv/vsetvli_rs1_eq_zero.S
@@ -1,7 +1,7 @@
 # See LICENSE for license details
 
 #include "riscv_test.h"
-#undef RVTEST_RV64U
-#define RVTEST_RV64U RVTEST_RV32U
+#undef RVTEST_RV64UV
+#define RVTEST_RV64UV RVTEST_RV32UV
 
 #include "../rv64uv/vsetvli_rs1_eq_zero.S"
diff --git a/tests/isa/rv32uv/vsetvli_vl_lt_vlmax.S b/tests/isa/rv32uv/vsetvli_vl_lt_vlmax.S
index 404141b704..d85e2f8c14 100644
--- a/tests/isa/rv32uv/vsetvli_vl_lt_vlmax.S
+++ b/tests/isa/rv32uv/vsetvli_vl_lt_vlmax.S
@@ -1,7 +1,7 @@
 # See LICENSE for license details
 
 #include "riscv_test.h"
-#undef RVTEST_RV64U
-#define RVTEST_RV64U RVTEST_RV32U
+#undef RVTEST_RV64UV
+#define RVTEST_RV64UV RVTEST_RV32UV
 
 #include "../rv64uv/vsetvli_vl_lt_vlmax.S"
diff --git a/tests/isa/rv64uv/Makefrag b/tests/isa/rv64uv/Makefrag
index 0f8aa6bcec..4f2c8af9fe 100644
--- a/tests/isa/rv64uv/Makefrag
+++ b/tests/isa/rv64uv/Makefrag
@@ -2,7 +2,7 @@
 # Makefrag for rv64uv tests
 #-----------------------------------------------------------------------
 
-rv64uv_sc_tests = \
+rv64uv_sc_tests = vadd.vv vle8 \
     vsetivli \
     vsetvl \
     vsetvli_vl_lt_vlmax \
diff --git a/tests/isa/rv64uv/vadd.vv.S b/tests/isa/rv64uv/vadd.vv.S
new file mode 100644
index 0000000000..5f2f469c3b
--- /dev/null
+++ b/tests/isa/rv64uv/vadd.vv.S
@@ -0,0 +1,58 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+# See LICENSE for license details
+
+#*****************************************************************************
+# vadd.vv.S
+#-----------------------------------------------------------------------------
+#
+# Test vadd.vv instruction.
+#
+# vadd.vv
+
+#include "riscv_test.h"
+#include "test_macros.h"
+
+RVTEST_RV64UV
+RVTEST_CODE_BEGIN
+
+# Write tests for vadd.vv
+
+# SEW=8, LMUL=1. x6 should contain VL == 2 (AVL, since its < VLMAX==128)
+# Use x5 to hold addresses
+# v2 == 0x2d2c
+# Store 0x2d2c to result[0..1] leaving other bits as 1
+# Load 4-bytes from result into x2 to compare with expected value
+
+  TEST_CASE(1, x2, 0xFFFFffffFFFF524F, \
+    vsetivli x6, 2, e8, m1; \
+    la      x5, input1; \
+    vle8.v  v2, (x5); \
+    la      x5, input2; \
+    vle8.v  v3, (x5); \
+    vadd.vv v4, v2, v3; \
+    la      x5, result; \
+    vse8.v  v4, (x5); \
+    lw      x2, (x5); \
+    )
+  TEST_PASSFAIL
+
+RVTEST_CODE_END
+
+# Input data section.
+# This section is optional, and this data is NOT saved in the output.
+.data
+  .balign 4
+input1:
+  .word 0xBABA2d2c
+input2:
+  .word 0xCAFE2523
+
+# Output data section.
+# This section is optional, but this data IS saved in the output.
+RVTEST_DATA_BEGIN
+
+result:
+  .word -1
+
+RVTEST_DATA_END
diff --git a/tests/isa/rv64uv/vle8.S b/tests/isa/rv64uv/vle8.S
new file mode 100644
index 0000000000..8b402a6223
--- /dev/null
+++ b/tests/isa/rv64uv/vle8.S
@@ -0,0 +1,54 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+# See LICENSE for license details
+
+#*****************************************************************************
+# vle8.S
+#-----------------------------------------------------------------------------
+#
+# Test vle8 instruction.
+#
+# vle8
+
+#include "riscv_test.h"
+#include "test_macros.h"
+
+RVTEST_RV64UV
+RVTEST_CODE_BEGIN
+
+# Write tests for vle8
+
+# SEW=8, LMUL=1. x6 should contain VL == 2 (AVL, since its < VLMAX==128)
+# Use x5 to hold addresses
+# v2 == 0x2d2c
+# Store 0x2d2c to result[0..1] leaving other bits as 1
+# Load 4-bytes from result into x2 to compare with expected value
+
+  TEST_CASE(1, x2, 0xFFFFffffFFFF2d2c, \
+    vsetivli x6, 2, e8, m1; \
+    la      x5, input; \
+    vle8.v  v2, (x5); \
+    la      x5, result; \
+    vse8.v  v2, (x5); \
+    lw      x2, (x5); \
+    )
+
+  TEST_PASSFAIL
+
+RVTEST_CODE_END
+
+# Input data section.
+# This section is optional, and this data is NOT saved in the output.
+.data
+  .balign 4
+input:
+  .word 0xBABA2d2c
+
+# Output data section.
+# This section is optional, but this data IS saved in the output.
+RVTEST_DATA_BEGIN
+
+result:
+  .word -1
+
+RVTEST_DATA_END
diff --git a/tools/scripts/new-rvtest b/tools/scripts/new-rvtest
new file mode 100755
index 0000000000..5f8c129f7f
--- /dev/null
+++ b/tools/scripts/new-rvtest
@@ -0,0 +1,73 @@
+#!/bin/bash
+# SPDX-FileCopyrightText: Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+
+help() {
+    echo "Usage: $0 <new-test-name>"
+    echo "This script copies new-rvtest-{64,32}-template.S files "
+    echo "and updates Makefrag files for RV64 and RV32."
+    echo
+    echo "Please provide the new test name as the first argument."
+    exit 1
+}
+
+# Check if the first argument is provided
+if [ -z "$1" ]; then
+    help
+fi
+
+# Validate that $1 only contains letters, numbers and punctuation '.', '-' and '_'
+if [[ ! "$1" =~ ^[A-Za-z0-9._-]+$ ]]; then
+    echo "Error: test name must contain only letters, numbers, '.', '-' or '_'"
+    exit 1
+fi
+
+source_dir="tools/scripts"
+dest_dir="tests/isa"
+
+# Copy new-rvtest-{64,32}-template.S from tests/isa/rv64uv directory
+
+source_file="$source_dir/new-rvtest-64-template.S"
+dest_file="$dest_dir/rv64uv/$1.S"
+
+if [ -f "$source_file" ]; then
+    cp "$source_file" "$dest_file"
+    echo "Copied $source_file to $dest_file"
+    # replace __INSTR__ with the provided test name (escape sed delimiters)
+    sed -i "s/__INSTR__/$1/g" "$dest_file"
+else
+    echo "Error: $source_file not found"
+    exit 1
+fi
+
+source_file="$source_dir/new-rvtest-32-template.S"
+dest_file="$dest_dir/rv32uv/$1.S"
+
+if [ -f "$source_file" ]; then
+    cp "$source_file" "$dest_file"
+    echo "Copied $source_file to $dest_file"
+    # replace __INSTR__ with the provided test name (escape sed delimiters)
+    sed -i "s/__INSTR__/$1/g" "$dest_file"
+else
+    echo "Error: $source_file not found"
+    exit 1
+fi
+
+# Append to Makefrag
+makefrag_file="tests/isa/rv64uv/Makefrag"
+if [ -f "$makefrag_file" ]; then
+    sed -i "s/rv64uv_sc_tests =/rv64uv_sc_tests = $1/" "$makefrag_file"
+    echo "Updated $makefrag_file"
+else
+    echo "Error: $makefrag_file not found"
+    exit 1
+fi
+
+makefrag_file="tests/isa/rv32uv/Makefrag"
+if [ -f "$makefrag_file" ]; then
+    sed -i "s/rv32uv_sc_tests =/rv32uv_sc_tests = $1/" "$makefrag_file"
+    echo "Updated $makefrag_file"
+else
+    echo "Error: $makefrag_file not found"
+    exit 1
+fi
diff --git a/tools/scripts/new-rvtest-32-template.S b/tools/scripts/new-rvtest-32-template.S
new file mode 100644
index 0000000000..963131e549
--- /dev/null
+++ b/tools/scripts/new-rvtest-32-template.S
@@ -0,0 +1,9 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+# See LICENSE for license details
+
+#include "riscv_test.h"
+#undef RVTEST_RV64UV
+#define RVTEST_RV64UV RVTEST_RV32UV
+
+#include "../rv64uv/__INSTR__.S"
diff --git a/tools/scripts/new-rvtest-64-template.S b/tools/scripts/new-rvtest-64-template.S
new file mode 100644
index 0000000000..d0a00cb303
--- /dev/null
+++ b/tools/scripts/new-rvtest-64-template.S
@@ -0,0 +1,48 @@
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+# See LICENSE for license details
+
+#*****************************************************************************
+# __INSTR__.S
+#-----------------------------------------------------------------------------
+#
+# Test __INSTR__ instruction.
+#
+# __INSTR__
+
+#include "riscv_test.h"
+#include "test_macros.h"
+
+RVTEST_RV64UV
+RVTEST_CODE_BEGIN
+
+# Write tests for __INSTR__
+  li      x3, 1         # x3 (gp) is TESTNUM register
+  la      x5, input     # Use x5 to hold addresses.
+  lw      x2, (x5)
+  addi    x2, x2, 1     # Should be 45 in x2.
+  la      x5, result
+  sw      x2, (x5)      # Store result into memory overwriting 1s.
+  li      x4, 45        # Desired result.
+  bne     x2, x4, fail  # Fail out if doesn't match.
+  RVTEST_PASS           # Signal success.
+fail:
+  RVTEST_FAIL
+
+RVTEST_CODE_END
+
+# Input data section.
+# This section is optional, and this data is NOT saved in the output.
+.data
+  .balign 8
+input:
+  .word 44
+
+# Output data section.
+# This section is optional, but this data IS saved in the output.
+RVTEST_DATA_BEGIN
+
+result:
+  .word -1
+
+RVTEST_DATA_END
