diff --git a/.devcontainer/Dockerfile b/.devcontainer/Dockerfile
index 3170bcb4bd..a1b5cfd6be 100644
--- a/.devcontainer/Dockerfile
+++ b/.devcontainer/Dockerfile
@@ -1,4 +1,5 @@
 FROM ubuntu:24.04
+LABEL org.opencontainers.image.source=https://github.com/riscv-software-src/riscv-unified-db
 
 ENV DEBIAN_FRONTEND=noninteractive
 
diff --git a/.github/workflows/container.yml b/.github/workflows/container.yml
index d07d663cdb..aaffde25b5 100644
--- a/.github/workflows/container.yml
+++ b/.github/workflows/container.yml
@@ -47,6 +47,7 @@ jobs:
       - name: Upload singularity container
         run: |
           singularity registry login -u ${{ secrets.DOCKERHUB_USERNAME }} -p ${{ secrets.DOCKERHUB_TOKEN }} oras://docker.io
+          singularity push .singularity/image.sif oras://gchr.io/riscvintl/spec-generator:${{ env.TAG }}
           singularity push .singularity/image.sif oras://docker.io/riscvintl/spec-generator:${{ env.TAG }}
 
   build-docker-container:
@@ -140,3 +141,21 @@ jobs:
       - name: Inspect image
         run: |
           docker buildx imagetools inspect riscvintl/udb:${{ env.TAG }}
+
+      - name: Login to GHCR
+        uses: docker/login-action@5e57cd118135c172c3672efd75eb46360885c0ef # v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Create manifest list and push
+        working-directory: /tmp/digests
+        run: |
+          docker buildx imagetools create \
+            --tag riscvintl/udb:${{ env.TAG }} \
+            $(printf 'riscvintl/udb@sha256:%s ' *)
+
+      - name: Inspect image
+        run: |
+          docker buildx imagetools inspect riscvintl/udb:${{ env.TAG }}
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 357ad54bcf..5f0b78dc70 100755
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -20,7 +20,7 @@ repos:
       - id: end-of-file-fixer
         stages: [pre-commit]
         # Exclude reference and generated files
-        exclude: (cfgs/profile.*|.*\.golden\.adoc$|.*\.svg|amo.*\.yaml|\.github/workflows/regress\.yml$)
+        exclude: (cfgs/profile.*|.*\.golden\.adoc$|.*\.svg|amo.*\.yaml|container\.def|\.github/workflows/regress\.yml$)
       - id: trailing-whitespace
         stages: [pre-commit]
         args: [--markdown-linebreak-ext=md]
diff --git a/bin/.container-tag b/bin/.container-tag
index 2856407c03..50653ad0a6 100644
--- a/bin/.container-tag
+++ b/bin/.container-tag
@@ -1 +1 @@
-0.15
+0.17
diff --git a/bin/setup b/bin/setup
index 6ffc41e242..82f6ca9f7b 100755
--- a/bin/setup
+++ b/bin/setup
@@ -5,6 +5,8 @@ cd "$ROOT" || exit
 
 CONTAINER_TAG=$(cat "${ROOT}/bin/.container-tag")
 
+REGISTRY=${REGISTRY:="gchr.io"}
+
 # shellcheck source=.functions.sh
 source "${ROOT}/bin/.functions.sh"
 
@@ -86,7 +88,7 @@ if [ -v GITHUB_ACTIONS ] && [ "${CONTAINER_TYPE}" == "singularity" ]; then
 elif [ "${CONTAINER_TYPE}" == "docker" ] || [ "${CONTAINER_TYPE}" == "podman" ]; then
     if ! ${CONTAINER_TYPE} images "riscvintl/udb:${CONTAINER_TAG}" | grep -q udb ; then
         echo "Image not found locally. Attempting to pull from registry..."
-        if ${CONTAINER_TYPE} pull "docker.io/riscvintl/udb:${CONTAINER_TAG}" 2>/dev/null; then
+        if ${CONTAINER_TYPE} pull "$REGISTRY/riscvintl/udb:${CONTAINER_TAG}" 2>/dev/null; then
             echo "Successfully pulled image from registry."
         else
             echo "Could not pull image from registry. Building locally..."
@@ -148,7 +150,7 @@ elif [ "${CONTAINER_TYPE}" == "singularity" ]; then
       if [ ! -d "${ROOT}/.singularity" ]; then
         mkdir -p "${ROOT}/.singularity"
       fi
-      singularity pull "${CONTAINER_PATH}" "oras://docker.io/riscvintl/spec-generator:$CONTAINER_TAG"
+      singularity pull "${CONTAINER_PATH}" "oras://$REGISTRY/riscvintl/spec-generator:$CONTAINER_TAG"
     fi
     if [ -v UDB_CONTAINER_BIND ]; then
       EXTRA_BIND="--bind ${UDB_CONTAINER_BIND}:${UDB_CONTAINER_BIND}"
diff --git a/container.def b/container.def
index e9a0a18f6b..5f2fa63049 100644
--- a/container.def
+++ b/container.def
@@ -28,6 +28,8 @@ tools/ruby-gems/udb_helpers/lib/udb_helpers/version.rb /opt/tools/ruby-gems/udb_
 tools/ruby-gems/idlc/bin/idlc /workspace/bin/container
 tools/ruby-gems/udb/bin/udb /workspace/bin/container
 tools/ruby-gems/udb-gen/bin/udb-gen /workspace/bin/container
+%labels
+org.opencontainers.image.source=https://github.com/riscv-software-src/riscv-unified-db
 %post -c /bin/bash
 
 DEBIAN_FRONTEND=noninteractive
