diff --git a/tools/test/regress-cli.rb b/tools/test/regress-cli.rb
index 24e0f5a7a4..bc12a55dec 100644
--- a/tools/test/regress-cli.rb
+++ b/tools/test/regress-cli.rb
@@ -159,11 +159,11 @@ def cmd_run_single_test(test_name)
     end
   end
 
-  sig { void }
+  sig { params(tag_name: String).void }
   def cmd_run_tagged_tests(tag_name)
     nran = 0
     test_data.fetch("tests").each do |tname, tdata|
-      next unless tdata.key?("tag") && tdata.fetch("tags").include?(tag_name)
+      next unless tdata.key?("tags") && tdata.fetch("tags").include?(tag_name)
       nran += 1
       cmd_run_single_test(tname)
     end
