diff --git a/.github/workflows/pages.yml b/.github/workflows/pages.yml
index 46d615bdb2..a9d89c2f15 100644
--- a/.github/workflows/pages.yml
+++ b/.github/workflows/pages.yml
@@ -234,7 +234,7 @@ jobs:
           run-id: ${{ github.event.workflow_run.id }}
       - name: Create index
         env:
-          PAGES_URL: https://riscv-software-src.github.io/riscv-unified-db
+          PAGES_URL: https://riscv.github.io/riscv-unified-db
         run: >
           cp doc/udb-block.svg _site/ &&
           ruby -r erb -r date -e "File.write('_site/index.html', ERB.new(File.read('tools/scripts/pages.html.erb'), trim_mode: '-').result(binding))"
diff --git a/CONTRIBUTING.adoc b/CONTRIBUTING.adoc
index b5dacfbcc4..d2f0d22a67 100644
--- a/CONTRIBUTING.adoc
+++ b/CONTRIBUTING.adoc
@@ -7,23 +7,23 @@ We welcome all contributions that meet our xref:CODE_OF_CONDUCT.adoc[code of con
 
 === Bug Reports
 
-Bug reports should be filed as a https://github.com/riscv-software-src/riscv-unified-db/issues[GitHub Issue] using the `Bug report` template.
+Bug reports should be filed as a https://github.com/riscv/riscv-unified-db/issues[GitHub Issue] using the `Bug report` template.
 
 === Feature Requests
 
-Feature requests should be filed as a https://github.com/riscv-software-src/riscv-unified-db/issues[GitHub Issue] using the `Feature request` template.
+Feature requests should be filed as a https://github.com/riscv/riscv-unified-db/issues[GitHub Issue] using the `Feature request` template.
 
 === Bug Fixes
 
 Bug fixes should be submitted as a Pull Request.
 
-There should be a corresponding https://github.com/riscv-software-src/riscv-unified-db/issues[GitHub Issue] for the PR. Ensure the two are linked by adding "closes #<issue number>" in the PR description.
+There should be a corresponding https://github.com/riscv/riscv-unified-db/issues[GitHub Issue] for the PR. Ensure the two are linked by adding "closes #<issue number>" in the PR description.
 
 === New data, tools, and features
 
 New data (for example, adding an extension), tools, or features should be submitted as a Pull Request.
 
-There should be a corresponding https://github.com/riscv-software-src/riscv-unified-db/issues[GitHub Issue] for the PR. Ensure the two are linked by adding "closes #<issue number>" in the PR description.
+There should be a corresponding https://github.com/riscv/riscv-unified-db/issues[GitHub Issue] for the PR. Ensure the two are linked by adding "closes #<issue number>" in the PR description.
 
 == Submitting a Pull Request
 
@@ -155,7 +155,7 @@ Code Owners are maintained in `.github/CODEOWNERS`.
 
 == Finding tasks
 
-If you are looking to contribute but are unsure what to do, browse through the https://github.com/riscv-software-src/riscv-unified-db/issues[issues]. We try to keep them tagged by area and mark easier tasks with "good first issue".
+If you are looking to contribute but are unsure what to do, browse through the https://github.com/riscv/riscv-unified-db/issues[issues]. We try to keep them tagged by area and mark easier tasks with "good first issue".
 
 == Legal
 
diff --git a/README.adoc b/README.adoc
index c4e1be7e48..ec810810f4 100644
--- a/README.adoc
+++ b/README.adoc
@@ -1,6 +1,6 @@
 image::doc/udb.svg[UDB banner]
 
-image::https://api.reuse.software/badge/github.com/riscv-software-src/riscv-unified-db[REUSE status, link="https://api.reuse.software/info/github.com/riscv-software-src/riscv-unified-db"]
+image::https://api.reuse.software/badge/github.com/riscv/riscv-unified-db[REUSE status, link="https://api.reuse.software/info/github.com/riscv/riscv-unified-db"]
 
 
 [WARNING]
@@ -17,7 +17,7 @@ That said, the following data has been validated against third-party sources:
 
 == Quick links
 
-* https://riscv-software-src.github.io/riscv-unified-db/index.html[The latest UDB-generated **UNOFFICIAL** RISC-V specifications].
+* https://riscv.github.io/riscv-unified-db/index.html[The latest UDB-generated **UNOFFICIAL** RISC-V specifications].
 * UDB Documentation
 ** xref:doc/schemas.adoc[Schema documentation]
 ** xref:doc/idl.adoc[ISA Description Language (IDL)]
diff --git a/REUSE.toml b/REUSE.toml
index 6c459c858f..6eb76599b1 100644
--- a/REUSE.toml
+++ b/REUSE.toml
@@ -27,7 +27,7 @@ path = [
 ]
 SPDX-FileCopyrightText = [
     "Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.",
-    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv-software-src/riscv-unified-db>",
+    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv/riscv-unified-db>",
 ]
 SPDX-License-Identifier = "BSD-3-Clause-Clear"
 
@@ -51,7 +51,7 @@ path = [
 precedence = "aggregate"
 SPDX-FileCopyrightText = [
     "Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.  ",
-    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv-software-src/riscv-unified-db>",
+    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv/riscv-unified-db>",
     # For import isa-manual docs: https://github.com/riscv/riscv-isa-manual
     "2017-2025 Contributors to the RISCV ISA Manual <https://github.com/riscv/riscv-isa-manual>",
 ]
diff --git a/backends/cfg_html_doc/templates/landing.adoc.erb b/backends/cfg_html_doc/templates/landing.adoc.erb
index 2f24553934..fad8673ae6 100644
--- a/backends/cfg_html_doc/templates/landing.adoc.erb
+++ b/backends/cfg_html_doc/templates/landing.adoc.erb
@@ -14,4 +14,4 @@ Additionally, the following documentation is also included:
  * xref:prose:idl.adoc[IDL language reference]
 
 This site was generated using
-https://github.com/riscv-software-src/riscv-unified-db[the RISC-V Unified Database].
+https://github.com/riscv/riscv-unified-db[the RISC-V Unified Database].
diff --git a/doc/ci.adoc b/doc/ci.adoc
index dd5802de70..b0b53c6e48 100644
--- a/doc/ci.adoc
+++ b/doc/ci.adoc
@@ -35,7 +35,7 @@ The tests are specified in xref:.github/workflows/regress.yml[`.github/workflows
 ==== Deployment Tests / Actions
 
 The second group of tests run when a PR enters the merge queue.
-They generate artifacts that are deployed to https://riscv-software-src.github.io/riscv-unified-db/[GitHub pages].
+They generate artifacts that are deployed to https://riscv.github.io/riscv-unified-db/[GitHub pages].
 If any test in the second group fails, a PR is removed from the merge queue.
 While in the merge queue, the CI tests generate artifacts but do not upload/update pages yet.
 The upload occurs after the PR merges to main and the test runs again.
diff --git a/ext/auto-inst/parsing.py b/ext/auto-inst/parsing.py
index 6460da46ad..d6e4481e6c 100755
--- a/ext/auto-inst/parsing.py
+++ b/ext/auto-inst/parsing.py
@@ -1,5 +1,5 @@
 # SPDX-FileCopyrightText: Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
-# SPDX-FileCopyrightText: 2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv-software-src/riscv-unified-db>
+# SPDX-FileCopyrightText: 2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv/riscv-unified-db>
 # SPDX-License-Identifier: BSD-3-Clause-Clear
 
 import os
diff --git a/ext/auto-inst/test_parsing.py b/ext/auto-inst/test_parsing.py
index 706b48f0cd..c89fc9f1db 100755
--- a/ext/auto-inst/test_parsing.py
+++ b/ext/auto-inst/test_parsing.py
@@ -1,5 +1,5 @@
 # SPDX-FileCopyrightText: Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
-# SPDX-FileCopyrightText: 2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv-software-src/riscv-unified-db>
+# SPDX-FileCopyrightText: 2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv/riscv-unified-db>
 # SPDX-License-Identifier: BSD-3-Clause-Clear
 
 import pytest
diff --git a/spec/std/isa/profile/RVB23S64.yaml b/spec/std/isa/profile/RVB23S64.yaml
index 3518fc946e..8bd45d5b59 100644
--- a/spec/std/isa/profile/RVB23S64.yaml
+++ b/spec/std/isa/profile/RVB23S64.yaml
@@ -194,7 +194,7 @@ extensions:
       The set of augmented hypervisor extensions: H, Ssstateen,
       Shcounterenw, Shvstvala, Shtvala, Shvstvecd, Shvsatpa, Shgatpa
 
-  # TODO: See https://github.com/riscv-software-src/riscv-unified-db/issues/373
+  # TODO: See https://github.com/riscv/riscv-unified-db/issues/373
   #  __NEED_AN_EXTENSION_NAME_HERE__:
   #    presence:
   #      optional: expansion
diff --git a/tools/python/profile_extensions.golden.license b/tools/python/profile_extensions.golden.license
index d24c2d02b0..9fd5feda8a 100644
--- a/tools/python/profile_extensions.golden.license
+++ b/tools/python/profile_extensions.golden.license
@@ -1,2 +1,2 @@
-# SPDX-FileCopyrightText: 2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv-software-src/riscv-unified-db>
+# SPDX-FileCopyrightText: 2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv/riscv-unified-db>
 # SPDX-License-Identifier: BSD-3-Clause-Clear
diff --git a/tools/ruby-gems/idl_highlighter/idl_highlighter.gemspec b/tools/ruby-gems/idl_highlighter/idl_highlighter.gemspec
index 51f2f8ddb7..f679601769 100644
--- a/tools/ruby-gems/idl_highlighter/idl_highlighter.gemspec
+++ b/tools/ruby-gems/idl_highlighter/idl_highlighter.gemspec
@@ -3,7 +3,7 @@
 
 # frozen_string_literal: true
 
-require_relative 'lib/idl_highlighter/version'
+require_relative "lib/idl_highlighter/version"
 
 Gem::Specification.new do |s|
   s.name        = "idl_highlighter"
@@ -17,14 +17,14 @@ Gem::Specification.new do |s|
   s.date        = Time.now.strftime("%Y-%m-%d")
   s.authors     = ["Derek Hower"]
   s.email       = ["dhower@qti.qualcomm.com"]
-  s.homepage    = "https://github.com/riscv-software-src/riscv-unified-db"
+  s.homepage    = "https://github.com/riscv/riscv-unified-db"
   s.platform    = Gem::Platform::RUBY
   s.files       = Dir["lib/**/*.rb", "LICENSE"]
   s.license     = "BSD-3-Clause-Clear"
   s.metadata    = {
-    "homepage_uri"      => "https://github.com/riscv-software-src/riscv-unified-db",
+    "homepage_uri"      => "https://github.com/riscv/riscv-unified-db",
     "mailing_list_uri"  => "https://lists.riscv.org/g/tech-unifieddb",
-    "bug_tracker_uri"   => "https://github.com/riscv-software-src/riscv-unified-db/issues"
+    "bug_tracker_uri"   => "https://github.com/riscv/riscv-unified-db/issues"
   }
   s.required_ruby_version = "~> 3.2"
 
diff --git a/tools/ruby-gems/idlc/REUSE.toml b/tools/ruby-gems/idlc/REUSE.toml
index 2f54deaf8d..c1727bc210 100644
--- a/tools/ruby-gems/idlc/REUSE.toml
+++ b/tools/ruby-gems/idlc/REUSE.toml
@@ -7,7 +7,7 @@ path = [
 ]
 SPDX-FileCopyrightText = [
     "Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.",
-    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv-software-src/riscv-unified-db>",
+    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv/riscv-unified-db>",
 ]
 SPDX-License-Identifier = "BSD-3-Clause-Clear"
 
diff --git a/tools/ruby-gems/idlc/idlc.gemspec b/tools/ruby-gems/idlc/idlc.gemspec
index ae151d4a0d..3e45f43ee1 100644
--- a/tools/ruby-gems/idlc/idlc.gemspec
+++ b/tools/ruby-gems/idlc/idlc.gemspec
@@ -18,14 +18,14 @@ Gem::Specification.new do |s|
   s.date        = Time.now.strftime("%Y-%m-%d")
   s.authors     = ["Derek Hower"]
   s.email       = ["dhower@qti.qualcomm.com"]
-  s.homepage    = "https://github.com/riscv-software-src/riscv-unified-db"
+  s.homepage    = "https://github.com/riscv/riscv-unified-db"
   s.platform    = Gem::Platform::RUBY
   s.files       = Dir["lib/**/*.rb", "LICENSE"]
   s.license     = "BSD-3-Clause-Clear"
   s.metadata    = {
-    "homepage_uri" => "https://github.com/riscv-software-src/riscv-unified-db",
+    "homepage_uri" => "https://github.com/riscv/riscv-unified-db",
     "mailing_list_uri" => "https://lists.riscv.org/g/tech-unifieddb",
-    "bug_tracker_uri" => "https://github.com/riscv-software-src/riscv-unified-db/issues"
+    "bug_tracker_uri" => "https://github.com/riscv/riscv-unified-db/issues"
   }
   s.required_ruby_version = "~> 3.2"
 
diff --git a/tools/ruby-gems/udb-gen/lib/udb-gen/subcommands/isa-explorer.rb b/tools/ruby-gems/udb-gen/lib/udb-gen/subcommands/isa-explorer.rb
index 5521f40774..30f08aadb2 100644
--- a/tools/ruby-gems/udb-gen/lib/udb-gen/subcommands/isa-explorer.rb
+++ b/tools/ruby-gems/udb-gen/lib/udb-gen/subcommands/isa-explorer.rb
@@ -48,7 +48,7 @@ def self.arch2ext_table(arch, skip)
           { name: "Extension Name", formatter: "link", sorter: "alphanum", headerFilter: true, frozen: true, formatterParams:
             {
             labelField: "Extension Name",
-            urlPrefix: "https://riscv-software-src.github.io/riscv-unified-db/manual/html/isa/isa_20240411/exts/"
+            urlPrefix: "https://riscv.github.io/riscv-unified-db/manual/html/isa/isa_20240411/exts/"
             }
           },
           { name: "Description", formatter: "textarea", sorter: "alphanum", headerFilter: true },
@@ -133,7 +133,7 @@ def self.arch2inst_table(arch, skip)
           { name: "Instruction Name", formatter: "link", sorter: "alphanum", headerFilter: true, frozen: true, formatterParams:
             {
             labelField: "Instruction Name",
-            urlPrefix: "https://riscv-software-src.github.io/riscv-unified-db/manual/html/isa/isa_20240411/insts/"
+            urlPrefix: "https://riscv.github.io/riscv-unified-db/manual/html/isa/isa_20240411/insts/"
             }
           },
           { name: "Description", formatter: "textarea", sorter: "alphanum", headerFilter: true },
@@ -184,7 +184,7 @@ def self.arch2csr_table(arch, skip)
           { name: "CSR Name", formatter: "link", sorter: "alphanum", headerFilter: true, frozen: true, formatterParams:
             {
             labelField: "CSR Name",
-            urlPrefix: "https://riscv-software-src.github.io/riscv-unified-db/manual/html/isa/isa_20240411/csrs/"
+            urlPrefix: "https://riscv.github.io/riscv-unified-db/manual/html/isa/isa_20240411/csrs/"
             }
           },
           { name: "Address", formatter: "textarea", sorter: "number", headerFilter: true },
diff --git a/tools/ruby-gems/udb-gen/templates/ext-doc/ext_pdf.adoc.erb b/tools/ruby-gems/udb-gen/templates/ext-doc/ext_pdf.adoc.erb
index d98e894baf..bee4738e9a 100644
--- a/tools/ruby-gems/udb-gen/templates/ext-doc/ext_pdf.adoc.erb
+++ b/tools/ruby-gems/udb-gen/templates/ext-doc/ext_pdf.adoc.erb
@@ -126,7 +126,7 @@ This document is released under the <%= primary_ext.doc_license.nil? ? "unknown"
 
 Copyright <%= ext_reqs[0].satisfying_versions.max.ratification_date.nil? ? Date.today.year : ext_reqs[0].satisfying_versions.max.ratification_date.split("-")[0] %> by <%= primary_ext.company.nil? ? "unknown" : primary_ext.company.name %>.
 
-This document was created using the http://github.org/riscv-software-src/riscv-unified-db[RISC-V Unified Database] at commit <%= `git rev-parse HEAD` %>.
+This document was created using the http://github.org/riscv/riscv-unified-db[RISC-V Unified Database] at commit <%= `git rev-parse HEAD` %>.
 
 [preface]
 == Acknowledgements
diff --git a/tools/ruby-gems/udb-gen/templates/manual/index.adoc.erb b/tools/ruby-gems/udb-gen/templates/manual/index.adoc.erb
index ad4248b455..45eb0304fd 100644
--- a/tools/ruby-gems/udb-gen/templates/manual/index.adoc.erb
+++ b/tools/ruby-gems/udb-gen/templates/manual/index.adoc.erb
@@ -12,5 +12,5 @@ The following versions of the manual can be found here:
 * xref:<%= version_name %>@<%= manual.name %>:ROOT:index.adoc[Version <%= version.marketing_version %>]
 <%- end -%>
 
-This site was generated using https://github.com/riscv-software-src/riscv-unified-db[riscv-unified-db],
+This site was generated using https://github.com/riscv/riscv-unified-db[riscv-unified-db],
 at commit <%= `git rev-parse HEAD` %><%- unless `git status -s`.empty? -%>, with dirty state<%- end -%>.
diff --git a/tools/ruby-gems/udb-gen/udb-gen.gemspec b/tools/ruby-gems/udb-gen/udb-gen.gemspec
index f08bd3677d..2ecc16517f 100644
--- a/tools/ruby-gems/udb-gen/udb-gen.gemspec
+++ b/tools/ruby-gems/udb-gen/udb-gen.gemspec
@@ -16,14 +16,14 @@ Gem::Specification.new do |s|
   s.date        = Time.now.strftime("%Y-%m-%d")
   s.authors     = ["Derek Hower"]
   s.email       = ["dhower@qti.qualcomm.com"]
-  s.homepage    = "https://github.com/riscv-software-src/riscv-unified-db"
+  s.homepage    = "https://github.com/riscv/riscv-unified-db"
   s.platform    = Gem::Platform::RUBY
   s.files       = Dir["lib/**/*.rb", "templates/*.erb", "LICENSE"]
   s.license     = "BSD-3-Clause-Clear"
   s.metadata    = {
-    "homepage_uri" => "https://github.com/riscv-software-src/riscv-unified-db",
+    "homepage_uri" => "https://github.com/riscv/riscv-unified-db",
     "mailing_list_uri" => "https://lists.riscv.org/g/tech-unifieddb",
-    "bug_tracker_uri" => "https://github.com/riscv-software-src/riscv-unified-db/issues"
+    "bug_tracker_uri" => "https://github.com/riscv/riscv-unified-db/issues"
   }
   s.required_ruby_version = "~> 3.2"
 
diff --git a/tools/ruby-gems/udb/REUSE.toml b/tools/ruby-gems/udb/REUSE.toml
index 9b0211aee2..afd24f933d 100644
--- a/tools/ruby-gems/udb/REUSE.toml
+++ b/tools/ruby-gems/udb/REUSE.toml
@@ -7,7 +7,7 @@ path = [
 ]
 SPDX-FileCopyrightText = [
     "Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.",
-    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv-software-src/riscv-unified-db>",
+    "2024-2025 Contributors to the RISCV UnifiedDB <https://github.com/riscv/riscv-unified-db>",
 ]
 SPDX-License-Identifier = "BSD-3-Clause-Clear"
 
diff --git a/tools/ruby-gems/udb/lib/udb/obj/portfolio.rb b/tools/ruby-gems/udb/lib/udb/obj/portfolio.rb
index 8999d1abe7..7b2e2791bc 100644
--- a/tools/ruby-gems/udb/lib/udb/obj/portfolio.rb
+++ b/tools/ruby-gems/udb/lib/udb/obj/portfolio.rb
@@ -745,7 +745,7 @@ def in_scope_csrs(design)
 
     # @param design [Design] The design
     # @return [Array<ExceptionCode>] Unsorted list of all in-scope exception codes.
-    # TODO: See https://github.com/riscv-software-src/riscv-unified-db/issues/291
+    # TODO: See https://github.com/riscv/riscv-unified-db/issues/291
     # TODO: Still needs work and haven't created in_scope_interrupt_codes yet.
     # TODO: Extensions should provide conditional information ("when" statements?)
     #       that we evaluate here to determine if a particular exception code can
diff --git a/tools/ruby-gems/udb/test/test_yaml_loader.rb b/tools/ruby-gems/udb/test/test_yaml_loader.rb
index 1c08717aa3..40e6ed7256 100644
--- a/tools/ruby-gems/udb/test/test_yaml_loader.rb
+++ b/tools/ruby-gems/udb/test/test_yaml_loader.rb
@@ -328,7 +328,7 @@ def test_that_invalid_inherits_raise
     assert_nil doc
   end
 
-  # Commented out until https://github.com/riscv-software-src/riscv-unified-db/issues/369 is fixed.
+  # Commented out until https://github.com/riscv/riscv-unified-db/issues/369 is fixed.
 #   def test_copy_in_the_same_document
 #     yaml = <<~YAML
 #       $defs:
diff --git a/tools/ruby-gems/udb/udb.gemspec b/tools/ruby-gems/udb/udb.gemspec
index af6dbf19b9..5eef7c07b8 100644
--- a/tools/ruby-gems/udb/udb.gemspec
+++ b/tools/ruby-gems/udb/udb.gemspec
@@ -18,14 +18,14 @@ Gem::Specification.new do |s|
   s.date        = Time.now.strftime("%Y-%m-%d")
   s.authors     = ["Derek Hower", "James Ball"]
   s.email       = ["dhower@qti.qualcomm.com", "jamesball@qti.qualcomm.com"]
-  s.homepage    = "https://github.com/riscv-software-src/riscv-unified-db"
+  s.homepage    = "https://github.com/riscv/riscv-unified-db"
   s.platform    = Gem::Platform::RUBY
   s.files       = Dir["lib/**/*.rb", "LICENSE"]
   s.license     = "BSD-3-Clause-Clear"
   s.metadata    = {
-    "homepage_uri" => "https://github.com/riscv-software-src/riscv-unified-db",
+    "homepage_uri" => "https://github.com/riscv/riscv-unified-db",
     "mailing_list_uri" => "https://lists.riscv.org/g/tech-unifieddb",
-    "bug_tracker_uri" => "https://github.com/riscv-software-src/riscv-unified-db/issues"
+    "bug_tracker_uri" => "https://github.com/riscv/riscv-unified-db/issues"
   }
   s.required_ruby_version = "~> 3.2"
 
diff --git a/tools/ruby-gems/udb_helpers/udb_helpers.gemspec b/tools/ruby-gems/udb_helpers/udb_helpers.gemspec
index e1f533a475..5aecaa5062 100644
--- a/tools/ruby-gems/udb_helpers/udb_helpers.gemspec
+++ b/tools/ruby-gems/udb_helpers/udb_helpers.gemspec
@@ -15,14 +15,14 @@ Gem::Specification.new do |s|
   s.date        = Time.now.strftime("%Y-%m-%d")
   s.authors     = ["Derek Hower", "James Ball", "Afonso Olivera"]
   s.email       = ["dhower@qti.qualcomm.com", "jamesball@qti.qualcomm.com", "Afonso.Oliveira@synopsys.com"]
-  s.homepage    = "https://github.com/riscv-software-src/riscv-unified-db"
+  s.homepage    = "https://github.com/riscv/riscv-unified-db"
   s.platform    = Gem::Platform::RUBY
   s.files       = Dir["lib/**/*.rb", "LICENSE"]
   s.license     = "BSD-3-Clause-Clear"
   s.metadata    = {
-    "homepage_uri"      => "https://github.com/riscv-software-src/riscv-unified-db",
+    "homepage_uri"      => "https://github.com/riscv/riscv-unified-db",
     "mailing_list_uri"  => "https://lists.riscv.org/g/tech-unifieddb",
-    "bug_tracker_uri"   => "https://github.com/riscv-software-src/riscv-unified-db/issues"
+    "bug_tracker_uri"   => "https://github.com/riscv/riscv-unified-db/issues"
   }
   s.required_ruby_version = "~> 3.2"
 
diff --git a/tools/scripts/pages.html.erb b/tools/scripts/pages.html.erb
index 3999422e10..ab72f638cc 100644
--- a/tools/scripts/pages.html.erb
+++ b/tools/scripts/pages.html.erb
@@ -156,7 +156,7 @@
           <h3>Repository development</h3>
           <ul>
             <li><a href="<%= ENV["PAGES_URL"] %>/reuse_bom.txt">Reuse Bill of Materials (LICENSE information)</a></li>
-            <li><a href="https://dhower-qc.github.io/udb-stats/riscv-software-src/riscv-unified-db/latest-report/report.html">GitHub usage statistics (updated daily)</a></li>
+            <li><a href="https://dhower-qc.github.io/udb-stats/riscv/riscv-unified-db/latest-report/report.html">GitHub usage statistics (updated daily)</a></li>
           </ul>
         </div>
       </div>
